{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7e64e85a-de23-4240-8c9b-fce7230854b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from collections import defaultdict\n",
    "from copy import deepcopy\n",
    "import time\n",
    "import random\n",
    "from scipy import stats\n",
    "from scipy.stats import mode\n",
    "import seaborn as sns\n",
    "import warnings\n",
    "from collections import defaultdict\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "from scipy.stats import chisquare"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "512b422e-0884-4607-a1e5-1c3749842bbb",
   "metadata": {},
   "source": [
    "## Hits where antibiotic resistance evolves first\n",
    "#### Exclude hits with two non-related antibiotics -> these go in co-linearity section"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a917b7e8-c7e6-4456-9123-e56b03a58053",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge sequential and simultaneous hits into a single table\n",
    "\n",
    "all_hits_sequential = pd.read_csv(\"output/04A.pair_annotation/results_allpairs_annotation_withcategories.csv\", index_col=0)\n",
    "all_simultaneous = pd.read_csv(\"output/04B.pair_annotation/results_allpairs_annotation_withcategories.csv\", index_col=0)\n",
    "\n",
    "sequential_merged = all_hits_sequential.merge(all_simultaneous[[\"position_i\", \"position_j\", \"pval_beta_i_on_j\", 'pval_beta_i_on_j_BH_sig']], \n",
    "                                                on=[\"position_i\", \"position_j\"], how=\"left\", suffixes=[\"_sequential\", \"_simultaneous\"])\n",
    "\n",
    "simultaneous_merged = all_simultaneous.merge(all_hits_sequential[[\"position_i\", \"position_j\", \"pval_beta_i_on_j\", 'pval_beta_i_on_j_BH_sig']], \n",
    "                                                on=[\"position_i\", \"position_j\"], how=\"left\", suffixes=[ \"_simultaneous\", \"_sequential\"])\n",
    "\n",
    "all_hits = pd.concat([simultaneous_merged, sequential_merged])\n",
    "all_hits = all_hits.drop_duplicates(subset=[\"position_i\", \"position_j\"]).reset_index(drop=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "e3b33ce7-de51-400d-87cc-c8ac15e12fa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "display_positions =['position_i', 'gene_name_i', 'position_j', 'gene_name_j',\n",
    "       'n_mutations_i', 'n_mutations_j', 'n_branches_with_i',\n",
    "    'n_muts_i_on_branch_j', 'n_muts_same_branch','pval_beta_i_on_j',\n",
    "       'pval_beta_i_on_j_BH_sig']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "56388898-58d4-47d7-89c8-3fd6f8c53df7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position_i</th>\n",
       "      <th>gene_name_i</th>\n",
       "      <th>position_j</th>\n",
       "      <th>gene_name_j</th>\n",
       "      <th>n_mutations_i</th>\n",
       "      <th>n_mutations_j</th>\n",
       "      <th>n_branches_with_i</th>\n",
       "      <th>n_muts_i_on_branch_j</th>\n",
       "      <th>n_muts_same_branch</th>\n",
       "      <th>pval_beta_i_on_j</th>\n",
       "      <th>pval_beta_i_on_j_BH_sig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16077</th>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>2726142</td>\n",
       "      <td>None</td>\n",
       "      <td>1811</td>\n",
       "      <td>39</td>\n",
       "      <td>12564</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0.180799</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       position_i gene_name_i  position_j gene_name_j  n_mutations_i  \\\n",
       "16077     2155168        katG     2726142        None           1811   \n",
       "\n",
       "       n_mutations_j  n_branches_with_i  n_muts_i_on_branch_j  \\\n",
       "16077             39              12564                     0   \n",
       "\n",
       "       n_muts_same_branch  pval_beta_i_on_j  pval_beta_i_on_j_BH_sig  \n",
       "16077                   3          0.180799                    False  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_simultaneous.query(\"position_j==2726142 and position_i==2155168\")[display_positions]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2c5c1db3-2ee1-4daa-9a15-94b0d0240e15",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "961\n"
     ]
    }
   ],
   "source": [
    "known_antibiotic_first = all_hits.dropna(subset=[\"known_drug_j\"]).query(\"pval_beta_i_on_j_BH_sig_sequential or pval_beta_i_on_j_BH_sig_simultaneous\")\n",
    "\n",
    "known_antibiotic_first['num_dependent'] = known_antibiotic_first.n_muts_i_on_branch_j + known_antibiotic_first.n_muts_same_branch\n",
    "know_ab_first = known_antibiotic_first.loc[known_antibiotic_first.possible_drug_i.isnull(), :]\n",
    "know_ab_first.sort_values('num_dependent', ascending=False).head(40)\n",
    "print(len(know_ab_first))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "eef7949c-fc61-47ad-aa93-2856efad3d7e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2043\n"
     ]
    }
   ],
   "source": [
    "def _process_known_hits(input_list):\n",
    "    \"\"\"\n",
    "    splits the possible associated drugs (string separated by ,) into a list\n",
    "    \"\"\"\n",
    "    output_list = []\n",
    "    for item in input_list:\n",
    "        #print(item)\n",
    "        if type(item) != str:\n",
    "            output_list.append(np.nan)\n",
    "        else:\n",
    "            splitted = item.split(\",\")\n",
    "            exclusive= list(set(splitted))\n",
    "            output_list.append(exclusive)\n",
    "    return output_list\n",
    "\n",
    "known_antibiotic_first = all_hits.dropna(subset=[\"known_drug_j\"]).query(\"pval_beta_i_on_j_BH_sig_sequential or pval_beta_i_on_j_BH_sig_simultaneous\")\n",
    "\n",
    "known_antibiotic_first[\"known_drug_i\"] = _process_known_hits(known_antibiotic_first.known_drug_i)\n",
    "known_antibiotic_first[\"possible_drug_i\"] = _process_known_hits(known_antibiotic_first.possible_drug_i)\n",
    "\n",
    "known_antibiotic_first[\"known_drug_j\"] = _process_known_hits(known_antibiotic_first.known_drug_j)\n",
    "known_antibiotic_first[\"possible_drug_j\"] = _process_known_hits(known_antibiotic_first.possible_drug_j)\n",
    "\n",
    "print(len(known_antibiotic_first))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c9917889-1006-4883-9a23-cdd3c497f6da",
   "metadata": {},
   "outputs": [],
   "source": [
    "between_antibiotic_indices = []\n",
    "antibiotic_indices = []\n",
    "\n",
    "known_antibiotic_first[\"within_same_drug\"] = False\n",
    "\n",
    "for idx, row in known_antibiotic_first.iterrows():\n",
    "    \n",
    "    set_of_known_first = set(row.known_drug_j)\n",
    "    set_of_known_second = set()\n",
    "    known_drug_second = False\n",
    "    \n",
    "    if type(row.known_drug_i) is list:\n",
    "        set_of_known_second= set(row.known_drug_i)\n",
    "        known_drug_second = True\n",
    "        \n",
    "    elif type(row.possible_drug_i) is list:\n",
    "        set_of_known_second= set(row.possible_drug_i)\n",
    "        known_drug_second = True\n",
    "        \n",
    "    if not known_drug_second:\n",
    "        antibiotic_indices.append(idx)\n",
    "    \n",
    "    elif len(set_of_known_first.intersection(set_of_known_second)) > 0:\n",
    "        antibiotic_indices.append(idx)\n",
    "        known_antibiotic_first.loc[idx, \"within_same_drug\"] = True\n",
    "        \n",
    "    else:\n",
    "        between_antibiotic_indices.append(idx)\n",
    "        \n",
    "between_antibiotic_hits = known_antibiotic_first.loc[between_antibiotic_indices]\n",
    "antibiotic_hits = known_antibiotic_first.loc[antibiotic_indices]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4de021a5-b3e0-4d0b-a0f6-183fb04b5249",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘output/05.antibiotic’: File exists\n"
     ]
    }
   ],
   "source": [
    "!mkdir output/05.antibiotic\n",
    "antibiotic_hits.to_csv(\"output/05.antibiotic/antibiotic_hits.csv\")\n",
    "between_antibiotic_hits.to_csv(\"output/05.antibiotic/between_antibiotic_hits.csv\")\n",
    "all_hits.to_csv(\"output/05.antibiotic/all_pairs.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87d07b62-5fe3-42f1-a77d-7ed04033d3bf",
   "metadata": {},
   "source": [
    "## Compute per-drug output tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6f2c9bf1-24f6-46bf-955e-67422e094ae0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ETH 46\n",
      "LZD 68\n",
      "STM 215\n",
      "EMB 271\n",
      "CAP 67\n",
      "PZA 129\n",
      "LEV 15\n",
      "MXF 15\n",
      "KAN 101\n",
      "INH 174\n",
      "RIF 152\n",
      "AMI 101\n"
     ]
    }
   ],
   "source": [
    "## These will contain all hits except the between-antibiotic hits\n",
    "\n",
    "## First get the list of all names of drugs in the dataset\n",
    "drug_list = []\n",
    "for idx, row in antibiotic_hits.iterrows():\n",
    "    drugs1 = row.possible_drug_j\n",
    "    drugs2 = row.possible_drug_i\n",
    "    \n",
    "    if type(drugs1) != list:\n",
    "        drugs1= []\n",
    "    if type(drugs2) != list:\n",
    "        drugs2=[]\n",
    "    \n",
    "    for drug in drugs1 + drugs2:\n",
    "        drug_list.append(drug)\n",
    "\n",
    "drug_list = list(set(drug_list))\n",
    "\n",
    "## Then create an empty df for each\n",
    "df_dict = {drug:pd.DataFrame() for drug in drug_list}\n",
    "\n",
    "## now concatenate each row to the appropriate drug dataframe\n",
    "for idx, row in antibiotic_hits.iterrows():\n",
    "    drugs1 = row.possible_drug_j\n",
    "    drugs2 = row.possible_drug_i\n",
    "    \n",
    "    if type(drugs1) != list:\n",
    "        drugs1= []\n",
    "    if type(drugs2) != list:\n",
    "        drugs2=[]\n",
    "    \n",
    "    for drug in drugs1:\n",
    "        df_dict[drug] = df_dict[drug].append(row)\n",
    "        \n",
    "    for drug in drugs2:\n",
    "        df_dict[drug] = df_dict[drug].append(row)\n",
    "        \n",
    "for drug, df in df_dict.items():\n",
    "    print(drug, len(df))\n",
    "    df.sort_values(\"n_muts_i_on_branch_j\", ascending=False).to_csv(f\"output/05.antibiotic/{drug}_antibiotic_hits.csv\")\n",
    "    df.query(\"pval_beta_i_on_j_BH_sig_sequential != 0\").query(\"pval_beta_i_on_j_BH_sig_sequential\").sort_values(\n",
    "        \"n_muts_i_on_branch_j\", ascending=False\n",
    "    ).to_csv(f\"output/05.antibiotic/{drug}_antibiotic_hits_sequential.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "35a53d11-66c4-43e3-a07d-3314adc112a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>position_i</th>\n",
       "      <th>gene_name_i</th>\n",
       "      <th>position_j</th>\n",
       "      <th>gene_name_j</th>\n",
       "      <th>n_mutations_i</th>\n",
       "      <th>n_mutations_j</th>\n",
       "      <th>n_branches_with_i</th>\n",
       "      <th>n_branches_with_j</th>\n",
       "      <th>n_muts_i_on_branch_j</th>\n",
       "      <th>...</th>\n",
       "      <th>in_antigen_epitope_j</th>\n",
       "      <th>in_antigen_protein_j</th>\n",
       "      <th>in_known_ABR_position_j</th>\n",
       "      <th>known_drug_j</th>\n",
       "      <th>in_possible_ABR_gene_j</th>\n",
       "      <th>possible_drug_j</th>\n",
       "      <th>annotation</th>\n",
       "      <th>pval_beta_i_on_j_sequential</th>\n",
       "      <th>pval_beta_i_on_j_BH_sig_sequential</th>\n",
       "      <th>within_same_drug</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>58212</th>\n",
       "      <td>55219</td>\n",
       "      <td>27490</td>\n",
       "      <td>None</td>\n",
       "      <td>1472359</td>\n",
       "      <td>None</td>\n",
       "      <td>6</td>\n",
       "      <td>233</td>\n",
       "      <td>8</td>\n",
       "      <td>870</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[STM]</td>\n",
       "      <td>True</td>\n",
       "      <td>[LZD, STM, CAP, KAN, AMI]</td>\n",
       "      <td>known_AB_first</td>\n",
       "      <td>6.726209e-02</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58228</th>\n",
       "      <td>55236</td>\n",
       "      <td>13638</td>\n",
       "      <td>None</td>\n",
       "      <td>1472359</td>\n",
       "      <td>None</td>\n",
       "      <td>35</td>\n",
       "      <td>233</td>\n",
       "      <td>14</td>\n",
       "      <td>870</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[STM]</td>\n",
       "      <td>True</td>\n",
       "      <td>[LZD, STM, CAP, KAN, AMI]</td>\n",
       "      <td>known_AB_first</td>\n",
       "      <td>7.910976e-02</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58243</th>\n",
       "      <td>55254</td>\n",
       "      <td>1254056</td>\n",
       "      <td>Rv1129c</td>\n",
       "      <td>1472359</td>\n",
       "      <td>None</td>\n",
       "      <td>50</td>\n",
       "      <td>233</td>\n",
       "      <td>42</td>\n",
       "      <td>870</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[STM]</td>\n",
       "      <td>True</td>\n",
       "      <td>[LZD, STM, CAP, KAN, AMI]</td>\n",
       "      <td>known_AB_first</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58270</th>\n",
       "      <td>55281</td>\n",
       "      <td>1338877</td>\n",
       "      <td>None</td>\n",
       "      <td>1472359</td>\n",
       "      <td>None</td>\n",
       "      <td>15</td>\n",
       "      <td>233</td>\n",
       "      <td>10</td>\n",
       "      <td>870</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[STM]</td>\n",
       "      <td>True</td>\n",
       "      <td>[LZD, STM, CAP, KAN, AMI]</td>\n",
       "      <td>known_AB_first</td>\n",
       "      <td>6.937649e-04</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58307</th>\n",
       "      <td>55320</td>\n",
       "      <td>13624</td>\n",
       "      <td>None</td>\n",
       "      <td>1472359</td>\n",
       "      <td>None</td>\n",
       "      <td>12</td>\n",
       "      <td>233</td>\n",
       "      <td>48</td>\n",
       "      <td>870</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[STM]</td>\n",
       "      <td>True</td>\n",
       "      <td>[LZD, STM, CAP, KAN, AMI]</td>\n",
       "      <td>known_AB_first</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>733445</th>\n",
       "      <td>816128</td>\n",
       "      <td>1388042</td>\n",
       "      <td>Rv1245c</td>\n",
       "      <td>1473246</td>\n",
       "      <td>None</td>\n",
       "      <td>6</td>\n",
       "      <td>560</td>\n",
       "      <td>12</td>\n",
       "      <td>1893</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[CAP, KAN, AMI]</td>\n",
       "      <td>True</td>\n",
       "      <td>[LZD, STM, CAP, KAN, AMI]</td>\n",
       "      <td>known_AB_first</td>\n",
       "      <td>1.254120e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>733536</th>\n",
       "      <td>816239</td>\n",
       "      <td>1662236</td>\n",
       "      <td>Rv1473</td>\n",
       "      <td>1473246</td>\n",
       "      <td>None</td>\n",
       "      <td>6</td>\n",
       "      <td>560</td>\n",
       "      <td>16</td>\n",
       "      <td>1893</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[CAP, KAN, AMI]</td>\n",
       "      <td>True</td>\n",
       "      <td>[LZD, STM, CAP, KAN, AMI]</td>\n",
       "      <td>known_AB_first</td>\n",
       "      <td>1.226512e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>790872</th>\n",
       "      <td>1445064</td>\n",
       "      <td>1472359</td>\n",
       "      <td>None</td>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>233</td>\n",
       "      <td>783</td>\n",
       "      <td>870</td>\n",
       "      <td>5303</td>\n",
       "      <td>38</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[STM]</td>\n",
       "      <td>True</td>\n",
       "      <td>[STM]</td>\n",
       "      <td>between_known_AB</td>\n",
       "      <td>3.367794e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>790946</th>\n",
       "      <td>1445155</td>\n",
       "      <td>3568795</td>\n",
       "      <td>None</td>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>15</td>\n",
       "      <td>783</td>\n",
       "      <td>64</td>\n",
       "      <td>5303</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[STM]</td>\n",
       "      <td>True</td>\n",
       "      <td>[STM]</td>\n",
       "      <td>known_AB_first</td>\n",
       "      <td>7.254204e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>791232</th>\n",
       "      <td>1445474</td>\n",
       "      <td>3568759</td>\n",
       "      <td>None</td>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>31</td>\n",
       "      <td>783</td>\n",
       "      <td>62</td>\n",
       "      <td>5303</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[STM]</td>\n",
       "      <td>True</td>\n",
       "      <td>[STM]</td>\n",
       "      <td>known_AB_first</td>\n",
       "      <td>3.249941e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>101 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Unnamed: 0  position_i gene_name_i  position_j gene_name_j  \\\n",
       "58212        55219       27490        None     1472359        None   \n",
       "58228        55236       13638        None     1472359        None   \n",
       "58243        55254     1254056     Rv1129c     1472359        None   \n",
       "58270        55281     1338877        None     1472359        None   \n",
       "58307        55320       13624        None     1472359        None   \n",
       "...            ...         ...         ...         ...         ...   \n",
       "733445      816128     1388042     Rv1245c     1473246        None   \n",
       "733536      816239     1662236      Rv1473     1473246        None   \n",
       "790872     1445064     1472359        None      781687        rpsL   \n",
       "790946     1445155     3568795        None      781687        rpsL   \n",
       "791232     1445474     3568759        None      781687        rpsL   \n",
       "\n",
       "        n_mutations_i  n_mutations_j  n_branches_with_i  n_branches_with_j  \\\n",
       "58212               6            233                  8                870   \n",
       "58228              35            233                 14                870   \n",
       "58243              50            233                 42                870   \n",
       "58270              15            233                 10                870   \n",
       "58307              12            233                 48                870   \n",
       "...               ...            ...                ...                ...   \n",
       "733445              6            560                 12               1893   \n",
       "733536              6            560                 16               1893   \n",
       "790872            233            783                870               5303   \n",
       "790946             15            783                 64               5303   \n",
       "791232             31            783                 62               5303   \n",
       "\n",
       "        n_muts_i_on_branch_j  ...  in_antigen_epitope_j  in_antigen_protein_j  \\\n",
       "58212                      1  ...                 False                 False   \n",
       "58228                      2  ...                 False                 False   \n",
       "58243                      0  ...                 False                 False   \n",
       "58270                      3  ...                 False                 False   \n",
       "58307                      0  ...                 False                 False   \n",
       "...                      ...  ...                   ...                   ...   \n",
       "733445                     3  ...                 False                 False   \n",
       "733536                     3  ...                 False                 False   \n",
       "790872                    38  ...                 False                 False   \n",
       "790946                     9  ...                 False                 False   \n",
       "791232                    10  ...                 False                 False   \n",
       "\n",
       "        in_known_ABR_position_j     known_drug_j  in_possible_ABR_gene_j  \\\n",
       "58212                      True            [STM]                    True   \n",
       "58228                      True            [STM]                    True   \n",
       "58243                      True            [STM]                    True   \n",
       "58270                      True            [STM]                    True   \n",
       "58307                      True            [STM]                    True   \n",
       "...                         ...              ...                     ...   \n",
       "733445                     True  [CAP, KAN, AMI]                    True   \n",
       "733536                     True  [CAP, KAN, AMI]                    True   \n",
       "790872                     True            [STM]                    True   \n",
       "790946                     True            [STM]                    True   \n",
       "791232                     True            [STM]                    True   \n",
       "\n",
       "                  possible_drug_j        annotation  \\\n",
       "58212   [LZD, STM, CAP, KAN, AMI]    known_AB_first   \n",
       "58228   [LZD, STM, CAP, KAN, AMI]    known_AB_first   \n",
       "58243   [LZD, STM, CAP, KAN, AMI]    known_AB_first   \n",
       "58270   [LZD, STM, CAP, KAN, AMI]    known_AB_first   \n",
       "58307   [LZD, STM, CAP, KAN, AMI]    known_AB_first   \n",
       "...                           ...               ...   \n",
       "733445  [LZD, STM, CAP, KAN, AMI]    known_AB_first   \n",
       "733536  [LZD, STM, CAP, KAN, AMI]    known_AB_first   \n",
       "790872                      [STM]  between_known_AB   \n",
       "790946                      [STM]    known_AB_first   \n",
       "791232                      [STM]    known_AB_first   \n",
       "\n",
       "       pval_beta_i_on_j_sequential  pval_beta_i_on_j_BH_sig_sequential  \\\n",
       "58212                 6.726209e-02                               False   \n",
       "58228                 7.910976e-02                               False   \n",
       "58243                          NaN                               False   \n",
       "58270                 6.937649e-04                               False   \n",
       "58307                          NaN                               False   \n",
       "...                            ...                                 ...   \n",
       "733445                1.254120e-04                                True   \n",
       "733536                1.226512e-04                                True   \n",
       "790872                3.367794e-05                                True   \n",
       "790946                7.254204e-09                                True   \n",
       "791232                3.249941e-05                                True   \n",
       "\n",
       "        within_same_drug  \n",
       "58212              False  \n",
       "58228              False  \n",
       "58243              False  \n",
       "58270              False  \n",
       "58307              False  \n",
       "...                  ...  \n",
       "733445             False  \n",
       "733536             False  \n",
       "790872              True  \n",
       "790946              True  \n",
       "791232              True  \n",
       "\n",
       "[101 rows x 38 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "35390a43-ecb2-457c-b08f-b472b4b9c1a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ETH 66\n",
      "BDQ 2\n",
      "LZD 153\n",
      "STM 317\n",
      "CAP 153\n",
      "LEV 154\n",
      "PZA 334\n",
      "EMB 418\n",
      "CFZ 2\n",
      "MXF 154\n",
      "KAN 227\n",
      "INH 205\n",
      "RIF 465\n",
      "AMI 227\n"
     ]
    }
   ],
   "source": [
    "## These will contain all hits except the between-antibiotic hits\n",
    "\n",
    "## First get the list of all names of drugs in the dataset\n",
    "drug_list = []\n",
    "for idx, row in between_antibiotic_hits.iterrows():\n",
    "    drugs1 = row.possible_drug_j\n",
    "    drugs2 = row.possible_drug_i\n",
    "    \n",
    "    if type(drugs1) != list:\n",
    "        drugs1= []\n",
    "    if type(drugs2) != list:\n",
    "        drugs2=[]\n",
    "    \n",
    "    for drug in drugs1 + drugs2:\n",
    "        drug_list.append(drug)\n",
    "\n",
    "drug_list = list(set(drug_list))\n",
    "\n",
    "## Then create an empty df for each\n",
    "df_dict = {drug:pd.DataFrame() for drug in drug_list}\n",
    "\n",
    "## now concatenate each row to the appropriate drug dataframe\n",
    "for idx, row in between_antibiotic_hits.iterrows():\n",
    "    drugs1 = row.possible_drug_j\n",
    "    drugs2 = row.possible_drug_i\n",
    "    \n",
    "    if type(drugs1) != list:\n",
    "        drugs1= []\n",
    "    if type(drugs2) != list:\n",
    "        drugs2=[]\n",
    "    \n",
    "    for drug in drugs1:\n",
    "        df_dict[drug] = df_dict[drug].append(row)\n",
    "        \n",
    "    for drug in drugs2:\n",
    "        df_dict[drug] = df_dict[drug].append(row)\n",
    "        \n",
    "for drug, df in df_dict.items():\n",
    "    print(drug, len(df))\n",
    "    df.sort_values(\"n_muts_i_on_branch_j\", ascending=False).to_csv(f\"output/05.antibiotic/{drug}_between_antibiotic_hits.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a7d5458-73d3-44dd-ae70-f12630fd21cf",
   "metadata": {},
   "source": [
    "## Fraction of resistant strains with a dependent mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a756e317-e72c-4fc1-93a8-dd2c3043a1c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "561\n"
     ]
    }
   ],
   "source": [
    "# Here let strains with a resistance-conferring mutation be defined as resistant\n",
    "# Then count the number of resistant strains that have another mutation\n",
    "\n",
    "snps = pd.read_csv(\"output/03.annotation/snps_with_all_annotation.csv\", index_col=0)\n",
    "snps = snps.query(\"not is_lineage\")\n",
    "\n",
    "# list of resistance mutations\n",
    "resistance_mutations = snps.query(\"in_known_ABR_position\")\n",
    "resistance_mutations[\"known_drug\"] = _process_known_hits(resistance_mutations.known_drug)\n",
    "\n",
    "antibiotic_hits = antibiotic_hits.query(\"pval_beta_i_on_j_BH_sig_sequential\")\n",
    "print(len(antibiotic_hits))\n",
    "# list of all other mutations\n",
    "other_positions = list(set(antibiotic_hits.position_i))\n",
    "other_mutations = snps.query(\"pos in @other_positions\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "52f41be3-a205-4f79-b4ff-cf494cb7d04b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Read in the matrix of isolate x genotype\n",
    "# # Input information from Roger\n",
    "genotype_translator = {\n",
    "0:\"A\", 1:\"C\", 2:\"G\", 3:\"T\", 9:\"N\"\n",
    "}\n",
    "isolate_annotation_file = \"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_isolate_annotation.pkl\"\n",
    "genotype_array_file = \"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_matrix.npy\"\n",
    "snp_annotation_file = \"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_SNP_annotation.pkl\"\n",
    "genotype_translator_r = {y:x for x,y in genotype_translator.items()}\n",
    "\n",
    "isolates = pd.read_pickle(isolate_annotation_file)\n",
    "genotypes = np.load(genotype_array_file)\n",
    "snps_older = pd.read_pickle(snp_annotation_file)\n",
    "# # Filter out the snps we don't need\n",
    "snps_to_retain = snps_older.query(\"pos in @other_positions or pos in @resistance_mutations.pos\")\n",
    "indices_to_keep = snps_to_retain.index\n",
    "\n",
    "g = genotypes[indices_to_keep, :]\n",
    "del genotypes\n",
    "np.save(\"output/05.antibiotic/antibiotic_genotypes_matrix.npy\", g)\n",
    "g = np.load(\"output/05.antibiotic/antibiotic_genotypes_matrix.npy\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "521fb9db-79cd-4a92-ab98-96f35d08d4e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "snps_to_retain = snps_to_retain.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "034d5293-f18c-417a-bb36-4e1227d56692",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "no file for BDQ\n",
      "no file for CFZ\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug</th>\n",
       "      <th>num_R_strains</th>\n",
       "      <th>num_R_strains_with_dependency</th>\n",
       "      <th>fraction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ETH</td>\n",
       "      <td>5744</td>\n",
       "      <td>1513</td>\n",
       "      <td>0.263405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LZD</td>\n",
       "      <td>109</td>\n",
       "      <td>53</td>\n",
       "      <td>0.486239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STM</td>\n",
       "      <td>10967</td>\n",
       "      <td>8707</td>\n",
       "      <td>0.793927</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CAP</td>\n",
       "      <td>3194</td>\n",
       "      <td>1153</td>\n",
       "      <td>0.360989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LEV</td>\n",
       "      <td>3153</td>\n",
       "      <td>297</td>\n",
       "      <td>0.094196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PZA</td>\n",
       "      <td>9563</td>\n",
       "      <td>3979</td>\n",
       "      <td>0.416083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>EMB</td>\n",
       "      <td>7235</td>\n",
       "      <td>3464</td>\n",
       "      <td>0.478784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MXF</td>\n",
       "      <td>3153</td>\n",
       "      <td>297</td>\n",
       "      <td>0.094196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>KAN</td>\n",
       "      <td>4410</td>\n",
       "      <td>1916</td>\n",
       "      <td>0.434467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>INH</td>\n",
       "      <td>11143</td>\n",
       "      <td>6236</td>\n",
       "      <td>0.559634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>RIF</td>\n",
       "      <td>10083</td>\n",
       "      <td>5092</td>\n",
       "      <td>0.505008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AMI</td>\n",
       "      <td>3431</td>\n",
       "      <td>1394</td>\n",
       "      <td>0.406296</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   drug  num_R_strains  num_R_strains_with_dependency  fraction\n",
       "0   ETH           5744                           1513  0.263405\n",
       "1   LZD            109                             53  0.486239\n",
       "2   STM          10967                           8707  0.793927\n",
       "3   CAP           3194                           1153  0.360989\n",
       "4   LEV           3153                            297  0.094196\n",
       "5   PZA           9563                           3979  0.416083\n",
       "6   EMB           7235                           3464  0.478784\n",
       "7   MXF           3153                            297  0.094196\n",
       "8   KAN           4410                           1916  0.434467\n",
       "9   INH          11143                           6236  0.559634\n",
       "10  RIF          10083                           5092  0.505008\n",
       "11  AMI           3431                           1394  0.406296"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For each drug, get the set of mutations we are looking for\n",
    "drug_to_sensitive_sites = {drug:[] for drug in drug_list}\n",
    "for idx, row in resistance_mutations.iterrows():\n",
    "    for drug in row.known_drug:\n",
    "        drug_to_sensitive_sites[drug].append((row.pos, row.ref))\n",
    "\n",
    "# Find strains that have any of those mutations\n",
    "drug_to_indices_R_strains = {}\n",
    "\n",
    "_num_sequential_dependencies = []\n",
    "for drug, list_of_tuples in drug_to_sensitive_sites.items():\n",
    "    \n",
    "    has_sensitive_alleles = np.ones(shape=(len(list_of_tuples), g.shape[1]))\n",
    "    #print(drug)\n",
    "    \n",
    "    for idx, (pos, nt) in enumerate(list_of_tuples):\n",
    "    \n",
    "        matrix_index = snps_to_retain.query(\"pos==@pos\").index\n",
    "        has_sensitive_alleles[idx,:] = g[matrix_index, :] == genotype_translator_r[nt]\n",
    "        \n",
    "    #print(np.sum(np.sum(has_sensitive_alleles, axis=0) < has_sensitive_alleles.shape[0]))\n",
    "    indices_R_strains = np.where(np.sum(has_sensitive_alleles, axis=0) < has_sensitive_alleles.shape[0])[0]\n",
    "    drug_to_indices_R_strains[drug] = indices_R_strains\n",
    "    \n",
    "    try:\n",
    "        drug_specific_dependencies = pd.read_csv(f\"output/05.antibiotic/{drug}_antibiotic_hits.csv\", index_col=0)\n",
    "    except:\n",
    "        print(\"no file for\", drug)\n",
    "        continue\n",
    "    drug_specific_dependencies = drug_specific_dependencies.query(\n",
    "        \"pval_beta_i_on_j_BH_sig_sequential\")\n",
    "    \n",
    "    # Get the dependent positions - exclude any known resistance mutation from dependent sites\n",
    "    dependent_positions = set(drug_specific_dependencies.position_i) - set(resistance_mutations.pos)\n",
    "    dependencies = snps_to_retain.query(\"pos in @dependent_positions\")\n",
    "    #print(drug, \"has\", len(dependencies), \"dependencies excluding known R positions\")\n",
    "    assert(len(dependent_positions) == len(dependencies))\n",
    "    \n",
    "    # Create a matrix of which strains have a dependent mutation\n",
    "    _dependency_df = []\n",
    "    has_dependent_alleles = np.ones(shape=(len(dependencies), len(indices_R_strains)))\n",
    "    for mat_idx, (idx, row) in enumerate(dependencies.iterrows()):\n",
    "        has_dependency = g[:, indices_R_strains][idx, :] != genotype_translator_r[row.ref]\n",
    "        total_has_dependency = g[idx, :] != genotype_translator_r[row.ref]\n",
    "        _dependency_df.append([idx, row.pos, row.ref, np.sum(has_dependency), np.sum(total_has_dependency)])\n",
    "        has_dependent_alleles[mat_idx, :] = has_dependency\n",
    "    \n",
    "    dependency_df = pd.DataFrame(_dependency_df, columns=[\"matrix_idx\", \"pos\", \"ref\", \"num_R_strains_with_dependency\", \"total_strains_with_mutation\"])\n",
    "    dependency_df.to_csv(f\"output/05.antibiotic/strains_with_dependencies_per_AB/{drug}.csv\")\n",
    "    \n",
    "    _num_sequential_dependencies.append([drug, len(indices_R_strains), np.sum(np.sum(has_dependent_alleles, axis=0) > 0)])\n",
    "\n",
    "df = pd.DataFrame(_num_sequential_dependencies, columns=[\"drug\", \"num_R_strains\", \"num_R_strains_with_dependency\"])\n",
    "df[\"fraction\"] = df.num_R_strains_with_dependency / df.num_R_strains\n",
    "df.to_csv(\"output/05.antibiotic/num_strains_with_dependency.csv\")    \n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f1c71a71-c3c0-49eb-b1d3-9a93c00378d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1002\n",
      "992\n"
     ]
    }
   ],
   "source": [
    "# Here let strains with a resistance-conferring mutation be defined as resistant\n",
    "# Then count the number of resistant strains that have another mutation\n",
    "#include simultaenous and sequential\n",
    "\n",
    "snps = pd.read_csv(\"output/03.annotation/snps_with_all_annotation.csv\", index_col=0)\n",
    "snps = snps.query(\"not is_lineage\")\n",
    "\n",
    "# list of resistance mutations\n",
    "resistance_mutations = snps.query(\"in_known_ABR_position\")\n",
    "resistance_mutations[\"known_drug\"] = _process_known_hits(resistance_mutations.known_drug)\n",
    "antibiotic_hits = pd.read_csv(\"output/05.antibiotic/antibiotic_hits.csv\")\n",
    "print(len(antibiotic_hits))\n",
    "# list of all other mutations\n",
    "antibiotic_hits[\"dist\"] = np.abs(antibiotic_hits.position_j - antibiotic_hits.position_i)\n",
    "antibiotic_hits = antibiotic_hits.query(\"pval_beta_i_on_j_BH_sig_sequential or (pval_beta_i_on_j_BH_sig_simultaneous and dist >100)\")\n",
    "print(len(antibiotic_hits))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2e65d604-0c8e-44fa-a219-3c3a31f80021",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "other_positions = list(set(antibiotic_hits.position_i))\n",
    "other_mutations = snps.query(\"pos in @other_positions\")\n",
    "\n",
    "# # Read in the matrix of isolate x genotype\n",
    "# # Input information from Roger\n",
    "genotype_translator = {\n",
    "0:\"A\", 1:\"C\", 2:\"G\", 3:\"T\", 9:\"N\"\n",
    "}\n",
    "isolate_annotation_file = \"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_isolate_annotation.pkl\"\n",
    "genotype_array_file = \"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_matrix.npy\"\n",
    "snp_annotation_file = \"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_SNP_annotation.pkl\"\n",
    "genotype_translator_r = {y:x for x,y in genotype_translator.items()}\n",
    "\n",
    "isolates = pd.read_pickle(isolate_annotation_file)\n",
    "genotypes = np.load(genotype_array_file)\n",
    "snps_older = pd.read_pickle(snp_annotation_file)\n",
    "# # Filter out the snps we don't need\n",
    "snps_to_retain = snps_older.query(\"pos in @other_positions or pos in @resistance_mutations.pos\")\n",
    "indices_to_keep = snps_to_retain.index\n",
    "\n",
    "g = genotypes[indices_to_keep, :]\n",
    "del genotypes\n",
    "np.save(\"output/05.antibiotic/antibiotic_genotypes_matrix.npy\", g)\n",
    "g = np.load(\"output/05.antibiotic/antibiotic_genotypes_matrix.npy\")\n",
    "\n",
    "snps_to_retain = snps_to_retain.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2df64579-2a05-4ccb-8f52-3c1acc18b903",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "no file for BDQ\n",
      "no file for CFZ\n"
     ]
    }
   ],
   "source": [
    "# For each drug, get the set of mutations we are looking for\n",
    "drug_to_sensitive_sites = {drug:[] for drug in drug_list}\n",
    "for idx, row in resistance_mutations.iterrows():\n",
    "    for drug in row.known_drug:\n",
    "        drug_to_sensitive_sites[drug].append((row.pos, row.ref))\n",
    "\n",
    "# Find strains that have any of those mutations\n",
    "drug_to_indices_R_strains = {}\n",
    "\n",
    "_num_sequential_dependencies = []\n",
    "for drug, list_of_tuples in drug_to_sensitive_sites.items():\n",
    "    \n",
    "    has_sensitive_alleles = np.ones(shape=(len(list_of_tuples), g.shape[1]))\n",
    "    #print(drug)\n",
    "    \n",
    "    # Determine if each isolate has any sensitive allele\n",
    "    for idx, (pos, nt) in enumerate(list_of_tuples):\n",
    "    \n",
    "        matrix_index = snps_to_retain.query(\"pos==@pos\").index\n",
    "        has_sensitive_alleles[idx,:] = g[matrix_index, :] == genotype_translator_r[nt]\n",
    "        \n",
    "    #print(np.sum(np.sum(has_sensitive_alleles, axis=0) < has_sensitive_alleles.shape[0]))\n",
    "    indices_R_strains = np.where(np.sum(has_sensitive_alleles, axis=0) < has_sensitive_alleles.shape[0])[0]\n",
    "    drug_to_indices_R_strains[drug] = indices_R_strains\n",
    "    \n",
    "    try:\n",
    "        drug_specific_dependencies = pd.read_csv(f\"output/05.antibiotic/{drug}_antibiotic_hits.csv\", index_col=0)\n",
    "    except:\n",
    "        print(\"no file for\", drug)\n",
    "        continue\n",
    "    #drug_specific_dependencies = drug_specific_dependencies.query(\"pval_beta_i_on_j_BH_sig_sequential == 'True'\")\n",
    "    \n",
    "    # Get the dependent positions - exclude any known resistance mutation from dependent sites\n",
    "    dependent_positions = set(drug_specific_dependencies.position_i) - set(resistance_mutations.pos)\n",
    "    dependencies = snps_to_retain.query(\"pos in @dependent_positions\")\n",
    "    #print(drug, \"has\", len(dependencies), \"dependencies excluding known R positions\")\n",
    "    assert(len(dependent_positions) == len(dependencies))\n",
    "    \n",
    "    # Create a matrix of which strains have a dependent mutation\n",
    "    _dependency_df = []\n",
    "    has_dependent_alleles = np.ones(shape=(len(dependencies), len(indices_R_strains)))\n",
    "    for mat_idx, (idx, row) in enumerate(dependencies.iterrows()):\n",
    "        has_dependency = g[:, indices_R_strains][idx, :] != genotype_translator_r[row.ref]\n",
    "        total_has_dependency = g[idx, :] != genotype_translator_r[row.ref]\n",
    "        _dependency_df.append([idx, row.pos, row.ref, np.sum(has_dependency), np.sum(total_has_dependency)])\n",
    "        has_dependent_alleles[mat_idx, :] = has_dependency\n",
    "    \n",
    "    dependency_df = pd.DataFrame(_dependency_df, columns=[\"matrix_idx\", \"pos\", \"ref\", \"num_R_strains_with_dependency\", \"total_strains_with_mutation\"])\n",
    "    dependency_df.to_csv(f\"output/05.antibiotic/strains_with_dependencies_per_AB/{drug}.csv\")\n",
    "    \n",
    "    _num_sequential_dependencies.append([drug, len(indices_R_strains), np.sum(np.sum(has_dependent_alleles, axis=0) > 0)])\n",
    "    \n",
    "df = pd.DataFrame(_num_sequential_dependencies, columns=[\"drug\", \"num_R_strains\", \"num_R_strains_with_dependency\"])\n",
    "df[\"fraction\"] = df.num_R_strains_with_dependency / df.num_R_strains\n",
    "df.to_csv(\"output/05.antibiotic/num_strains_with_dependency_simultaneous.csv\")    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1949acd-ce76-40af-96c5-9fc53f0d0ce3",
   "metadata": {},
   "source": [
    "### Plot fraction of resistant strains with a dependent mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "fff17b40-128e-4a8a-9a49-290842f24c0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAmEAAAGpCAYAAADFpuEPAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8qNh9FAAAACXBIWXMAAAsTAAALEwEAmpwYAAAf2ElEQVR4nO3df5xdd13n8debhGJrgSoEkaSFWiO1KhQaqm6pFBFJa9lQYSEB6YJiHkUriOJSfy5uQQqsothCHhVLQV2CWrCxhlbFVqIIJNDwI8WyoWAbIksKCLa2lrSf/eOcgdubmWTa5tzvncnr+XjMI/ec87133pm5c+c933PuOakqJEmSNFn3ax1AkiTpUGQJkyRJasASJkmS1IAlTJIkqQFLmCRJUgNLWwe4p1avXl1XXnll6xiSJEnzkbk2LLiZsJtvvrl1BEmSpPtswZUwSZKkxcASJkmS1IAlTJIkqQFLmCRJUgOWMEmSpAYsYZIkSQ1YwiRJkhqwhEmSJDVgCZMkSWrAEiZJktSAJUySJKmBQUtYktVJrk+yM8l5s2x/cJK/TPLRJDuSvHDIPJIkSdNisBKWZAlwEXA6cAKwLskJY8N+Friuqh4LnAb8dpLDhsokSZI0LYacCTsZ2FlVN1TVHcBGYM3YmAIemCTAkcCXgL0DZpIkSZoKQ5aw5cBNI8u7+nWjLgS+G9gNfBx4aVXdNf5ASdYn2ZZk2549e4bKK0mSNDFDlrDMsq7Glp8GbAceAZwIXJjkQfvcqeriqlpVVauWLVt2sHNKkiRN3NIBH3sXcPTI8gq6Ga9RLwQuqKoCdib5DHA88KEBc0kL0pYXbG4dYR+nXnpG6wiStGANORO2FViZ5Nj+YPu1wKaxMTcCTwFI8m3Ao4EbBswkSZI0FQabCauqvUnOBa4ClgCXVNWOJOf02zcA5wOXJvk43e7LV1TVzUNlkiRJmhZD7o6kqjYDm8fWbRi5vRv40SEzSJIkTSPPmC9JktSAJUySJKkBS5gkSVIDljBJkqQGLGGSJEkNWMIkSZIasIRJkiQ1YAmTJElqwBImSZLUgCVMkiSpAUuYJElSA5YwSZKkBixhkiRJDVjCJEmSGrCESZIkNWAJkyRJasASJkmS1IAlTJIkqQFLmCRJUgOWMEmSpAYsYZIkSQ1YwiRJkhqwhEmSJDVgCZMkSWrAEiZJktSAJUySJKkBS5gkSVIDljBJkqQGLGGSJEkNWMIkSZIasIRJkiQ1YAmTJElqwBImSZLUgCVMkiSpAUuYJElSA5YwSZKkBgYtYUlWJ7k+yc4k582y/ZeSbO8/PpHkziTfOmQmSZKkaTBYCUuyBLgIOB04AViX5ITRMVX1+qo6sapOBH4Z+Puq+tJQmSRJkqbFkDNhJwM7q+qGqroD2Ais2c/4dcA7BswjSZI0NYYsYcuBm0aWd/Xr9pHkCGA1cNkc29cn2ZZk2549ew56UEmSpEkbsoRllnU1x9inA/84167Iqrq4qlZV1aply5YdtICSJEmtDFnCdgFHjyyvAHbPMXYt7oqUJEmHkCFL2FZgZZJjkxxGV7Q2jQ9K8mDgScDlA2aRJEmaKkuHeuCq2pvkXOAqYAlwSVXtSHJOv31DP/Qs4K+r6tahskiSJE2bwUoYQFVtBjaPrdswtnwpcOmQOSRJkqaNZ8yXJElqwBImSZLUgCVMkiSpAUuYJElSA5YwSZKkBixhkiRJDVjCJEmSGrCESZIkNWAJkyRJasASJkmS1IAlTJIkqQFLmCRJUgOWMEmSpAYsYZIkSQ1YwiRJkhqwhEmSJDVgCZMkSWrAEiZJktSAJUySJKkBS5gkSVIDljBJkqQGLGGSJEkNWMIkSZIasIRJkiQ1YAmTJElqwBImSZLUgCVMkiSpAUuYJElSA5YwSZKkBixhkiRJDVjCJEmSGrCESZIkNWAJkyRJasASJkmS1IAlTJIkqQFLmCRJUgOWMEmSpAYGLWFJVie5PsnOJOfNMea0JNuT7Ejy90PmkSRJmhZLh3rgJEuAi4CnAruArUk2VdV1I2OOAt4ErK6qG5M8bKg8kiRJ02TImbCTgZ1VdUNV3QFsBNaMjXku8K6quhGgqr4wYB5JkqSpMWQJWw7cNLK8q1836ruAb0lyTZIPJzl7tgdKsj7JtiTb9uzZM1BcSZKkyRmyhGWWdTW2vBQ4Cfgx4GnAryf5rn3uVHVxVa2qqlXLli07+EklSZImbLBjwuhmvo4eWV4B7J5lzM1VdStwa5L3AY8FPjVgLkmSpOaGnAnbCqxMcmySw4C1wKaxMZcDpyZZmuQI4PuBTw6YSZIkaSoMNhNWVXuTnAtcBSwBLqmqHUnO6bdvqKpPJrkS+BhwF/CWqvrEUJkkSZKmxZC7I6mqzcDmsXUbxpZfD7x+yBySJEnTxjPmS5IkNWAJkyRJasASJkmS1IAlTJIkqQFLmCRJUgOWMEmSpAYsYZIkSQ1YwiRJkhqwhEmSJDVgCZMkSWrAEiZJktSAJUySJKmBQS/gPSmPu+IRrSPs49ozd7eOIEmSppgzYZIkSQ1YwiRJkhqwhEmSJDVgCZMkSWrAEiZJktSAJUySJKkBS5gkSVIDljBJkqQGLGGSJEkNWMIkSZIasIRJkiQ1YAmTJElqwBImSZLUgCVMkiSpAUuYJElSA5YwSZKkBixhkiRJDVjCJEmSGrCESZIkNWAJkyRJasASJkmS1IAlTJIkqQFLmCRJUgODlrAkq5Ncn2RnkvNm2X5akq8k2d5//MaQeSRJkqbF0qEeOMkS4CLgqcAuYGuSTVV13djQLVV15lA5ptmyZ69oHWFWe/50V+sIkiQtekPOhJ0M7KyqG6rqDmAjsGbAzydJkrRgDFnClgM3jSzv6teN+8EkH03yniTfM2AeSZKkqTHY7kggs6yrseWPAI+sqluSnAH8BbBynwdK1gPrAY455piDHFOSJGnyhpwJ2wUcPbK8Atg9OqCqvlpVt/S3NwP3T/LQ8QeqqouralVVrVq2bNmAkSVJkiZjyBK2FViZ5NgkhwFrgU2jA5I8PEn62yf3eb44YCZJkqSpMNjuyKram+Rc4CpgCXBJVe1Ick6/fQPwLODFSfYCtwFrq2p8l6UkSdKiM+QxYTO7GDePrdswcvtC4MIhM0iSJE0jz5gvSZLUwKAzYZIkSUN42BPObh1hVl/Y+vZ5j3UmTJIkqQFLmCRJUgPujtS98pW3vrl1hH08+IUvbh1BkqR5cyZMkiSpAUuYJElSA5YwSZKkBixhkiRJDVjCJEmSGrCESZIkNWAJkyRJasASJkmS1IAlTJIkqQFLmCRJUgOWMEmSpAYsYZIkSQ0c8ALeSZYBPw08anR8Vf3kcLEkSZIWtwOWMOByYAvwt8Cdw8aRJEk6NMynhB1RVa8YPIkkSdIhZD7HhF2R5IzBk0iSJB1C5lPCXkpXxG5P8u/9x1eHDiZJkrSYHXB3ZFU9cBJBJEmSDiXzOSaMJP8V+KF+8ZqqumK4SJIkSYvffE5RcQHwBOBP+lUvTfLEqjpv0GSS1NiWF2xuHWEfp17qIbrSYjGfmbAzgBOr6i6AJG8DrgUsYZIkSffSfM+Yf9TI7QcPkEOSJOmQMp+ZsNcA1ya5GgjdsWG/PGgqSZKkRW4+7458R5Jr6I4LC/CKqvr80MEkSZIWszl3RyY5vv/38cC3A7uAm4BH9OskSZJ0L+1vJuwXgPXAb8+yrYAfHiSRJEnSIWDOElZV6/ubp1fV7aPbknzToKkkSZIWufm8O/L981wnSZKkeZpzJizJw4HlwOFJHkd3UD7Ag4AjJpBNOuj+8c2PbR1hVqe8+KOtI0iSJmx/x4Q9DXgBsILuuLCZEvZV4FeGjSVJkrS47e+YsLcBb0vyzKq6bIKZJEmSFr35HBN2UpKjZhaSfEuSVw0XSZIkafGbzxnzT6+qr+9+rKovJzkD+LUD3THJauD3gCXAW6rqgjnGPQH4APCcqvrzeSWXtCBM40WwwQthS2pvPjNhS5I8YGYhyeHAA/YzfmbcEuAi4HTgBGBdkhPmGPda4Kr5hpYkSVro5jMT9sfAe5O8le4krT8JvG0e9zsZ2FlVNwAk2QisAa4bG/dzwGV0l0WSJEk6JMzn2pGvS/Jx4Cl075A8v6rmM2u1nO4yRzN2Ad8/OiDJcuAsurPvz1nCkqynO3s/xxxzzDw+tSRJ0nSbz0wYVfUe4D338LEzy7oaW/5duguC35nMNvzrn/9i4GKAVatWjT+GJEnSgnPAEpbkB4DfB74bOIzuIPtbq+pBB7jrLuDokeUVwO6xMauAjX0BeyhwRpK9VfUX80ovSZK0QM1nJuxCYC3wZ3Sl6WzgO+dxv63AyiTHAp/rH+O5owOq6tiZ20kuBa6wgEmSpEPBfHdH7kyypKruBN6a5IDXjqyqvUnOpXvX4xLgkqrakeScfvuG+xJckiRpIZtPCfuPJIcB25O8DvhX4Jvn8+BVtRnYPLZu1vJVVS+Yz2NKkiQtBvM5T9jz+3HnArfSHef1zCFDSZIkLXb7nQnrT6T66qr6CeB24DcnkkqSJGmR2+9MWH8M2LJ+d6QkSZIOkvkcE/ZZ4B+TbKLbHQlAVf3OUKEkSZIWuzlnwpL8UX/zOcAV/dgHjnxIkiTpXtrfTNhJSR4J3Eh3slZJkiQdJPsrYRuAK4FjgW0j60N3+aHvGDCXJEnSojbn7siqemNVfTfw1qr6jpGPY6vKAiZJknQfHPA8YVX14kkEkSRJOpTM52StkiRJOsgsYZIkSQ1YwiRJkhqYz8laJUkLyJYXbG4dYVanXnpG6wjSVHEmTJIkqQFLmCRJUgOWMEmSpAYsYZIkSQ1YwiRJkhqwhEmSJDVgCZMkSWrAEiZJktSAJUySJKkBS5gkSVIDljBJkqQGLGGSJEkNWMIkSZIasIRJkiQ1YAmTJElqwBImSZLUgCVMkiSpAUuYJElSA5YwSZKkBixhkiRJDVjCJEmSGrCESZIkNWAJkyRJasASJkmS1MCgJSzJ6iTXJ9mZ5LxZtq9J8rEk25NsS/LEIfNIkiRNi6VDPXCSJcBFwFOBXcDWJJuq6rqRYe8FNlVVJXkM8KfA8UNlkiRJmhZDzoSdDOysqhuq6g5gI7BmdEBV3VJV1S9+M1BIkiQdAoYsYcuBm0aWd/Xr7ibJWUn+Gfgr4Cdne6Ak6/vdldv27NkzSFhJkqRJGrKEZZZ1+8x0VdW7q+p44BnA+bM9UFVdXFWrqmrVsmXLDm5KSZKkBoYsYbuAo0eWVwC75xpcVe8Djkvy0AEzSZIkTYUhS9hWYGWSY5McBqwFNo0OSPKdSdLffjxwGPDFATNJkiRNhcHeHVlVe5OcC1wFLAEuqaodSc7pt28AngmcneRrwG3Ac0YO1JckSVq0BithAFW1Gdg8tm7DyO3XAq8dMoMkSdI08oz5kiRJDVjCJEmSGrCESZIkNWAJkyRJasASJkmS1IAlTJIkqQFLmCRJUgOWMEmSpAYsYZIkSQ1YwiRJkhqwhEmSJDUw6LUjJUnSdHvcFY9oHWEf1565u3WEiXAmTJIkqQFLmCRJUgOWMEmSpAYsYZIkSQ1YwiRJkhqwhEmSJDVgCZMkSWrAEiZJktSAJUySJKkBS5gkSVIDljBJkqQGLGGSJEkNWMIkSZIasIRJkiQ1YAmTJElqwBImSZLUgCVMkiSpAUuYJElSA5YwSZKkBixhkiRJDVjCJEmSGrCESZIkNWAJkyRJasASJkmS1MCgJSzJ6iTXJ9mZ5LxZtj8vycf6j/cneeyQeSRJkqbFYCUsyRLgIuB04ARgXZITxoZ9BnhSVT0GOB+4eKg8kiRJ02TImbCTgZ1VdUNV3QFsBNaMDqiq91fVl/vFDwArBswjSZI0NYYsYcuBm0aWd/Xr5vJTwHsGzCNJkjQ1lg742JllXc06MHkyXQl74hzb1wPrAY455piDlU+SJKmZIWfCdgFHjyyvAHaPD0ryGOAtwJqq+uJsD1RVF1fVqqpatWzZskHCSpIkTdKQJWwrsDLJsUkOA9YCm0YHJDkGeBfw/Kr61IBZJEmSpspguyOram+Sc4GrgCXAJVW1I8k5/fYNwG8ADwHelARgb1WtGiqTJEnStBjymDCqajOweWzdhpHbLwJeNGQGSZKkaeQZ8yVJkhqwhEmSJDVgCZMkSWrAEiZJktSAJUySJKkBS5gkSVIDljBJkqQGLGGSJEkNWMIkSZIasIRJkiQ1YAmTJElqwBImSZLUgCVMkiSpAUuYJElSA5YwSZKkBixhkiRJDVjCJEmSGrCESZIkNWAJkyRJasASJkmS1IAlTJIkqQFLmCRJUgOWMEmSpAYsYZIkSQ1YwiRJkhqwhEmSJDVgCZMkSWrAEiZJktSAJUySJKkBS5gkSVIDljBJkqQGLGGSJEkNWMIkSZIasIRJkiQ1YAmTJElqwBImSZLUgCVMkiSpgUFLWJLVSa5PsjPJebNsPz7JPyX5zyQvHzKLJEnSNFk61AMnWQJcBDwV2AVsTbKpqq4bGfYl4CXAM4bKIUmSNI2GnAk7GdhZVTdU1R3ARmDN6ICq+kJVbQW+NmAOSZKkqTNkCVsO3DSyvKtfd48lWZ9kW5Jte/bsOSjhJEmSWhqyhGWWdXVvHqiqLq6qVVW1atmyZfcxliRJUntDlrBdwNEjyyuA3QN+PkmSpAVjsAPzga3AyiTHAp8D1gLPHfDzSZIWuPNOe0nrCPu44Jo3HnDMNOaG+WVXO4OVsKram+Rc4CpgCXBJVe1Ick6/fUOShwPbgAcBdyX5eeCEqvrqULkkSZKmwZAzYVTVZmDz2LoNI7c/T7ebUpIk6ZDiGfMlSZIasIRJkiQ1YAmTJElqwBImSZLUgCVMkiSpAUuYJElSA5YwSZKkBixhkiRJDVjCJEmSGrCESZIkNWAJkyRJasASJkmS1IAlTJIkqQFLmCRJUgOWMEmSpAYsYZIkSQ1YwiRJkhqwhEmSJDVgCZMkSWrAEiZJktSAJUySJKkBS5gkSVIDljBJkqQGLGGSJEkNWMIkSZIasIRJkiQ1YAmTJElqwBImSZLUgCVMkiSpAUuYJElSA5YwSZKkBixhkiRJDVjCJEmSGrCESZIkNWAJkyRJasASJkmS1IAlTJIkqYFBS1iS1UmuT7IzyXmzbE+SN/bbP5bk8UPmkSRJmhaDlbAkS4CLgNOBE4B1SU4YG3Y6sLL/WA+8eag8kiRJ02TImbCTgZ1VdUNV3QFsBNaMjVkDvL06HwCOSvLtA2aSJEmaCqmqYR44eRawuqpe1C8/H/j+qjp3ZMwVwAVV9Q/98nuBV1TVtrHHWk83UwbwaOD6QULDQ4GbB3rsoS3U7As1Nyzc7As1Nyzc7As1Nyzc7As1Nyzc7As1Nwyb/eaqWj3bhqUDfUKAzLJuvPHNZwxVdTFw8cEItT9JtlXVqqE/zxAWavaFmhsWbvaFmhsWbvaFmhsWbvaFmhsWbvaFmhvaZR9yd+Qu4OiR5RXA7nsxRpIkadEZsoRtBVYmOTbJYcBaYNPYmE3A2f27JH8A+EpV/euAmSRJkqbCYLsjq2pvknOBq4AlwCVVtSPJOf32DcBm4AxgJ/AfwAuHyjNPg+/yHNBCzb5Qc8PCzb5Qc8PCzb5Qc8PCzb5Qc8PCzb5Qc0Oj7IMdmC9JkqS5ecZ8SZKkBixhkiRJDRxyJSzJLf2/j0pSSX5uZNuFSV7Q3760P9fZPvdtJcmdSbYn+USSv0xyVL/+UUk+0d8+LclX+nHbk/ztFOSd+TivX39NkhuTZGTsX4x9b27r7/PRJO9P8uhG2T+R5M+SHJHkrLH/z/YkdyU5feR+L0tye5IHTzLvWPZfTbKjvxTY9iRX9//uHHtu/JcDfS8a5T+r/9k8vl+e+Vk9f2TMQ5N8LcmF/fIrk7y8Ud5bRm6fkeT/JjmmX16a5OYkrxm7zzVJto0sr0pyzcRC3z3Lw5NsTPLpJNcl2Zzku/pt+zyfR15jrk3yyST/s0HmSvJHI8tLk+xJd+5JkvxCkj8c2f68JH/V335lks+N/Bxc0CD/Pj9fs+TanmR5ki+Ov570P6PPnlzir3/efV7Tk7z7AK8vq0bu//XfVRPOPdvX+6qx/8vuJB/st12a5DPpfv98KsnbkywfItshV8LGfAF4abp3by4Et1XViVX1vcCXgJ+dY9yWftyJVfUjE8w37raRHCdW1eiL3b8BpwCkK5PjV0r4dH+fxwJvA35lEoFHjH6t7wDOqap3j/5/gDcBW+jefDJjHd07g8+acF4AkvwgcCbw+Kp6DPAjwPP6vC/i7s+N9/d3+zf2/72YtHXAP9C9o3rGDXT/rxn/DdgxyVAHkuQpwO/TnaT6xn71j9KdXPrZo0W397DRAt9Cn+ndwDVVdVxVnUD3s/Zt/ZC5ns9bqupxwCrgJ5KcNKnMvVuB701yeL/8VOBzI9vfCJyU5JT+Of0q4OdGtr9h5Odgn+saN/SGsdfMzwF/DTxjZkBfyJ4IXNEg3z6v6VV11gFeX6ZSVT1t5LX8FOCrwK+NDPml/vfPo4FrgauH6AqHegnbA7wX+O+tg9wL/wQM0swnZCPf+CX748C79jP2QcCXB080ty3Ad46u6GcKfgN4flXd1a87DjiS7gd53aRD9r6d7uzM/wlQVTdX1YHOvXdPvheDSnIk3QviT3H3EnYb8MmRv6qfA/zphOPNKcmpwB8AP1ZVnx7ZtA74PeBG4AfG7vZ67v6i38KTga/171YHoKq2V9WW+Tyfq+pW4MPAcZMIO+Y9wI/1t9cB7xjJtRf4GbrrF7+O7t35N0w84cHxDu7+s3AWcGVV/UejPIvR7wGbq+pvxjf0l1V8A/B5uutdH1SHegkDuAD4xXQXHB/3+tHpygnnmlOf9Snse961GaeO5P7VCUYbd/jYdO9zRra9F/ih/v+yFnjn2H2P6+/zaeAXgN+ZUOa7SbKU7gfv4yPr7g/8H+DlIzMe8I1fBFuARyd52CSz9v4aOLqfQn9TkifN4z4H+l5M0jPofsF8CvhSksePbNsIrE2yAriT6Tmx8wOAy4FnVNU/z6zsZ2meQjdj8Q72LTL/BPxnkidPKugsvpeuRM3mgM/nJA+hK5ctZiVnng/fBDwG+ODoxn4m5pN0s8GvG7vvy0Zel542kbTzM5rr6n7dlXSzeg/pl9cyUjgnbH+v6XP5k5HfoZsHznePJTmLbkb3lw8w9CPA8Qf78x/yJayqPgN8CHjuLJt/aWz3U2uH90/kLwLfCuzT2nujU8Kvnli6fY1PXY/+cr+TbpfTc4DDq+qzY/ed2R15HPDzTP4cLjNf6210sxh/OLLtfGBHVW0cu89aYGM/M/Yuul1mE1VVtwAn0V1rdQ/wzvTHOe7Hgb4Xk7SO7pcr/b+jxeVKut1O62hbFMd9DXg/3ezdqDOBq/sZi8uAs2b5Y+9VtJ8Nm8v+ns+nJrmWrvRfUFUTL2FV9THgUXTPh31+ufezqquA+wPLxjaP7va7avy+DY3mejJAVd1B9wf3s5I8FDiR7uvewv5e0+fyvJHfoWcMnO8e6Y/zeiPw3Jm9B/sbPkSGIa8duZD8FvDnwPtaBzmA26rqxP6YgCvojgl7Y+NM98VGuuNRXnmAcZuAtw6e5u5um614JzkNeCbw+LH1jwFWAn/TH/pzGN1xTBcNnHMfVXUncA1wTZKP0+1uv/QAd5vv92Iw/V/6P0x3rE/RneS56I69o6ruSPJh4BeB7wGe3irrmLuAZwN/m+RXquq3+vXrgFOSfLZffgjd7r+vv1mmqv4u3RsOxndVTsoO4FnjK+fxfN5SVWeO36+BTcD/Bk6j+/qO+k3gj4H/B7yBBn8UHUTvoCvrAS6vqq81zrPg9cdDvo3uj4jr5nGXx9HtNTioDvmZMIB+F8J13P3A36lVVV8BXgK8vN81tlBtAV7DgafWnwh8+gBjBpfkW+jK4NlV9e9jm9cBr6yqR/UfjwCWJ3nkhDM+OsnKkVUnAv8yj7vO93sxpGcBb6+qR/Zfw6OBz9BdU3bGbwOvqKovNkk4h36260zgeUl+KsmD6J63x8w8J+j+aJrt2KpXA/9jYmHv7u+AByT56ZkVSZ5Ad4xM8+fzPFwC/K+q+vjoyiTfR3e82GvpZtEfmeSpDfIdLFfTleKfpe3P6GLycuD2qtrvH8rpvITueNsrD3YIZ8K+4dV074BYEKrq2iQfpdtlsKV1njnM7NKbceXoO5Gqquj+ip3Ncf19Q/fuxBcNFfIeOAd4GPDmsTe6vYbu+zB+0Oa7+/WvnUi6zpHA7/fvCNtLd0mw9Qe60wG+F5Oyju4YzVGXMfLO2H6311S9K3JGVX0pyWq6GfUjgb8b28VxOfC6JA8Yu9/mJHsmGHX0c1d/TMzvpjuFzO3AZ+lmll48Nnzm+fxBpkRV7aIrjF/Xz3C8GXhZVd3er/sZ4O1JTpx4yNkdkWTXyPLMMa8vS/ITI+ufUVWfraq7klxGN5vXco/Nfl/Tp9hsX+9XAbvG/j9fntkNTHdM+K8DRwAfAJ7c7xo+qLxskSRJUgPujpQkSWrAEiZJktSAJUySJKkBS5gkSVIDljBJkqQGLGGSDhlJXpnk5a1zSBJYwiQd4vrrg0rSxPniI2lR6y9ifzZwE931ND+c5Bq66z2eAmzqz7B+RVX9eX+fW6rqyCT3Ay4EnkR39v77AZfMjJOk+8ISJmnRSnIS3VneH0f3evcR4MP95qOq6kn9uEvneIgfp7tI9PfRXS3hk3SXypGk+8wSJmkxOxV4d39tR5JsGtn2znnc/4nAn1XVXcDnk1w9QEZJhyiPCZO02M11bbZbR27vpX897K89eFi/PuN3kqSDxRImaTF7H3BWksOTPBB4+hzjPguc1N9eA9y/v/0PwDOT3C/Jt9Fd2FqSDgp3R0patKrqI0neCWwH/gXYMsfQPwAuT/Ih4L18Y5bsMuApwCeATwEfBL4yZGZJh45UzTVTL0lKcmRV3ZLkIcCHgFOq6vOtc0la+JwJk6T9uyLJUXTHiZ1vAZN0sDgTJkmS1IAH5kuSJDVgCZMkSWrAEiZJktSAJUySJKkBS5gkSVID/x+nlVL1AkVaAAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 720x504 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "list_of_colors = [\n",
    "    (30, 204, 0),\n",
    "    (9, 96, 19),\n",
    "    (255, 146, 135),\n",
    "    (221, 160, 21),\n",
    "    (210,78,193),\n",
    "    (210,78,193),\n",
    "    (210,78,193),\n",
    "    \n",
    "    (210,78,193),\n",
    "    \n",
    "    (128,54,115),\n",
    "    (128,54,115),\n",
    "    (30,204,0),\n",
    "    (9, 50, 105),\n",
    "    \n",
    "    (128,54,115),\n",
    "    (128,54,115),\n",
    "    (128,54,115),\n",
    "    (128,54,115),\n",
    "    (139,192,221),\n",
    "    (9, 50, 105),\n",
    "\n",
    "    (189,189,189)\n",
    "]\n",
    "list_of_colors = [(x/255,y/255, z/255) for x,y,z in list_of_colors]\n",
    "\n",
    "# Make plot\n",
    "\n",
    "df_both = pd.read_csv(\"output/05.antibiotic/num_strains_with_dependency_simultaneous.csv\", index_col=0) \n",
    "df_single = pd.read_csv(\"output/05.antibiotic/num_strains_with_dependency.csv\", index_col=0)  \n",
    "\n",
    "df_both[\"label\"]= \"both\"\n",
    "df_single[\"label\"] = \"sequential_only\"\n",
    "\n",
    "sorted_drug_list = [ 'INH', \"RIF\", \"EMB\", \"PZA\", \n",
    "                     'STM', 'AMI','KAN', 'CAP',\n",
    "                     'MXF', 'LEV', 'ETH',\n",
    "                    'DLM','CFZ','LZD','BDQ'\n",
    "]\n",
    "\n",
    "#df = pd.concat([df_both, df_single])\n",
    "df = df_single\n",
    "# Generate a rank column that will be used to sort\n",
    "# the dataframe numerically\n",
    "sorterIndex = dict(zip(sorted_drug_list, range(len(sorted_drug_list))))\n",
    "df['ab_rank'] = df['drug'].map(sorterIndex)\n",
    "df = df.sort_values(\"ab_rank\")\n",
    "sns.set_palette(sns.color_palette(list_of_colors))\n",
    "figure=plt.figure(figsize=(10,7))\n",
    "sns.barplot(data=df,x=\"drug\", y=\"fraction\", palette= sns.color_palette(list_of_colors))\n",
    "sns.despine()\n",
    "plt.savefig(\"figures/percent_strains_with_dependency.pdf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d593dbbb-0cd9-4e5f-99ca-36899e32e645",
   "metadata": {},
   "source": [
    "## Prepare Cytoscape plot of between-antibiotic dependencies"
   ]
  },
  {
   "cell_type": "raw",
   "id": "becf51d7-0a13-4b4a-a3e4-67d9f72cc7c5",
   "metadata": {},
   "source": [
    "For each drug:\n",
    "    read in the \"after AB\" file\n",
    "    read in the \"between AB\" file\n",
    "    for each hit\n",
    "        group hits by gene\n",
    "        take hits that happen at least 10 times\n",
    "        take only the top 3 hits in terms of number of happenings\n",
    "    save to output file\n",
    "    \n",
    "    make output file of node characteristics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b10216a0-a7f3-4d83-93d9-24a89b119b7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘output/05.antibiotic/cytoscape’: File exists\n"
     ]
    }
   ],
   "source": [
    "!mkdir output/05.antibiotic/cytoscape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "85add2f7-6dcd-413e-9506-e18dba73200e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ETH 46\n",
      "       position_i gene_name_i  position_j gene_name_j  overlap drug\n",
      "51629     1673425        None     1674481        inhA       27  ETH\n",
      "       position_i gene_name_i  position_j gene_name_j  overlap drug\n",
      "51582      761155        rpoB     1674481        inhA       26  ETH\n",
      "LZD 68\n",
      "        position_i gene_name_i  position_j gene_name_j  overlap drug\n",
      "664738      781687        rpsL     1472362        None       41  LZD\n",
      "       position_i gene_name_i  position_j gene_name_j  overlap drug\n",
      "49143     1473246        None     2155168        katG      254  LZD\n",
      "STM 211\n",
      "        position_i gene_name_i  position_j gene_name_j  overlap drug\n",
      "664738      781687        rpsL     1472362        None       41  STM\n",
      "       position_i gene_name_i  position_j gene_name_j  overlap drug\n",
      "49053      781687        rpsL     2155168        katG      434  STM\n",
      "CAP 67\n",
      "        position_i gene_name_i  position_j gene_name_j  overlap drug\n",
      "664738      781687        rpsL     1472362        None       41  CAP\n",
      "       position_i gene_name_i  position_j gene_name_j  overlap drug\n",
      "49143     1473246        None     2155168        katG      254  CAP\n",
      "LEV 11\n",
      "Empty DataFrame\n",
      "Columns: [position_i, gene_name_i, position_j, gene_name_j, overlap, drug]\n",
      "Index: []\n",
      "       position_i gene_name_i  position_j gene_name_j  overlap drug\n",
      "49077        7582        gyrA     2155168        katG      339  LEV\n",
      "PZA 119\n",
      "Empty DataFrame\n",
      "Columns: [position_i, gene_name_i, position_j, gene_name_j, overlap, drug]\n",
      "Index: []\n",
      "        position_i gene_name_i  position_j gene_name_j  overlap drug\n",
      "693242     2289252        None     2155168        katG       87  PZA\n",
      "EMB 250\n",
      "       position_i gene_name_i  position_j gene_name_j  overlap drug\n",
      "62472     4247431        embB     4247730        embB       41  EMB\n",
      "        position_i gene_name_i  position_j gene_name_j  overlap drug\n",
      "693229     4247429        embB     2155168        katG      454  EMB\n",
      "MXF 11\n",
      "Empty DataFrame\n",
      "Columns: [position_i, gene_name_i, position_j, gene_name_j, overlap, drug]\n",
      "Index: []\n",
      "       position_i gene_name_i  position_j gene_name_j  overlap drug\n",
      "49077        7582        gyrA     2155168        katG      339  MXF\n",
      "KAN 100\n",
      "        position_i gene_name_i  position_j gene_name_j  overlap drug\n",
      "664738      781687        rpsL     1472362        None       41  KAN\n",
      "       position_i gene_name_i  position_j gene_name_j  overlap drug\n",
      "49143     1473246        None     2155168        katG      254  KAN\n",
      "INH 172\n",
      "       position_i gene_name_i  position_j gene_name_j  overlap drug\n",
      "49209     1673425        None     2155168        katG      174  INH\n",
      "       position_i gene_name_i  position_j gene_name_j  overlap drug\n",
      "49046      761155        rpoB     2155168        katG      815  INH\n",
      "RIF 129\n",
      "        position_i gene_name_i  position_j gene_name_j  overlap drug\n",
      "686655      764817        rpoC      761155        rpoB      183  RIF\n",
      "       position_i gene_name_i  position_j gene_name_j  overlap drug\n",
      "49046      761155        rpoB     2155168        katG      815  RIF\n",
      "AMI 100\n",
      "        position_i gene_name_i  position_j gene_name_j  overlap drug\n",
      "664738      781687        rpsL     1472362        None       41  AMI\n",
      "       position_i gene_name_i  position_j gene_name_j  overlap drug\n",
      "49143     1473246        None     2155168        katG      254  AMI\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position_i</th>\n",
       "      <th>gene_name_i</th>\n",
       "      <th>position_j</th>\n",
       "      <th>gene_name_j</th>\n",
       "      <th>overlap</th>\n",
       "      <th>drug</th>\n",
       "      <th>name_i</th>\n",
       "      <th>name_j</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1673425</td>\n",
       "      <td>None</td>\n",
       "      <td>1674481</td>\n",
       "      <td>inhA</td>\n",
       "      <td>27</td>\n",
       "      <td>ETH</td>\n",
       "      <td>None_1673425</td>\n",
       "      <td>inhA_1674481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>761155</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>1674481</td>\n",
       "      <td>inhA</td>\n",
       "      <td>26</td>\n",
       "      <td>ETH</td>\n",
       "      <td>rpoB_761155</td>\n",
       "      <td>inhA_1674481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>1472362</td>\n",
       "      <td>None</td>\n",
       "      <td>41</td>\n",
       "      <td>LZD</td>\n",
       "      <td>rpsL_781687</td>\n",
       "      <td>None_1472362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1473246</td>\n",
       "      <td>None</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>254</td>\n",
       "      <td>LZD</td>\n",
       "      <td>None_1473246</td>\n",
       "      <td>katG_2155168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>1472362</td>\n",
       "      <td>None</td>\n",
       "      <td>41</td>\n",
       "      <td>STM</td>\n",
       "      <td>rpsL_781687</td>\n",
       "      <td>None_1472362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>434</td>\n",
       "      <td>STM</td>\n",
       "      <td>rpsL_781687</td>\n",
       "      <td>katG_2155168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>1472362</td>\n",
       "      <td>None</td>\n",
       "      <td>41</td>\n",
       "      <td>CAP</td>\n",
       "      <td>rpsL_781687</td>\n",
       "      <td>None_1472362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1473246</td>\n",
       "      <td>None</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>254</td>\n",
       "      <td>CAP</td>\n",
       "      <td>None_1473246</td>\n",
       "      <td>katG_2155168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7582</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>339</td>\n",
       "      <td>LEV</td>\n",
       "      <td>gyrA_7582</td>\n",
       "      <td>katG_2155168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2289252</td>\n",
       "      <td>None</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>87</td>\n",
       "      <td>PZA</td>\n",
       "      <td>None_2289252</td>\n",
       "      <td>katG_2155168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>4247431</td>\n",
       "      <td>embB</td>\n",
       "      <td>4247730</td>\n",
       "      <td>embB</td>\n",
       "      <td>41</td>\n",
       "      <td>EMB</td>\n",
       "      <td>embB_4247431</td>\n",
       "      <td>embB_4247730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>4247429</td>\n",
       "      <td>embB</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>454</td>\n",
       "      <td>EMB</td>\n",
       "      <td>embB_4247429</td>\n",
       "      <td>katG_2155168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>7582</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>339</td>\n",
       "      <td>MXF</td>\n",
       "      <td>gyrA_7582</td>\n",
       "      <td>katG_2155168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>1472362</td>\n",
       "      <td>None</td>\n",
       "      <td>41</td>\n",
       "      <td>KAN</td>\n",
       "      <td>rpsL_781687</td>\n",
       "      <td>None_1472362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1473246</td>\n",
       "      <td>None</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>254</td>\n",
       "      <td>KAN</td>\n",
       "      <td>None_1473246</td>\n",
       "      <td>katG_2155168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1673425</td>\n",
       "      <td>None</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>174</td>\n",
       "      <td>INH</td>\n",
       "      <td>None_1673425</td>\n",
       "      <td>katG_2155168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>761155</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>815</td>\n",
       "      <td>INH</td>\n",
       "      <td>rpoB_761155</td>\n",
       "      <td>katG_2155168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>764817</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>761155</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>183</td>\n",
       "      <td>RIF</td>\n",
       "      <td>rpoC_764817</td>\n",
       "      <td>rpoB_761155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>761155</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>815</td>\n",
       "      <td>RIF</td>\n",
       "      <td>rpoB_761155</td>\n",
       "      <td>katG_2155168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>1472362</td>\n",
       "      <td>None</td>\n",
       "      <td>41</td>\n",
       "      <td>AMI</td>\n",
       "      <td>rpsL_781687</td>\n",
       "      <td>None_1472362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1473246</td>\n",
       "      <td>None</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>254</td>\n",
       "      <td>AMI</td>\n",
       "      <td>None_1473246</td>\n",
       "      <td>katG_2155168</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    position_i gene_name_i  position_j gene_name_j  overlap drug  \\\n",
       "0      1673425        None     1674481        inhA       27  ETH   \n",
       "1       761155        rpoB     1674481        inhA       26  ETH   \n",
       "2       781687        rpsL     1472362        None       41  LZD   \n",
       "3      1473246        None     2155168        katG      254  LZD   \n",
       "4       781687        rpsL     1472362        None       41  STM   \n",
       "5       781687        rpsL     2155168        katG      434  STM   \n",
       "6       781687        rpsL     1472362        None       41  CAP   \n",
       "7      1473246        None     2155168        katG      254  CAP   \n",
       "8         7582        gyrA     2155168        katG      339  LEV   \n",
       "9      2289252        None     2155168        katG       87  PZA   \n",
       "10     4247431        embB     4247730        embB       41  EMB   \n",
       "11     4247429        embB     2155168        katG      454  EMB   \n",
       "12        7582        gyrA     2155168        katG      339  MXF   \n",
       "13      781687        rpsL     1472362        None       41  KAN   \n",
       "14     1473246        None     2155168        katG      254  KAN   \n",
       "15     1673425        None     2155168        katG      174  INH   \n",
       "16      761155        rpoB     2155168        katG      815  INH   \n",
       "17      764817        rpoC      761155        rpoB      183  RIF   \n",
       "18      761155        rpoB     2155168        katG      815  RIF   \n",
       "19      781687        rpsL     1472362        None       41  AMI   \n",
       "20     1473246        None     2155168        katG      254  AMI   \n",
       "\n",
       "          name_i        name_j  \n",
       "0   None_1673425  inhA_1674481  \n",
       "1    rpoB_761155  inhA_1674481  \n",
       "2    rpsL_781687  None_1472362  \n",
       "3   None_1473246  katG_2155168  \n",
       "4    rpsL_781687  None_1472362  \n",
       "5    rpsL_781687  katG_2155168  \n",
       "6    rpsL_781687  None_1472362  \n",
       "7   None_1473246  katG_2155168  \n",
       "8      gyrA_7582  katG_2155168  \n",
       "9   None_2289252  katG_2155168  \n",
       "10  embB_4247431  embB_4247730  \n",
       "11  embB_4247429  katG_2155168  \n",
       "12     gyrA_7582  katG_2155168  \n",
       "13   rpsL_781687  None_1472362  \n",
       "14  None_1473246  katG_2155168  \n",
       "15  None_1673425  katG_2155168  \n",
       "16   rpoB_761155  katG_2155168  \n",
       "17   rpoC_764817   rpoB_761155  \n",
       "18   rpoB_761155  katG_2155168  \n",
       "19   rpsL_781687  None_1472362  \n",
       "20  None_1473246  katG_2155168  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "N=1\n",
    "_p = []\n",
    "\n",
    "for drug in drug_list: \n",
    "    try:\n",
    "        ab_hits = pd.read_csv(f\"output/05.antibiotic/{drug}_antibiotic_hits.csv\", index_col=0).drop_duplicates()\n",
    "        between_hits = pd.read_csv(f\"output/05.antibiotic/{drug}_between_antibiotic_hits.csv\", index_col=0).drop_duplicates()\n",
    "        \n",
    "        ## ThyX is antibiotic associated\n",
    "        #ab_hits = ab_hits.query(\"gene_name_i != 'r_thyX' and gene_name_i != 'r_thyA'\")\n",
    "    except:\n",
    "        continue\n",
    "\n",
    "    allhits = ab_hits\n",
    "    allhits['overlap'] = allhits.n_muts_i_on_branch_j + allhits.n_muts_same_branch\n",
    "    print(drug, len(allhits))\n",
    "    \n",
    "    # filtering\n",
    "    allhits = allhits.query(\"overlap > 9\")\n",
    "    \n",
    "    allhits = allhits.sort_values(\"overlap\", ascending=False)\n",
    "    p = allhits.head(N)[[\"position_i\", \"gene_name_i\", \"position_j\", \"gene_name_j\", \"overlap\"]]\n",
    "    p[\"drug\"] = drug\n",
    "    print(p)\n",
    "    _p.append(p)\n",
    "    \n",
    "    between_ab = between_hits\n",
    "    between_ab['overlap'] = between_ab.n_muts_i_on_branch_j + between_ab.n_muts_same_branch\n",
    "    between_ab = between_ab.sort_values(\"overlap\", ascending=False)\n",
    "    \n",
    "    between_ab = between_ab.query(\"overlap > 9\")\n",
    "    \n",
    "    p = between_ab.head(N)[[\"position_i\", \"gene_name_i\", \"position_j\", \"gene_name_j\", \"overlap\"]]\n",
    "    p[\"drug\"] = drug\n",
    "    print(p)\n",
    "    _p.append(p)\n",
    "    \n",
    "df = pd.concat(_p)\n",
    "df = df.reset_index(drop=True)\n",
    "df.loc[df.query(\"position_i==3073679\").index, \"gene_name_i\"] = \"r_thyA\"\n",
    "\n",
    "df['name_i']=[f'{x}_{y}' for x,y in zip(df.gene_name_i, df.position_i)]\n",
    "df['name_j']=[f'{x}_{y}' for x,y in zip(df.gene_name_j, df.position_j)]\n",
    "\n",
    "df.to_csv(f\"output/05.antibiotic/cytoscape//top_{N}_hits_per_drug_2022.csv\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e75258e6-83e1-4958-bcca-c74d4b201f01",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>overlap</th>\n",
       "      <th>drug</th>\n",
       "      <th>name_i</th>\n",
       "      <th>name_j</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name_i</th>\n",
       "      <th>gene_name_j</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">None</th>\n",
       "      <th>inhA</th>\n",
       "      <td>27</td>\n",
       "      <td>ETH</td>\n",
       "      <td>None</td>\n",
       "      <td>inhA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>katG</th>\n",
       "      <td>1277</td>\n",
       "      <td>PZA</td>\n",
       "      <td>None</td>\n",
       "      <td>katG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">embB</th>\n",
       "      <th>embB</th>\n",
       "      <td>41</td>\n",
       "      <td>EMB</td>\n",
       "      <td>embB</td>\n",
       "      <td>embB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>katG</th>\n",
       "      <td>454</td>\n",
       "      <td>EMB</td>\n",
       "      <td>embB</td>\n",
       "      <td>katG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gyrA</th>\n",
       "      <th>katG</th>\n",
       "      <td>678</td>\n",
       "      <td>MXF</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>katG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">rpoB</th>\n",
       "      <th>inhA</th>\n",
       "      <td>26</td>\n",
       "      <td>ETH</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>inhA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>katG</th>\n",
       "      <td>1630</td>\n",
       "      <td>RIF</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>katG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpoC</th>\n",
       "      <th>rpoB</th>\n",
       "      <td>183</td>\n",
       "      <td>RIF</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>rpoB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">rpsL</th>\n",
       "      <th>None</th>\n",
       "      <td>205</td>\n",
       "      <td>STM</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>katG</th>\n",
       "      <td>434</td>\n",
       "      <td>STM</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>katG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         overlap drug name_i name_j\n",
       "gene_name_i gene_name_j                            \n",
       "None        inhA              27  ETH   None   inhA\n",
       "            katG            1277  PZA   None   katG\n",
       "embB        embB              41  EMB   embB   embB\n",
       "            katG             454  EMB   embB   katG\n",
       "gyrA        katG             678  MXF   gyrA   katG\n",
       "rpoB        inhA              26  ETH   rpoB   inhA\n",
       "            katG            1630  RIF   rpoB   katG\n",
       "rpoC        rpoB             183  RIF   rpoC   rpoB\n",
       "rpsL        None             205  STM   rpsL   None\n",
       "            katG             434  STM   rpsL   katG"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# mutations in same gene or region get condensed\n",
    "\n",
    "g= df.groupby([\"gene_name_i\", \"gene_name_j\"]).agg({\n",
    "    \"overlap\":\"sum\",\n",
    "    \"drug\":\"max\"\n",
    "})\n",
    "g[\"name_i\"] = g.index.get_level_values(0)\n",
    "g[\"name_j\"] = g.index.get_level_values(1)\n",
    "g.name_i = [x.replace(\"r_\",\"p_\") for x in g.name_i]\n",
    "g.name_j = [x.replace(\"r_\",\"p_\") for x in g.name_j]\n",
    "#g[\"drug\"] = [gene_to_ab[x] if x in gene_to_ab else None for x in g.index.get_level_values(0)]\n",
    "\n",
    "g.to_csv(f\"output/05.antibiotic/cytoscape/top_{N}_hits_per_drug_grouped.csv\")\n",
    "g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0c030e4d-cf4a-473e-aec4-67f8e69673f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_names = set(list(g.name_i.values) + list(g.name_j.values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b696c1d8-be7d-488f-b7a4-72fc8741964d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_name_to_drug = {\n",
    " 'embB':\"EMB\",\n",
    " 'gyrA':\"FQ\",\n",
    " 'inhA':\"INH_ETA\",\n",
    " 'katG':\"INH\",\n",
    " 'p_fabG1':\"INH_ETA\",\n",
    " 'p_pncA':\"PZA\",\n",
    " 'rpoB':\"RIF\",\n",
    " 'rpoC':\"RIF\",\n",
    " 'rpsL':\"STM\",\n",
    " 'rrs':\"AMG\"}\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f448a92c-13c6-4aac-8684-3d1f42725a35",
   "metadata": {},
   "outputs": [],
   "source": [
    "individual_attributes = pd.concat([\n",
    "    g[[\"name_i\", \"drug\"]],\n",
    "]).reset_index()\n",
    "\n",
    "individual_attributes['known'] = False\n",
    "\n",
    "individual_attributes[\"drug\"] = [gene_name_to_drug[x] if x in gene_name_to_drug else \"UK\" for x in individual_attributes.name_i]\n",
    "\n",
    "individual_attributes = individual_attributes.rename({\"name_i\": \"name\"}, axis=1)\n",
    "\n",
    "drug_to_color = {\n",
    "    'AMG':'#d24ec1',\n",
    "    'CYS': '#8bc0dd',\n",
    "    'EMB':'#FF9287',\n",
    "    'ETH':'#a0e85b',\n",
    "    'INH':'#a0e85b', \n",
    "    'INH_ETA':'#a0e85b',\n",
    "    'FQ':'#57377e',\n",
    "    'CIP':'#57377e',\n",
    "    'PAS':'#155392',\n",
    "    'PZA': '#6a3d9a',\n",
    "    'RIF':'#34a54f',\n",
    "    'RFB':'#34a54f',\n",
    "    'STM':'#fcc2fb',\n",
    "    \"UK\":\"#AAAAAA\"\n",
    "}\n",
    "\n",
    "\n",
    "individual_attributes['color'] = [drug_to_color[x] for x in \n",
    "    individual_attributes.drug]\n",
    "\n",
    "individual_attributes.to_csv(f'output/05.antibiotic/cytoscape/top_{N}_hits_per_drug_grouped_indiv.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46c2cf56-17df-477d-a337-6fc367769630",
   "metadata": {},
   "source": [
    "## Prepare for CIRCOS plot of after-antibiotic mutations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c149302f-139b-421a-90d2-a31092c84ba6",
   "metadata": {},
   "source": [
    "## Presence of Featured mutations in the paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "4938b144-f8f9-48e9-902d-b705e2857d42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position_i</th>\n",
       "      <th>gene_name_i</th>\n",
       "      <th>position_j</th>\n",
       "      <th>gene_name_j</th>\n",
       "      <th>n_mutations_i</th>\n",
       "      <th>n_mutations_j</th>\n",
       "      <th>n_branches_with_i</th>\n",
       "      <th>n_muts_i_on_branch_j</th>\n",
       "      <th>n_muts_same_branch</th>\n",
       "      <th>pval_beta_i_on_j_BH_sig_sequential</th>\n",
       "      <th>pval_beta_i_on_j_BH_sig_simultaneous</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>734624</td>\n",
       "      <td>nusG</td>\n",
       "      <td>761155</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>14</td>\n",
       "      <td>1487</td>\n",
       "      <td>32</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>734470</td>\n",
       "      <td>nusG</td>\n",
       "      <td>761155</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>6</td>\n",
       "      <td>1487</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     position_i gene_name_i  position_j gene_name_j  n_mutations_i  \\\n",
       "100      734624        nusG      761155        rpoB             14   \n",
       "109      734470        nusG      761155        rpoB              6   \n",
       "\n",
       "     n_mutations_j  n_branches_with_i  n_muts_i_on_branch_j  \\\n",
       "100           1487                 32                     4   \n",
       "109           1487                 10                     1   \n",
       "\n",
       "     n_muts_same_branch  pval_beta_i_on_j_BH_sig_sequential  \\\n",
       "100                   6                               False   \n",
       "109                   3                               False   \n",
       "\n",
       "    pval_beta_i_on_j_BH_sig_simultaneous  \n",
       "100                                 True  \n",
       "109                                 True  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "antibiotic_hits = pd.read_csv(\"output/05.antibiotic/antibiotic_hits.csv\")\n",
    "display_positions =['position_i', 'gene_name_i', 'position_j', 'gene_name_j',\n",
    "       'n_mutations_i', 'n_mutations_j', 'n_branches_with_i',\n",
    "    'n_muts_i_on_branch_j', 'n_muts_same_branch',\n",
    "       'pval_beta_i_on_j_BH_sig_sequential', 'pval_beta_i_on_j_BH_sig_simultaneous']\n",
    "antibiotic_hits.query('gene_name_i==\"nusG\" and gene_name_j==\"rpoB\"')[display_positions]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "538f358b-0761-4e46-a47b-5dfa82c24fd1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position_i</th>\n",
       "      <th>gene_name_i</th>\n",
       "      <th>position_j</th>\n",
       "      <th>gene_name_j</th>\n",
       "      <th>n_mutations_i</th>\n",
       "      <th>n_mutations_j</th>\n",
       "      <th>n_branches_with_i</th>\n",
       "      <th>n_muts_i_on_branch_j</th>\n",
       "      <th>n_muts_same_branch</th>\n",
       "      <th>pval_beta_i_on_j_BH_sig_sequential</th>\n",
       "      <th>pval_beta_i_on_j_BH_sig_simultaneous</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>629831</th>\n",
       "      <td>2726142</td>\n",
       "      <td>None</td>\n",
       "      <td>2155783</td>\n",
       "      <td>katG</td>\n",
       "      <td>39</td>\n",
       "      <td>9</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>694271</th>\n",
       "      <td>2726142</td>\n",
       "      <td>None</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>39</td>\n",
       "      <td>1811</td>\n",
       "      <td>20</td>\n",
       "      <td>14</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        position_i gene_name_i  position_j gene_name_j  n_mutations_i  \\\n",
       "629831     2726142        None     2155783        katG             39   \n",
       "694271     2726142        None     2155168        katG             39   \n",
       "\n",
       "        n_mutations_j  n_branches_with_i  n_muts_i_on_branch_j  \\\n",
       "629831              9                 20                     1   \n",
       "694271           1811                 20                    14   \n",
       "\n",
       "        n_muts_same_branch pval_beta_i_on_j_BH_sig_sequential  \\\n",
       "629831                   0                              False   \n",
       "694271                   3                              False   \n",
       "\n",
       "       pval_beta_i_on_j_BH_sig_simultaneous  \n",
       "629831                                  NaN  \n",
       "694271                                  NaN  "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_hits.query('position_i==2726142 and gene_name_j==\"katG\"')[display_positions]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "074d19a3-cde6-4b3f-a708-f5036bf7d57d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position_i</th>\n",
       "      <th>gene_name_i</th>\n",
       "      <th>position_j</th>\n",
       "      <th>gene_name_j</th>\n",
       "      <th>n_mutations_i</th>\n",
       "      <th>n_mutations_j</th>\n",
       "      <th>n_branches_with_i</th>\n",
       "      <th>n_muts_i_on_branch_j</th>\n",
       "      <th>n_muts_same_branch</th>\n",
       "      <th>pval_beta_i_on_j_BH_sig_sequential</th>\n",
       "      <th>pval_beta_i_on_j_BH_sig_simultaneous</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7004</th>\n",
       "      <td>2154083</td>\n",
       "      <td>katG</td>\n",
       "      <td>2726142</td>\n",
       "      <td>None</td>\n",
       "      <td>7</td>\n",
       "      <td>39</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7012</th>\n",
       "      <td>2155541</td>\n",
       "      <td>katG</td>\n",
       "      <td>2726142</td>\n",
       "      <td>None</td>\n",
       "      <td>35</td>\n",
       "      <td>39</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7014</th>\n",
       "      <td>2156110</td>\n",
       "      <td>katG</td>\n",
       "      <td>2726142</td>\n",
       "      <td>None</td>\n",
       "      <td>6</td>\n",
       "      <td>39</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7015</th>\n",
       "      <td>2155295</td>\n",
       "      <td>katG</td>\n",
       "      <td>2726142</td>\n",
       "      <td>None</td>\n",
       "      <td>6</td>\n",
       "      <td>39</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52763</th>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>2726142</td>\n",
       "      <td>None</td>\n",
       "      <td>1811</td>\n",
       "      <td>39</td>\n",
       "      <td>12564</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52774</th>\n",
       "      <td>2155167</td>\n",
       "      <td>katG</td>\n",
       "      <td>2726142</td>\n",
       "      <td>None</td>\n",
       "      <td>56</td>\n",
       "      <td>39</td>\n",
       "      <td>136</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>518330</th>\n",
       "      <td>2155783</td>\n",
       "      <td>katG</td>\n",
       "      <td>2726142</td>\n",
       "      <td>None</td>\n",
       "      <td>9</td>\n",
       "      <td>39</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        position_i gene_name_i  position_j gene_name_j  n_mutations_i  \\\n",
       "7004       2154083        katG     2726142        None              7   \n",
       "7012       2155541        katG     2726142        None             35   \n",
       "7014       2156110        katG     2726142        None              6   \n",
       "7015       2155295        katG     2726142        None              6   \n",
       "52763      2155168        katG     2726142        None           1811   \n",
       "52774      2155167        katG     2726142        None             56   \n",
       "518330     2155783        katG     2726142        None              9   \n",
       "\n",
       "        n_mutations_j  n_branches_with_i  n_muts_i_on_branch_j  \\\n",
       "7004               39                  0                     0   \n",
       "7012               39                 32                     0   \n",
       "7014               39                  0                     0   \n",
       "7015               39                  2                     0   \n",
       "52763              39              12564                     0   \n",
       "52774              39                136                     0   \n",
       "518330             39                 16                     0   \n",
       "\n",
       "        n_muts_same_branch pval_beta_i_on_j_BH_sig_sequential  \\\n",
       "7004                     1                                NaN   \n",
       "7012                     1                                NaN   \n",
       "7014                     1                                NaN   \n",
       "7015                     1                                NaN   \n",
       "52763                    3                              False   \n",
       "52774                    1                                NaN   \n",
       "518330                   0                              False   \n",
       "\n",
       "       pval_beta_i_on_j_BH_sig_simultaneous  \n",
       "7004                                   True  \n",
       "7012                                  False  \n",
       "7014                                   True  \n",
       "7015                                   True  \n",
       "52763                                 False  \n",
       "52774                                 False  \n",
       "518330                                  NaN  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_hits.query('position_j==2726142 and gene_name_i==\"katG\"')[display_positions]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "086d3f58-99ba-4cca-8f3a-594a2c72a79c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position_i</th>\n",
       "      <th>gene_name_i</th>\n",
       "      <th>position_j</th>\n",
       "      <th>gene_name_j</th>\n",
       "      <th>n_mutations_i</th>\n",
       "      <th>n_mutations_j</th>\n",
       "      <th>n_branches_with_i</th>\n",
       "      <th>n_muts_i_on_branch_j</th>\n",
       "      <th>n_muts_same_branch</th>\n",
       "      <th>pval_beta_i_on_j_BH_sig_sequential</th>\n",
       "      <th>pval_beta_i_on_j_BH_sig_simultaneous</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>732110</td>\n",
       "      <td>hadA</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>57</td>\n",
       "      <td>1811</td>\n",
       "      <td>84</td>\n",
       "      <td>12</td>\n",
       "      <td>28</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>732111</td>\n",
       "      <td>hadA</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>16</td>\n",
       "      <td>1811</td>\n",
       "      <td>42</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    position_i gene_name_i  position_j gene_name_j  n_mutations_i  \\\n",
       "20      732110        hadA     2155168        katG             57   \n",
       "23      732111        hadA     2155168        katG             16   \n",
       "\n",
       "    n_mutations_j  n_branches_with_i  n_muts_i_on_branch_j  \\\n",
       "20           1811                 84                    12   \n",
       "23           1811                 42                     3   \n",
       "\n",
       "    n_muts_same_branch  pval_beta_i_on_j_BH_sig_sequential  \\\n",
       "20                  28                               False   \n",
       "23                  10                               False   \n",
       "\n",
       "   pval_beta_i_on_j_BH_sig_simultaneous  \n",
       "20                                 True  \n",
       "23                                 True  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "antibiotic_hits.query('gene_name_i==\"hadA\" and gene_name_j==\"katG\"')[display_positions]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "bab4c560-338d-4fc6-b6d3-9a8f3e67f17d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position_i</th>\n",
       "      <th>gene_name_i</th>\n",
       "      <th>position_j</th>\n",
       "      <th>gene_name_j</th>\n",
       "      <th>n_mutations_i</th>\n",
       "      <th>n_mutations_j</th>\n",
       "      <th>n_branches_with_i</th>\n",
       "      <th>n_muts_i_on_branch_j</th>\n",
       "      <th>n_muts_same_branch</th>\n",
       "      <th>pval_beta_i_on_j_BH_sig_sequential</th>\n",
       "      <th>pval_beta_i_on_j_BH_sig_simultaneous</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>821891</th>\n",
       "      <td>3568795</td>\n",
       "      <td>None</td>\n",
       "      <td>4338473</td>\n",
       "      <td>whiB6</td>\n",
       "      <td>15</td>\n",
       "      <td>18</td>\n",
       "      <td>64</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>876097</th>\n",
       "      <td>3568795</td>\n",
       "      <td>None</td>\n",
       "      <td>4338292</td>\n",
       "      <td>whiB6</td>\n",
       "      <td>15</td>\n",
       "      <td>6</td>\n",
       "      <td>64</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>876240</th>\n",
       "      <td>3568795</td>\n",
       "      <td>None</td>\n",
       "      <td>4338371</td>\n",
       "      <td>whiB6</td>\n",
       "      <td>15</td>\n",
       "      <td>8</td>\n",
       "      <td>64</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        position_i gene_name_i  position_j gene_name_j  n_mutations_i  \\\n",
       "821891     3568795        None     4338473       whiB6             15   \n",
       "876097     3568795        None     4338292       whiB6             15   \n",
       "876240     3568795        None     4338371       whiB6             15   \n",
       "\n",
       "        n_mutations_j  n_branches_with_i  n_muts_i_on_branch_j  \\\n",
       "821891             18                 64                     0   \n",
       "876097              6                 64                     0   \n",
       "876240              8                 64                     4   \n",
       "\n",
       "        n_muts_same_branch pval_beta_i_on_j_BH_sig_sequential  \\\n",
       "821891                   0                              False   \n",
       "876097                   0                              False   \n",
       "876240                   0                               True   \n",
       "\n",
       "       pval_beta_i_on_j_BH_sig_simultaneous  \n",
       "821891                                  NaN  \n",
       "876097                                  NaN  \n",
       "876240                                  NaN  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# whiB7 upstream region\n",
    "all_hits.query('gene_name_i==\"r_whiB7\" and gene_name_j==\"whiB6\"')[display_positions]\n",
    "# whiB7\n",
    "all_hits.query('position_i > 3568679 and position_i <3569109 and gene_name_j==\"whiB6\"')[display_positions]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "0c394e43-6b69-4b00-a06c-1111e37844ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position_i</th>\n",
       "      <th>gene_name_i</th>\n",
       "      <th>position_j</th>\n",
       "      <th>gene_name_j</th>\n",
       "      <th>n_mutations_i</th>\n",
       "      <th>n_mutations_j</th>\n",
       "      <th>n_branches_with_i</th>\n",
       "      <th>n_muts_i_on_branch_j</th>\n",
       "      <th>n_muts_same_branch</th>\n",
       "      <th>pval_beta_i_on_j_BH_sig_sequential</th>\n",
       "      <th>pval_beta_i_on_j_BH_sig_simultaneous</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>57184</th>\n",
       "      <td>2715342</td>\n",
       "      <td>None</td>\n",
       "      <td>4338365</td>\n",
       "      <td>whiB6</td>\n",
       "      <td>139</td>\n",
       "      <td>48</td>\n",
       "      <td>350</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70925</th>\n",
       "      <td>2715344</td>\n",
       "      <td>None</td>\n",
       "      <td>4338473</td>\n",
       "      <td>whiB6</td>\n",
       "      <td>75</td>\n",
       "      <td>18</td>\n",
       "      <td>526</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>675482</th>\n",
       "      <td>2715369</td>\n",
       "      <td>None</td>\n",
       "      <td>4338473</td>\n",
       "      <td>whiB6</td>\n",
       "      <td>40</td>\n",
       "      <td>18</td>\n",
       "      <td>506</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>676765</th>\n",
       "      <td>2715369</td>\n",
       "      <td>None</td>\n",
       "      <td>4338365</td>\n",
       "      <td>whiB6</td>\n",
       "      <td>40</td>\n",
       "      <td>48</td>\n",
       "      <td>506</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>779963</th>\n",
       "      <td>2715342</td>\n",
       "      <td>None</td>\n",
       "      <td>4338371</td>\n",
       "      <td>whiB6</td>\n",
       "      <td>139</td>\n",
       "      <td>8</td>\n",
       "      <td>350</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>779969</th>\n",
       "      <td>2715369</td>\n",
       "      <td>None</td>\n",
       "      <td>4338371</td>\n",
       "      <td>whiB6</td>\n",
       "      <td>40</td>\n",
       "      <td>8</td>\n",
       "      <td>506</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>779974</th>\n",
       "      <td>2715344</td>\n",
       "      <td>None</td>\n",
       "      <td>4338371</td>\n",
       "      <td>whiB6</td>\n",
       "      <td>75</td>\n",
       "      <td>8</td>\n",
       "      <td>526</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>780000</th>\n",
       "      <td>2715346</td>\n",
       "      <td>None</td>\n",
       "      <td>4338371</td>\n",
       "      <td>whiB6</td>\n",
       "      <td>82</td>\n",
       "      <td>8</td>\n",
       "      <td>190</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809742</th>\n",
       "      <td>2715346</td>\n",
       "      <td>None</td>\n",
       "      <td>4338307</td>\n",
       "      <td>whiB6</td>\n",
       "      <td>82</td>\n",
       "      <td>5</td>\n",
       "      <td>190</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>876859</th>\n",
       "      <td>2714308</td>\n",
       "      <td>eis</td>\n",
       "      <td>4338471</td>\n",
       "      <td>whiB6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>106</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>876891</th>\n",
       "      <td>2715347</td>\n",
       "      <td>None</td>\n",
       "      <td>4338472</td>\n",
       "      <td>whiB6</td>\n",
       "      <td>13</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        position_i gene_name_i  position_j gene_name_j  n_mutations_i  \\\n",
       "57184      2715342        None     4338365       whiB6            139   \n",
       "70925      2715344        None     4338473       whiB6             75   \n",
       "675482     2715369        None     4338473       whiB6             40   \n",
       "676765     2715369        None     4338365       whiB6             40   \n",
       "779963     2715342        None     4338371       whiB6            139   \n",
       "779969     2715369        None     4338371       whiB6             40   \n",
       "779974     2715344        None     4338371       whiB6             75   \n",
       "780000     2715346        None     4338371       whiB6             82   \n",
       "809742     2715346        None     4338307       whiB6             82   \n",
       "876859     2714308         eis     4338471       whiB6              6   \n",
       "876891     2715347        None     4338472       whiB6             13   \n",
       "\n",
       "        n_mutations_j  n_branches_with_i  n_muts_i_on_branch_j  \\\n",
       "57184              48                350                     0   \n",
       "70925              18                526                     1   \n",
       "675482             18                506                     1   \n",
       "676765             48                506                     0   \n",
       "779963              8                350                    19   \n",
       "779969              8                506                     5   \n",
       "779974              8                526                     2   \n",
       "780000              8                190                     2   \n",
       "809742              5                190                     1   \n",
       "876859              6                106                     0   \n",
       "876891              8                  8                     1   \n",
       "\n",
       "        n_muts_same_branch pval_beta_i_on_j_BH_sig_sequential  \\\n",
       "57184                    1                                NaN   \n",
       "70925                    1                              False   \n",
       "675482                   0                              False   \n",
       "676765                   0                              False   \n",
       "779963                   0                               True   \n",
       "779969                   0                              False   \n",
       "779974                   0                              False   \n",
       "780000                   0                              False   \n",
       "809742                   0                              False   \n",
       "876859                   0                              False   \n",
       "876891                   0                              False   \n",
       "\n",
       "       pval_beta_i_on_j_BH_sig_simultaneous  \n",
       "57184                                 False  \n",
       "70925                                 False  \n",
       "675482                                  NaN  \n",
       "676765                                  NaN  \n",
       "779963                                  NaN  \n",
       "779969                                  NaN  \n",
       "779974                                  NaN  \n",
       "780000                                  NaN  \n",
       "809742                                  NaN  \n",
       "876859                                  NaN  \n",
       "876891                                  NaN  "
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# eis upstream region\n",
    "all_hits.query('position_i > 2714124 and position_i<2715472 and gene_name_j==\"whiB6\"')[display_positions]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "d40e766b-737d-4ed6-8d28-87db340a7693",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position_i</th>\n",
       "      <th>gene_name_i</th>\n",
       "      <th>position_j</th>\n",
       "      <th>gene_name_j</th>\n",
       "      <th>n_mutations_i</th>\n",
       "      <th>n_mutations_j</th>\n",
       "      <th>n_branches_with_i</th>\n",
       "      <th>n_muts_i_on_branch_j</th>\n",
       "      <th>n_muts_same_branch</th>\n",
       "      <th>pval_beta_i_on_j_BH_sig_sequential</th>\n",
       "      <th>pval_beta_i_on_j_BH_sig_simultaneous</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>664917</th>\n",
       "      <td>2715346</td>\n",
       "      <td>None</td>\n",
       "      <td>1472362</td>\n",
       "      <td>None</td>\n",
       "      <td>82</td>\n",
       "      <td>115</td>\n",
       "      <td>190</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>699923</th>\n",
       "      <td>2715342</td>\n",
       "      <td>None</td>\n",
       "      <td>1472359</td>\n",
       "      <td>None</td>\n",
       "      <td>139</td>\n",
       "      <td>233</td>\n",
       "      <td>350</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>733076</th>\n",
       "      <td>2715342</td>\n",
       "      <td>None</td>\n",
       "      <td>1473246</td>\n",
       "      <td>None</td>\n",
       "      <td>139</td>\n",
       "      <td>560</td>\n",
       "      <td>350</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        position_i gene_name_i  position_j gene_name_j  n_mutations_i  \\\n",
       "664917     2715346        None     1472362        None             82   \n",
       "699923     2715342        None     1472359        None            139   \n",
       "733076     2715342        None     1473246        None            139   \n",
       "\n",
       "        n_mutations_j  n_branches_with_i  n_muts_i_on_branch_j  \\\n",
       "664917            115                190                     8   \n",
       "699923            233                350                    15   \n",
       "733076            560                350                    14   \n",
       "\n",
       "        n_muts_same_branch pval_beta_i_on_j_BH_sig_sequential  \\\n",
       "664917                   3                               True   \n",
       "699923                   0                               True   \n",
       "733076                   1                               True   \n",
       "\n",
       "       pval_beta_i_on_j_BH_sig_simultaneous  \n",
       "664917                                  NaN  \n",
       "699923                                  NaN  \n",
       "733076                                  NaN  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# eis upstream region and rrs - 41 mutations\n",
    "all_hits.query('position_i > 2714124 and position_i<2715472 and position_j > 1471846 and position_j<1473382 and pval_beta_i_on_j_BH_sig_sequential ')[display_positions]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "71794573-94cf-4a34-8397-ba60c5e77c0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position_i</th>\n",
       "      <th>gene_name_i</th>\n",
       "      <th>position_j</th>\n",
       "      <th>gene_name_j</th>\n",
       "      <th>n_mutations_i</th>\n",
       "      <th>n_mutations_j</th>\n",
       "      <th>n_branches_with_i</th>\n",
       "      <th>n_muts_i_on_branch_j</th>\n",
       "      <th>n_muts_same_branch</th>\n",
       "      <th>pval_beta_i_on_j_BH_sig_sequential</th>\n",
       "      <th>pval_beta_i_on_j_BH_sig_simultaneous</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>57179</th>\n",
       "      <td>1472359</td>\n",
       "      <td>None</td>\n",
       "      <td>4338365</td>\n",
       "      <td>whiB6</td>\n",
       "      <td>233</td>\n",
       "      <td>48</td>\n",
       "      <td>870</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57196</th>\n",
       "      <td>1473247</td>\n",
       "      <td>None</td>\n",
       "      <td>4338365</td>\n",
       "      <td>whiB6</td>\n",
       "      <td>16</td>\n",
       "      <td>48</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70922</th>\n",
       "      <td>1472359</td>\n",
       "      <td>None</td>\n",
       "      <td>4338473</td>\n",
       "      <td>whiB6</td>\n",
       "      <td>233</td>\n",
       "      <td>18</td>\n",
       "      <td>870</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79162</th>\n",
       "      <td>1473246</td>\n",
       "      <td>None</td>\n",
       "      <td>4338471</td>\n",
       "      <td>whiB6</td>\n",
       "      <td>560</td>\n",
       "      <td>6</td>\n",
       "      <td>1893</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       position_i gene_name_i  position_j gene_name_j  n_mutations_i  \\\n",
       "57179     1472359        None     4338365       whiB6            233   \n",
       "57196     1473247        None     4338365       whiB6             16   \n",
       "70922     1472359        None     4338473       whiB6            233   \n",
       "79162     1473246        None     4338471       whiB6            560   \n",
       "\n",
       "       n_mutations_j  n_branches_with_i  n_muts_i_on_branch_j  \\\n",
       "57179             48                870                     2   \n",
       "57196             48                  6                     0   \n",
       "70922             18                870                     0   \n",
       "79162              6               1893                     0   \n",
       "\n",
       "       n_muts_same_branch pval_beta_i_on_j_BH_sig_sequential  \\\n",
       "57179                   4                              False   \n",
       "57196                   2                                NaN   \n",
       "70922                   2                                NaN   \n",
       "79162                   2                                NaN   \n",
       "\n",
       "      pval_beta_i_on_j_BH_sig_simultaneous  \n",
       "57179                                 True  \n",
       "57196                                 True  \n",
       "70922                                 True  \n",
       "79162                                 True  "
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# rrsto WhiB6 10 mutations\n",
    "all_hits.query('position_i > 1471846 and position_i<1473382  and gene_name_j==\"whiB6\"and pval_beta_i_on_j_BH_sig_simultaneous ')[display_positions].head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "e99bede5-fc3a-4d14-9d80-e4422df42ebb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position_i</th>\n",
       "      <th>gene_name_i</th>\n",
       "      <th>position_j</th>\n",
       "      <th>gene_name_j</th>\n",
       "      <th>n_mutations_i</th>\n",
       "      <th>n_mutations_j</th>\n",
       "      <th>n_branches_with_i</th>\n",
       "      <th>n_muts_i_on_branch_j</th>\n",
       "      <th>n_muts_same_branch</th>\n",
       "      <th>pval_beta_i_on_j_BH_sig_sequential</th>\n",
       "      <th>pval_beta_i_on_j_BH_sig_simultaneous</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>521</th>\n",
       "      <td>764817</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>761155</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>277</td>\n",
       "      <td>1487</td>\n",
       "      <td>921</td>\n",
       "      <td>78</td>\n",
       "      <td>105</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>641</th>\n",
       "      <td>2986827</td>\n",
       "      <td>Rv2670c</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>95</td>\n",
       "      <td>1811</td>\n",
       "      <td>671</td>\n",
       "      <td>41</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455</th>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>1472362</td>\n",
       "      <td>None</td>\n",
       "      <td>783</td>\n",
       "      <td>115</td>\n",
       "      <td>5303</td>\n",
       "      <td>41</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>892</th>\n",
       "      <td>1472359</td>\n",
       "      <td>None</td>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>233</td>\n",
       "      <td>783</td>\n",
       "      <td>870</td>\n",
       "      <td>38</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>596</th>\n",
       "      <td>3067961</td>\n",
       "      <td>None</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>78</td>\n",
       "      <td>1811</td>\n",
       "      <td>286</td>\n",
       "      <td>29</td>\n",
       "      <td>24</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>581</th>\n",
       "      <td>3073679</td>\n",
       "      <td>None</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>53</td>\n",
       "      <td>1811</td>\n",
       "      <td>192</td>\n",
       "      <td>25</td>\n",
       "      <td>5</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>513</th>\n",
       "      <td>3073679</td>\n",
       "      <td>None</td>\n",
       "      <td>761155</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>53</td>\n",
       "      <td>1487</td>\n",
       "      <td>192</td>\n",
       "      <td>21</td>\n",
       "      <td>7</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>544</th>\n",
       "      <td>764840</td>\n",
       "      <td>rpoC</td>\n",
       "      <td>761155</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>63</td>\n",
       "      <td>1487</td>\n",
       "      <td>192</td>\n",
       "      <td>20</td>\n",
       "      <td>17</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>1853001</td>\n",
       "      <td>rpmI</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>26</td>\n",
       "      <td>1811</td>\n",
       "      <td>280</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>512</th>\n",
       "      <td>1853001</td>\n",
       "      <td>rpmI</td>\n",
       "      <td>761155</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>26</td>\n",
       "      <td>1487</td>\n",
       "      <td>280</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593</th>\n",
       "      <td>2747480</td>\n",
       "      <td>folC</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>38</td>\n",
       "      <td>1811</td>\n",
       "      <td>74</td>\n",
       "      <td>16</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>583</th>\n",
       "      <td>2487024</td>\n",
       "      <td>Rv2219A</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>19</td>\n",
       "      <td>1811</td>\n",
       "      <td>632</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>587</th>\n",
       "      <td>2620781</td>\n",
       "      <td>dnaG</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>18</td>\n",
       "      <td>1811</td>\n",
       "      <td>128</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>523</th>\n",
       "      <td>761156</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>761155</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>37</td>\n",
       "      <td>1487</td>\n",
       "      <td>152</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584</th>\n",
       "      <td>2241091</td>\n",
       "      <td>ctpF</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>19</td>\n",
       "      <td>1811</td>\n",
       "      <td>632</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>894</th>\n",
       "      <td>3073679</td>\n",
       "      <td>None</td>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>53</td>\n",
       "      <td>783</td>\n",
       "      <td>192</td>\n",
       "      <td>15</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>893</th>\n",
       "      <td>1853001</td>\n",
       "      <td>rpmI</td>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>26</td>\n",
       "      <td>783</td>\n",
       "      <td>280</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>582</th>\n",
       "      <td>3521253</td>\n",
       "      <td>nuoJ</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>19</td>\n",
       "      <td>1811</td>\n",
       "      <td>632</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>588</th>\n",
       "      <td>3811039</td>\n",
       "      <td>Rv3395c</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>18</td>\n",
       "      <td>1811</td>\n",
       "      <td>128</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>896</th>\n",
       "      <td>2487024</td>\n",
       "      <td>Rv2219A</td>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>19</td>\n",
       "      <td>783</td>\n",
       "      <td>632</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>895</th>\n",
       "      <td>3521253</td>\n",
       "      <td>nuoJ</td>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>19</td>\n",
       "      <td>783</td>\n",
       "      <td>632</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>897</th>\n",
       "      <td>2241091</td>\n",
       "      <td>ctpF</td>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>19</td>\n",
       "      <td>783</td>\n",
       "      <td>632</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>766</th>\n",
       "      <td>2715342</td>\n",
       "      <td>None</td>\n",
       "      <td>1473246</td>\n",
       "      <td>None</td>\n",
       "      <td>139</td>\n",
       "      <td>560</td>\n",
       "      <td>350</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>905</th>\n",
       "      <td>2747480</td>\n",
       "      <td>folC</td>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>38</td>\n",
       "      <td>783</td>\n",
       "      <td>74</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899</th>\n",
       "      <td>2699402</td>\n",
       "      <td>Rv2402</td>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>19</td>\n",
       "      <td>783</td>\n",
       "      <td>204</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>586</th>\n",
       "      <td>2699402</td>\n",
       "      <td>Rv2402</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>19</td>\n",
       "      <td>1811</td>\n",
       "      <td>204</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>898</th>\n",
       "      <td>666107</td>\n",
       "      <td>pncB2</td>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>18</td>\n",
       "      <td>783</td>\n",
       "      <td>196</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666</th>\n",
       "      <td>3841083</td>\n",
       "      <td>alr</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>21</td>\n",
       "      <td>1811</td>\n",
       "      <td>234</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>514</th>\n",
       "      <td>3521253</td>\n",
       "      <td>nuoJ</td>\n",
       "      <td>761155</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>19</td>\n",
       "      <td>1487</td>\n",
       "      <td>632</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>515</th>\n",
       "      <td>2487024</td>\n",
       "      <td>Rv2219A</td>\n",
       "      <td>761155</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>19</td>\n",
       "      <td>1487</td>\n",
       "      <td>632</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>516</th>\n",
       "      <td>2241091</td>\n",
       "      <td>ctpF</td>\n",
       "      <td>761155</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>19</td>\n",
       "      <td>1487</td>\n",
       "      <td>632</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>585</th>\n",
       "      <td>666107</td>\n",
       "      <td>pncB2</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>18</td>\n",
       "      <td>1811</td>\n",
       "      <td>196</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>589</th>\n",
       "      <td>4195074</td>\n",
       "      <td>ctpJ</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>17</td>\n",
       "      <td>1811</td>\n",
       "      <td>198</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>518</th>\n",
       "      <td>3811039</td>\n",
       "      <td>Rv3395c</td>\n",
       "      <td>761155</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>18</td>\n",
       "      <td>1487</td>\n",
       "      <td>128</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>2620781</td>\n",
       "      <td>dnaG</td>\n",
       "      <td>761155</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>18</td>\n",
       "      <td>1487</td>\n",
       "      <td>128</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>900</th>\n",
       "      <td>2620781</td>\n",
       "      <td>dnaG</td>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>18</td>\n",
       "      <td>783</td>\n",
       "      <td>128</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>902</th>\n",
       "      <td>4195074</td>\n",
       "      <td>ctpJ</td>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>17</td>\n",
       "      <td>783</td>\n",
       "      <td>198</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>901</th>\n",
       "      <td>3811039</td>\n",
       "      <td>Rv3395c</td>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>18</td>\n",
       "      <td>783</td>\n",
       "      <td>128</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>590</th>\n",
       "      <td>2101193</td>\n",
       "      <td>ureD</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>17</td>\n",
       "      <td>1811</td>\n",
       "      <td>426</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>591</th>\n",
       "      <td>4090823</td>\n",
       "      <td>Rv3649</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>18</td>\n",
       "      <td>1811</td>\n",
       "      <td>426</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>904</th>\n",
       "      <td>4090823</td>\n",
       "      <td>Rv3649</td>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>18</td>\n",
       "      <td>783</td>\n",
       "      <td>426</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>903</th>\n",
       "      <td>2101193</td>\n",
       "      <td>ureD</td>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>17</td>\n",
       "      <td>783</td>\n",
       "      <td>426</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>602</th>\n",
       "      <td>259092</td>\n",
       "      <td>Rv0216</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>13</td>\n",
       "      <td>1811</td>\n",
       "      <td>208</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>603</th>\n",
       "      <td>96899</td>\n",
       "      <td>None</td>\n",
       "      <td>2155168</td>\n",
       "      <td>katG</td>\n",
       "      <td>14</td>\n",
       "      <td>1811</td>\n",
       "      <td>208</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>940</th>\n",
       "      <td>3568759</td>\n",
       "      <td>None</td>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>31</td>\n",
       "      <td>783</td>\n",
       "      <td>62</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>908</th>\n",
       "      <td>3073715</td>\n",
       "      <td>thyA</td>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>13</td>\n",
       "      <td>783</td>\n",
       "      <td>308</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>519</th>\n",
       "      <td>2101193</td>\n",
       "      <td>ureD</td>\n",
       "      <td>761155</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>17</td>\n",
       "      <td>1487</td>\n",
       "      <td>426</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>907</th>\n",
       "      <td>3568795</td>\n",
       "      <td>None</td>\n",
       "      <td>781687</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>15</td>\n",
       "      <td>783</td>\n",
       "      <td>64</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>520</th>\n",
       "      <td>4090823</td>\n",
       "      <td>Rv3649</td>\n",
       "      <td>761155</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>18</td>\n",
       "      <td>1487</td>\n",
       "      <td>426</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>525</th>\n",
       "      <td>259092</td>\n",
       "      <td>Rv0216</td>\n",
       "      <td>761155</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>13</td>\n",
       "      <td>1487</td>\n",
       "      <td>208</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     position_i gene_name_i  position_j gene_name_j  n_mutations_i  \\\n",
       "521      764817        rpoC      761155        rpoB            277   \n",
       "641     2986827     Rv2670c     2155168        katG             95   \n",
       "455      781687        rpsL     1472362        None            783   \n",
       "892     1472359        None      781687        rpsL            233   \n",
       "596     3067961        None     2155168        katG             78   \n",
       "581     3073679        None     2155168        katG             53   \n",
       "513     3073679        None      761155        rpoB             53   \n",
       "544      764840        rpoC      761155        rpoB             63   \n",
       "580     1853001        rpmI     2155168        katG             26   \n",
       "512     1853001        rpmI      761155        rpoB             26   \n",
       "593     2747480        folC     2155168        katG             38   \n",
       "583     2487024     Rv2219A     2155168        katG             19   \n",
       "587     2620781        dnaG     2155168        katG             18   \n",
       "523      761156        rpoB      761155        rpoB             37   \n",
       "584     2241091        ctpF     2155168        katG             19   \n",
       "894     3073679        None      781687        rpsL             53   \n",
       "893     1853001        rpmI      781687        rpsL             26   \n",
       "582     3521253        nuoJ     2155168        katG             19   \n",
       "588     3811039     Rv3395c     2155168        katG             18   \n",
       "896     2487024     Rv2219A      781687        rpsL             19   \n",
       "895     3521253        nuoJ      781687        rpsL             19   \n",
       "897     2241091        ctpF      781687        rpsL             19   \n",
       "766     2715342        None     1473246        None            139   \n",
       "905     2747480        folC      781687        rpsL             38   \n",
       "899     2699402      Rv2402      781687        rpsL             19   \n",
       "586     2699402      Rv2402     2155168        katG             19   \n",
       "898      666107       pncB2      781687        rpsL             18   \n",
       "666     3841083         alr     2155168        katG             21   \n",
       "514     3521253        nuoJ      761155        rpoB             19   \n",
       "515     2487024     Rv2219A      761155        rpoB             19   \n",
       "516     2241091        ctpF      761155        rpoB             19   \n",
       "585      666107       pncB2     2155168        katG             18   \n",
       "589     4195074        ctpJ     2155168        katG             17   \n",
       "518     3811039     Rv3395c      761155        rpoB             18   \n",
       "517     2620781        dnaG      761155        rpoB             18   \n",
       "900     2620781        dnaG      781687        rpsL             18   \n",
       "902     4195074        ctpJ      781687        rpsL             17   \n",
       "901     3811039     Rv3395c      781687        rpsL             18   \n",
       "590     2101193        ureD     2155168        katG             17   \n",
       "591     4090823      Rv3649     2155168        katG             18   \n",
       "904     4090823      Rv3649      781687        rpsL             18   \n",
       "903     2101193        ureD      781687        rpsL             17   \n",
       "602      259092      Rv0216     2155168        katG             13   \n",
       "603       96899        None     2155168        katG             14   \n",
       "940     3568759        None      781687        rpsL             31   \n",
       "908     3073715        thyA      781687        rpsL             13   \n",
       "519     2101193        ureD      761155        rpoB             17   \n",
       "907     3568795        None      781687        rpsL             15   \n",
       "520     4090823      Rv3649      761155        rpoB             18   \n",
       "525      259092      Rv0216      761155        rpoB             13   \n",
       "\n",
       "     n_mutations_j  n_branches_with_i  n_muts_i_on_branch_j  \\\n",
       "521           1487                921                    78   \n",
       "641           1811                671                    41   \n",
       "455            115               5303                    41   \n",
       "892            783                870                    38   \n",
       "596           1811                286                    29   \n",
       "581           1811                192                    25   \n",
       "513           1487                192                    21   \n",
       "544           1487                192                    20   \n",
       "580           1811                280                    18   \n",
       "512           1487                280                    17   \n",
       "593           1811                 74                    16   \n",
       "583           1811                632                    15   \n",
       "587           1811                128                    15   \n",
       "523           1487                152                    15   \n",
       "584           1811                632                    15   \n",
       "894            783                192                    15   \n",
       "893            783                280                    15   \n",
       "582           1811                632                    15   \n",
       "588           1811                128                    15   \n",
       "896            783                632                    14   \n",
       "895            783                632                    14   \n",
       "897            783                632                    14   \n",
       "766            560                350                    14   \n",
       "905            783                 74                    13   \n",
       "899            783                204                    13   \n",
       "586           1811                204                    13   \n",
       "898            783                196                    13   \n",
       "666           1811                234                    13   \n",
       "514           1487                632                    13   \n",
       "515           1487                632                    13   \n",
       "516           1487                632                    13   \n",
       "585           1811                196                    13   \n",
       "589           1811                198                    12   \n",
       "518           1487                128                    12   \n",
       "517           1487                128                    12   \n",
       "900            783                128                    12   \n",
       "902            783                198                    12   \n",
       "901            783                128                    12   \n",
       "590           1811                426                    11   \n",
       "591           1811                426                    11   \n",
       "904            783                426                    10   \n",
       "903            783                426                    10   \n",
       "602           1811                208                    10   \n",
       "603           1811                208                    10   \n",
       "940            783                 62                    10   \n",
       "908            783                308                     9   \n",
       "519           1487                426                     9   \n",
       "907            783                 64                     9   \n",
       "520           1487                426                     9   \n",
       "525           1487                208                     9   \n",
       "\n",
       "     n_muts_same_branch  pval_beta_i_on_j_BH_sig_sequential  \\\n",
       "521                 105                                True   \n",
       "641                  29                                True   \n",
       "455                   0                                True   \n",
       "892                   1                                True   \n",
       "596                  24                                True   \n",
       "581                   5                                True   \n",
       "513                   7                                True   \n",
       "544                  17                                True   \n",
       "580                   0                                True   \n",
       "512                   0                                True   \n",
       "593                   6                                True   \n",
       "583                   0                                True   \n",
       "587                   0                                True   \n",
       "523                  15                                True   \n",
       "584                   0                                True   \n",
       "894                   2                                True   \n",
       "893                   0                                True   \n",
       "582                   0                                True   \n",
       "588                   0                                True   \n",
       "896                   0                                True   \n",
       "895                   0                                True   \n",
       "897                   0                                True   \n",
       "766                   1                                True   \n",
       "905                   2                                True   \n",
       "899                   0                                True   \n",
       "586                   0                                True   \n",
       "898                   0                                True   \n",
       "666                   2                                True   \n",
       "514                   0                                True   \n",
       "515                   0                                True   \n",
       "516                   0                                True   \n",
       "585                   0                                True   \n",
       "589                   0                                True   \n",
       "518                   0                                True   \n",
       "517                   0                                True   \n",
       "900                   0                                True   \n",
       "902                   0                                True   \n",
       "901                   0                                True   \n",
       "590                   0                                True   \n",
       "591                   0                                True   \n",
       "904                   0                                True   \n",
       "903                   0                                True   \n",
       "602                   0                                True   \n",
       "603                   0                                True   \n",
       "940                   1                                True   \n",
       "908                   0                                True   \n",
       "519                   0                                True   \n",
       "907                   0                                True   \n",
       "520                   0                                True   \n",
       "525                   0                                True   \n",
       "\n",
       "    pval_beta_i_on_j_BH_sig_simultaneous  \n",
       "521                                  NaN  \n",
       "641                                  NaN  \n",
       "455                                  NaN  \n",
       "892                                  NaN  \n",
       "596                                  NaN  \n",
       "581                                  NaN  \n",
       "513                                  NaN  \n",
       "544                                  NaN  \n",
       "580                                  NaN  \n",
       "512                                  NaN  \n",
       "593                                  NaN  \n",
       "583                                  NaN  \n",
       "587                                  NaN  \n",
       "523                                  NaN  \n",
       "584                                  NaN  \n",
       "894                                  NaN  \n",
       "893                                  NaN  \n",
       "582                                  NaN  \n",
       "588                                  NaN  \n",
       "896                                  NaN  \n",
       "895                                  NaN  \n",
       "897                                  NaN  \n",
       "766                                  NaN  \n",
       "905                                  NaN  \n",
       "899                                  NaN  \n",
       "586                                  NaN  \n",
       "898                                  NaN  \n",
       "666                                  NaN  \n",
       "514                                  NaN  \n",
       "515                                  NaN  \n",
       "516                                  NaN  \n",
       "585                                  NaN  \n",
       "589                                  NaN  \n",
       "518                                  NaN  \n",
       "517                                  NaN  \n",
       "900                                  NaN  \n",
       "902                                  NaN  \n",
       "901                                  NaN  \n",
       "590                                  NaN  \n",
       "591                                  NaN  \n",
       "904                                  NaN  \n",
       "903                                  NaN  \n",
       "602                                  NaN  \n",
       "603                                  NaN  \n",
       "940                                  NaN  \n",
       "908                                  NaN  \n",
       "519                                  NaN  \n",
       "907                                  NaN  \n",
       "520                                  NaN  \n",
       "525                                  NaN  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "antibiotic_hits.query(\"pval_beta_i_on_j_BH_sig_sequential\").sort_values(\"n_muts_i_on_branch_j\", ascending=False)[display_positions].head(50)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8f008b6-fa10-433f-b3d7-043d0a4b6461",
   "metadata": {},
   "source": [
    "## Get the nucleotide changes for individual SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "422d4076-e3f6-4e7d-9dd9-cc7e245d0bfd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_pos</th>\n",
       "      <th>SNP_type</th>\n",
       "      <th>AA_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2153889_C</th>\n",
       "      <td>2153889</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>katG</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>2223</td>\n",
       "      <td>N</td>\n",
       "      <td>*741W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2153891_G</th>\n",
       "      <td>2153891</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>katG</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>2221</td>\n",
       "      <td>N</td>\n",
       "      <td>*741R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2153897_T</th>\n",
       "      <td>2153897</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>katG</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>2215</td>\n",
       "      <td>N</td>\n",
       "      <td>V739M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2153905_T</th>\n",
       "      <td>2153905</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>katG</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>2207</td>\n",
       "      <td>N</td>\n",
       "      <td>R736K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2153908_G</th>\n",
       "      <td>2153908</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>katG</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>2204</td>\n",
       "      <td>N</td>\n",
       "      <td>D735A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2156104_C</th>\n",
       "      <td>2156104</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>katG</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>8</td>\n",
       "      <td>N</td>\n",
       "      <td>E3G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2156106_A</th>\n",
       "      <td>2156106</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>katG</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>6</td>\n",
       "      <td>S</td>\n",
       "      <td>P2P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2156109_A</th>\n",
       "      <td>2156109</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>katG</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>3</td>\n",
       "      <td>S</td>\n",
       "      <td>V1V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2156109_T</th>\n",
       "      <td>2156109</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>katG</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>3</td>\n",
       "      <td>S</td>\n",
       "      <td>V1V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2156110_G</th>\n",
       "      <td>2156110</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>katG</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>2</td>\n",
       "      <td>N</td>\n",
       "      <td>V1A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>845 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               pos ref alt          gene_category gene_name  gene_id gene_pos  \\\n",
       "2153889_C  2153889   T   C  Antibiotic Resistance      katG  Rv1908c     2223   \n",
       "2153891_G  2153891   A   G  Antibiotic Resistance      katG  Rv1908c     2221   \n",
       "2153897_T  2153897   C   T  Antibiotic Resistance      katG  Rv1908c     2215   \n",
       "2153905_T  2153905   C   T  Antibiotic Resistance      katG  Rv1908c     2207   \n",
       "2153908_G  2153908   T   G  Antibiotic Resistance      katG  Rv1908c     2204   \n",
       "...            ...  ..  ..                    ...       ...      ...      ...   \n",
       "2156104_C  2156104   T   C  Antibiotic Resistance      katG  Rv1908c        8   \n",
       "2156106_A  2156106   G   A  Antibiotic Resistance      katG  Rv1908c        6   \n",
       "2156109_A  2156109   C   A  Antibiotic Resistance      katG  Rv1908c        3   \n",
       "2156109_T  2156109   C   T  Antibiotic Resistance      katG  Rv1908c        3   \n",
       "2156110_G  2156110   A   G  Antibiotic Resistance      katG  Rv1908c        2   \n",
       "\n",
       "          SNP_type AA_change  \n",
       "2153889_C        N     *741W  \n",
       "2153891_G        N     *741R  \n",
       "2153897_T        N     V739M  \n",
       "2153905_T        N     R736K  \n",
       "2153908_G        N     D735A  \n",
       "...            ...       ...  \n",
       "2156104_C        N       E3G  \n",
       "2156106_A        S       P2P  \n",
       "2156109_A        S       V1V  \n",
       "2156109_T        S       V1V  \n",
       "2156110_G        N       V1A  \n",
       "\n",
       "[845 rows x 9 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps = pd.read_pickle(\"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_SNP_functional_annotation.pkl\")\n",
    "snps.query(\"gene_name=='katG'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8140230f-1b6b-4fb7-88eb-91548bb043e1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
