{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Objective:\n",
    "For each drug, test heritability of \n",
    "all minor allele frequnecy (MAF) 0p001 loci\n",
    "all homoplasic sites\n",
    "all mutations in known ABR sites\n",
    "all mutations in possible ABR sites\n",
    "all mutations in known ABR sites + all mutations with epistatic relationship (single site model)\n",
    "all mutations in possible ABR sites + all mutations with epistatic relationship (single site model)\n",
    "all mutations in known ABR sites + positions with epistatic relationship (single site model)+ epistasis within and between AB sites and outside sites\n",
    "\n",
    "# Note: this code relies on having already run 00.run_gemma_cross_MIC.py in order to generate the GRMs for MAF 0p001"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Command line tool for executing GEMMA\n",
    "\n",
    "Authors:\n",
    "Anna G. Green\n",
    "\"\"\"\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import numpy as np\n",
    "import ruamel.yaml as yaml\n",
    "import sys\n",
    "from evcouplings.utils import valid_file, verify_resources, run\n",
    "from evcouplings.align import Alignment\n",
    "import os\n",
    "from copy import deepcopy\n",
    "from scipy import stats\n",
    "\n",
    "ALPHABET_NOGAP = \"ACGT\"\n",
    "GAP_CHAR = \"-\"\n",
    "ALPHABET = GAP_CHAR + ALPHABET_NOGAP\n",
    "\n",
    "genotype_translator = {\n",
    "    0:\"A\", 1:\"C\", 2:\"G\", 3:\"T\", 9:\"-\"\n",
    "}\n",
    "genotype_translator_r = {y:x for x,y in genotype_translator.items()}\n",
    "def map_matrix(matrix, map_):\n",
    "    \"\"\"\n",
    "    Map elements in a numpy array using alphabet\n",
    "    Parameters\n",
    "    ----------\n",
    "    matrix : np.array\n",
    "        Matrix that should be remapped\n",
    "    map_ : defaultdict\n",
    "        Map that will be applied to matrix elements\n",
    "    Returns\n",
    "    -------\n",
    "    np.array\n",
    "        Remapped matrix\n",
    "    \"\"\"\n",
    "    return np.vectorize(map_.__getitem__)(matrix)\n",
    "\n",
    "\n",
    "def calculate_frequencies(matrix, num_symbols):\n",
    "    \"\"\"\n",
    "    Calculate single-site frequencies of symbols in alignment\n",
    "    Parameters\n",
    "    ----------\n",
    "    matrix : np.array\n",
    "        N x L matrix containing N sequences of length L.\n",
    "        Matrix must be mapped to range(0, num_symbols) using\n",
    "        map_matrix function\n",
    "    num_symbols : int\n",
    "        Number of different symbols contained in alignment\n",
    "    Returns\n",
    "    -------\n",
    "    np.array\n",
    "        Matrix of size L x num_symbols containing relative\n",
    "        column frequencies of all characters\n",
    "    \"\"\"\n",
    "    N, L = matrix.shape\n",
    "    fi = np.zeros((L, num_symbols))\n",
    "    for s in range(N):\n",
    "        for i in range(L):\n",
    "            fi[i, matrix[s, i]] += 1\n",
    "    return np.divide(fi, N)\n",
    "\n",
    "class GenotypeMatrix:\n",
    "    \"\"\"\n",
    "    Takes inspiration from the evcouplings python package Alignment class\n",
    "    \"\"\"\n",
    "    def __init__(self, strain_list, position_list, genotype_data=None, matrix_data=None):\n",
    "\n",
    "        self.positions = position_list\n",
    "        self.position_to_index = {p:idx for idx, p in enumerate(self.positions)}\n",
    "        self.strains = strain_list\n",
    "        self.strain_to_index = {p:idx for idx, p in enumerate(self.strains)}\n",
    "\n",
    "        # If user provided genotype data, do some sanity checks\n",
    "        self.genotype_data = genotype_data\n",
    "        # Make sure that every position and every genotype is represented in the genotype_data\n",
    "        if not self.genotype_data is None:\n",
    "            assert set(self.positions).issubset(genotype_data.POS)\n",
    "            assert set(self.strains).issubset(genotype_data.STRAIN)\n",
    "\n",
    "        # if user provided matrix data, set it up\n",
    "        if not matrix_data is None:\n",
    "            assert matrix_data.shape[0] == len(self.strains)\n",
    "            assert matrix_data.shape[1] == len(self.positions)\n",
    "            self.matrix = matrix_data\n",
    "            self.reference_alleles = deepcopy(self.matrix[0, :])\n",
    "            \n",
    "            if 0 in matrix_data:\n",
    "                self.matrix_mapped = matrix_data\n",
    "            else:\n",
    "                self.matrix_mapped = None\n",
    "            \n",
    "        else:\n",
    "            self.matrix = np.zeros((len(self.strains), len(self.positions)), dtype=str)\n",
    "            self.reference_alleles = None\n",
    "            self.matrix_mapped = None\n",
    "\n",
    "        self.alphabet_map = {\n",
    "            c: i for i, c in enumerate(ALPHABET)\n",
    "        }\n",
    "        self.num_symbols = len(self.alphabet_map)\n",
    "        \n",
    "        self._frequencies = None\n",
    "        self._major_alleles = None\n",
    "        \n",
    "\n",
    "    @classmethod\n",
    "    def from_df(cls, file):\n",
    "        \"\"\"\n",
    "        Initializes a GenotypeMatrix using an input pd.DataFrame, where rows = strains and columns=positions, and\n",
    "        the data will become the contents of self.matrix\n",
    "        \"\"\"\n",
    "        df = pd.read_csv(file, index_col=0, header=0)\n",
    "\n",
    "        return cls(\n",
    "            list(df.index), list(df.columns), genotype_data=None, matrix_data=df.values\n",
    "        )\n",
    "\n",
    "    def make_matrix(self):\n",
    "        '''\n",
    "        Populates the matrix of strains by positions with genotype data from\n",
    "        the input genotype data file via the following heuristic:\n",
    "\n",
    "        0. Initialize every position in every genome as the reference allele\n",
    "        1. For each position in each strain, determine if there is any information\n",
    "            in the genotype data input. If not, leave that position as Reference. If yes,\n",
    "            proceed to step 2.\n",
    "        2. If the alternate allele is an insertion or a non-standard character, insert a gap\n",
    "            at that position.\n",
    "        3. If the alternate allele did not pass quality threshold (determined by QUAL flag being 0),\n",
    "            insert a gap at that position\n",
    "        4. If alternate allele is a SNP and is not low quality, replace reference with alternate allale\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        np.ndarray:\n",
    "            an N_strains x N_positions array of strings\n",
    "        '''\n",
    "        print(\"preparing the reference\")\n",
    "\n",
    "        position_subset = self.genotype_data.query(\"POS in @self.positions\")\n",
    "\n",
    "        position_groups = position_subset.groupby(\"POS\")\n",
    "        for position, subset in position_groups:\n",
    "            allele = subset.loc[subset.index[0], \"REF\"]\n",
    "            self.matrix[:, self.position_to_index[position]] = allele\n",
    "\n",
    "        self.reference_alleles = deepcopy(self.matrix[0, :])\n",
    "        #print(\"assigned the reference\", self.reference_alleles)\n",
    "\n",
    "        # iterate through each strain\n",
    "        print(\"filling in the isolate genotypes\")\n",
    "        strain_group = position_subset.query(\"STRAIN in @self.strains\").groupby(\"STRAIN\")\n",
    "\n",
    "        for strain, subset in strain_group:\n",
    "\n",
    "            # iterate through each position\n",
    "            position_groups = subset.groupby(\"POS\")\n",
    "            for position, subsubset in position_groups:\n",
    "\n",
    "                # if we didn't find any information for this position, leave as reference allele\n",
    "                if len(subsubset) == 0:\n",
    "                    continue\n",
    "\n",
    "                subsubset = subsubset.iloc[0,:]\n",
    "\n",
    "                # If we did find information but its an indel, treat as missing\n",
    "                if len(subsubset.ALT) > 1 or  not subsubset.ALT in ALPHABET:\n",
    "                    self.matrix[self.strain_to_index[strain],self.position_to_index[position]] = \"-\"\n",
    "\n",
    "                # If we found information but it didn't pass quality filter, treat as missing\n",
    "                elif subsubset.QUAL==0:\n",
    "                    self.matrix[self.strain_to_index[strain],self.position_to_index[position]] = \"-\"\n",
    "\n",
    "                # else, change to alterante allele\n",
    "                else:\n",
    "                    self.matrix[self.strain_to_index[strain],self.position_to_index[position]] = subsubset.ALT\n",
    "\n",
    "\n",
    "    def __ensure_mapped_matrix(self):\n",
    "        \"\"\"\n",
    "        Ensure self.matrix_mapped exists\n",
    "        \"\"\"\n",
    "        if self.matrix_mapped is None:\n",
    "            self.matrix_mapped = map_matrix(\n",
    "                self.matrix, self.alphabet_map\n",
    "            )\n",
    "\n",
    "    @property\n",
    "    def major_alleles(self):\n",
    "        \"\"\"\n",
    "        Returns/calculates identity of major allele for each position in alignment.\n",
    "        Also sets self._major_alleles member variable for later reuse.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        np.array\n",
    "            Reference to self._major_alleles\n",
    "        \"\"\"\n",
    "        if self._major_alleles is None:\n",
    "\n",
    "            self._major_alleles = np.argmax(\n",
    "                self.frequencies, axis=1\n",
    "            )\n",
    "\n",
    "        return self._major_alleles\n",
    "\n",
    "    @property\n",
    "    def frequencies(self):\n",
    "        \"\"\"\n",
    "        Returns/calculates single-site frequencies of symbols in alignment.\n",
    "        Also sets self._frequencies member variable for later reuse.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        np.array\n",
    "            Reference to self._frequencies\n",
    "        \"\"\"\n",
    "        if self._frequencies is None:\n",
    "            self.__ensure_mapped_matrix()\n",
    "\n",
    "            self._frequencies = calculate_frequencies(\n",
    "                self.matrix_mapped, self.num_symbols\n",
    "            )\n",
    "\n",
    "        return self._frequencies\n",
    "\n",
    "    def write_fasta(self, fileobj, strain_list=None):\n",
    "        \"\"\"\n",
    "        Creates a fasta-formatted file of pseudo-sequences for strains and positions in self.matrix\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        fileobj: filehandle\n",
    "            fileobject to which to write\n",
    "        strain_list: list of str, optional (default=None)\n",
    "            List of strain names, in order, which will be written to file. If None,\n",
    "            will write all strains in self.strains\n",
    "\n",
    "        \"\"\"\n",
    "        if strain_list is None:\n",
    "            strain_list = self.strains\n",
    "\n",
    "        for strain in strain_list:\n",
    "            seq = \"\".join(self.matrix[self.strain_to_index[strain],:])\n",
    "            fileobj.write(f\">{strain}\\n{seq}\\n\")\n",
    "\n",
    "    def write_GEMMA(self, fileobj, positions=None, major_allele_string=\"\\\"A\\\"\", minor_allele_string=\"TRUE\",\n",
    "        major_allele_code='0', minor_allele_code='1',gap_code=\"NA\"):\n",
    "        \"\"\"\n",
    "        Creates a file of loci in correct input format for GEMMA\n",
    "\n",
    "        User warning: Changing the optional variables to values other than their defaults may result in errors when\n",
    "        using the program GEMMA. Verify that all inputs are still valid before proceeding.\n",
    "\n",
    "        Gemma input format:\n",
    "        position_name,major_allele_string,minor_allele_string,{allele code for strain 1, ..., allele code for strain N}\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        fileobj: filehandle\n",
    "            file to which to write\n",
    "        positions: list of str, optional (default: None)\n",
    "            specify subset of positions to write. If None, use all positions\n",
    "        major_allele_string: str, optional (default \"A\")\n",
    "            string used to represent the major allele\n",
    "        minor_allele_string: str, optional (default \"T\")\n",
    "            string used to represent the minor allele\n",
    "        major_allele_code: str, optional (default \"1\")\n",
    "            string used to code for the major allele in the input\n",
    "        minor_allele_code: str, optional (default \"0\")\n",
    "            string used to code for the major allele in the input\n",
    "        gap_code: str, optional (default \"NaN\")\n",
    "            string used to code for the gap/missing data in the input\n",
    "\n",
    "        \"\"\"\n",
    "\n",
    "        if positions is None:\n",
    "            positions = self.positions\n",
    "\n",
    "        self.__ensure_mapped_matrix()\n",
    "        if not 0 in self.reference_alleles:\n",
    "            reference_alleles = [self.alphabet_map[x] for x in self.reference_alleles]\n",
    "        else:\n",
    "            reference_alleles = self.reference_alleles\n",
    "        print(\"calculating reference allele matrix with gap code\", gap_code)  \n",
    "        major_allele_matrix = calculate_reference_allele_matrix_for_pair_regression(\n",
    "            self.matrix_mapped,\n",
    "            reference_alleles,\n",
    "            encode_gaps=True,\n",
    "            gap_code=gap_code,\n",
    "            ref_allele_code=major_allele_code,\n",
    "            non_ref_allele_code=minor_allele_code\n",
    "        )\n",
    "\n",
    "        for position in positions:\n",
    "            # slice corresponding to all strains\n",
    "            vec = major_allele_matrix[:, self.position_to_index[position]]\n",
    "\n",
    "            fileobj.write(f'{position},{major_allele_string},{minor_allele_string},{\",\".join(vec)}\\n')\n",
    "\n",
    "    def to_df(self):\n",
    "        \"\"\"\n",
    "        Creates a pd.DataFrame, with rows corresponding to strains, columns corresponding to positions,\n",
    "        and entries corresponding to the alleles in self.matrix\n",
    "        Returns\n",
    "        -------\n",
    "        pd.DataFrame\n",
    "        \"\"\"\n",
    "        df = pd.DataFrame(self.matrix, index=self.strains, columns=self.positions)\n",
    "        return df\n",
    "\n",
    "def write_phenotype_file_MIC(gemma_input_phenotype_file, phenotypes_found, isolates_with_phenotype, drug):\n",
    "    \n",
    "    \"\"\"\n",
    "    Convert input phenotype file into correct format for GEMMA and saves file.\n",
    "    Adds an ancestral sequence and assumes it is sensitive\n",
    "    \n",
    "    Parameters\n",
    "    ---------\n",
    "    gemma_input_phenotype_file: str\n",
    "        name of file to save\n",
    "    \n",
    "    phenotypes_found: pd.DataFrame\n",
    "    \n",
    "    isolates: pd.DataFrame\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    phenotypes = isolates_with_phenotype[[\"isolate_ID\"]].merge(\n",
    "        phenotypes_found[[\"ROLLINGDB_ID\", drug]], right_on=\"ROLLINGDB_ID\", left_on=\"isolate_ID\", how=\"left\"\n",
    "    )\n",
    "    print(len(phenotypes))\n",
    "    phenotypes = phenotypes.drop_duplicates(subset=[\"isolate_ID\"])\n",
    "    phenotypes = phenotypes.drop_duplicates(subset=[\"ROLLINGDB_ID\"])\n",
    "    # Add the ancestor sequence, assuming ancestor is sensitive\n",
    "    ancestor = pd.DataFrame({\n",
    "        \"isolate_ID\": [\"MT_H37Rv_anc\"], \n",
    "        \"ROLLINGDB_ID\": [None], \n",
    "        drug: phenotypes[drug].min()\n",
    "    })\n",
    "    phenotypes = pd.concat([ancestor, phenotypes])\n",
    "    print(len(phenotypes))\n",
    "    phenotypes.iloc[:,-1].to_csv(gemma_input_phenotype_file, header=False, index=None, na_rep=\"NA\")\n",
    "    \n",
    "    \n",
    "def translate_genotypes(matrix, translator):\n",
    "    \"\"\"\n",
    "    Translate matrix of genotypes into string characters. Used for converting Roger's numeric genotype \n",
    "    matrix back into strings of ATCG-\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    \n",
    "    matrix: np.array \n",
    "        input genotype matrix\n",
    "    translator: dict\n",
    "        translator for converting entires of matrix to strings\n",
    "    Returns\n",
    "    -------\n",
    "    np.array of str, translated matrix\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    translated_matrix = np.zeros_like(matrix, dtype='str')\n",
    "    for i in range(matrix.shape[0]):\n",
    "        translated_matrix[i, :] = [translator[x] for x in matrix[i,:]]\n",
    "    return translated_matrix\n",
    "\n",
    "\n",
    "# Add the ancestral sequence to the sequence matrix\n",
    "def add_ancestor(mtb_ancestor_fasta, snps_to_keep, matrix, isolates_with_phenotype):\n",
    "    mtb_ancestor = Alignment.from_file(open(mtb_ancestor_fasta, \"r\"))\n",
    "    indices_to_slice = [x-1 for x in snps_to_keep.pos]\n",
    "    ancestor_sequence = mtb_ancestor.matrix[:, indices_to_slice]\n",
    "\n",
    "    matrix  = np.concatenate([ancestor_sequence, matrix])\n",
    "    isolate_labels = [\"MTB_ancestor\"] + list(isolates_with_phenotype.isolate_ID)\n",
    "    \n",
    "    return matrix, isolate_labels\n",
    "    \n",
    "def write_phenotype_file(gemma_input_phenotype_file, phenotypes_found, isolates_with_phenotype):\n",
    "    \n",
    "    \"\"\"\n",
    "    Convert input phenotype file into correct format for GEMMA and saves file.\n",
    "    Adds an ancestral sequence and assumes it is sensitive\n",
    "    \n",
    "    Parameters\n",
    "    ---------\n",
    "    gemma_input_phenotype_file: str\n",
    "        name of file to save\n",
    "    \n",
    "    phenotypes_found: pd.DataFrame\n",
    "    \n",
    "    isolates: pd.DataFrame\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    phenotypes = isolates_with_phenotype[[\"isolate_ID\"]].merge(\n",
    "        phenotypes_found[[\"Isolate\", kwargs['drug']]], left_on=\"isolate_ID\", right_on=\"Isolate\", how=\"left\"\n",
    "    )\n",
    "    phenotypes = phenotypes.drop_duplicates(subset=[\"isolate_ID\"])\n",
    "    # Add the ancestor sequence, assuming ancestor is sensitive\n",
    "    ancestor = pd.DataFrame({\n",
    "        \"Isolate\": [\"MT_H37Rv_anc\"], \n",
    "        \"isolate_ID\": [None], \n",
    "        kwargs['drug']: \"S\"\n",
    "    })\n",
    "    phenotypes = pd.concat([ancestor, phenotypes])\n",
    "    \n",
    "    encoding = {\"R\": 1, \"S\": 0}\n",
    "    phenotypes[kwargs['drug']] = [encoding[x] for x in phenotypes[kwargs['drug']]]\n",
    "\n",
    "    phenotypes.iloc[:,-1].to_csv(gemma_input_phenotype_file, header=False, index=None, na_rep=\"NA\")\n",
    "    \n",
    "def calculate_reference_allele_matrix_for_pair_regression(matrix, reference_alleles, encode_gaps=True, gap_code=np.nan,\n",
    "        ref_allele_code=0, non_ref_allele_code=1):\n",
    "    \"\"\"\n",
    "    Encodes a mapped genotype matrix with reference and non-reference allele codes\n",
    "    NOTE: This function currently only works for 5-character (with gaps) alphabet\n",
    "\n",
    "    ----------\n",
    "    matrix : np.array\n",
    "        N x L matrix containing N sequences of length L.\n",
    "        Matrix must be mapped to range(0, num_symbols) using\n",
    "        map_matrix function\n",
    "    reference_alleles : np.array\n",
    "        1 x L matrix with coding of reference allele for each position in L\n",
    "    encode_gaps: boolean, optional (default=True)\n",
    "        If True, gaps will be fille with gap_code. If False, gaps will be\n",
    "        eliminated (ie, replaced with the major allele code)\n",
    "    gap_code: str or numeric, optional (default=np.nan)\n",
    "        encoding to use for gap characters if nogap=False\n",
    "    ref_allele_code: str or numeric, optional (default=0)\n",
    "        encoding to use for major alleles\n",
    "    non_ref_allele_code: str or numeric, optional (default=1)\n",
    "        encoding to use for minor alleles\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    np.array\n",
    "        Matrix of size N x L with reference/non-reference allele encoding\n",
    "        for all positions in L\n",
    "    \"\"\"\n",
    "\n",
    "    if encode_gaps:\n",
    "        gap_code=gap_code\n",
    "        print(\"encoding gaps with gap code\", gap_code)\n",
    "    else:\n",
    "        gap_code=ref_allele_code\n",
    "        print(\"encoding gaps with reference allele\", ref_allele_code)\n",
    "\n",
    "    encoded_matrix = np.zeros(shape=matrix.shape, dtype=object)\n",
    "    \n",
    "    print(type(matrix))\n",
    "\n",
    "    # For each column\n",
    "    for i in np.arange(0, encoded_matrix.shape[1]):\n",
    "\n",
    "        #grab that column\n",
    "        vec = matrix[:, i]\n",
    "\n",
    "        # create an encoded version of the column, with minor allele code as default\n",
    "        encoded_vec = np.array([non_ref_allele_code] * len(vec), dtype=object)\n",
    "\n",
    "        # If the row has a major allele, replace in encoded vec\n",
    "        is_ref_allele = vec == reference_alleles[i]\n",
    "        encoded_vec[is_ref_allele] = ref_allele_code\n",
    "\n",
    "        # If the row has a gap character, replace in encoded vec\n",
    "        is_gap = vec ==9\n",
    "        encoded_vec[is_gap] = gap_code\n",
    "\n",
    "        # save in the encoded matrix\n",
    "        encoded_matrix[:, i] = encoded_vec\n",
    "\n",
    "    return encoded_matrix\n",
    "\n",
    "def calculate_reference_allele_matrix(matrix, reference_alleles, encode_gaps=True, gap_code=np.nan,\n",
    "        ref_allele_code=0, non_ref_allele_code=1):\n",
    "    \"\"\"\n",
    "    Encodes a mapped genotype matrix with reference and non-reference allele codes\n",
    "    NOTE: This function currently only works for 5-character (with gaps) alphabet\n",
    "\n",
    "    ----------\n",
    "    matrix : np.array\n",
    "        N x L matrix containing N sequences of length L.\n",
    "        Matrix must be mapped to range(0, num_symbols) using\n",
    "        map_matrix function\n",
    "    reference_alleles : np.array\n",
    "        1 x L matrix with coding of reference allele for each position in L\n",
    "    encode_gaps: boolean, optional (default=True)\n",
    "        If True, gaps will be fille with gap_code. If False, gaps will be\n",
    "        eliminated (ie, replaced with the major allele code)\n",
    "    gap_code: str or numeric, optional (default=np.nan)\n",
    "        encoding to use for gap characters if nogap=False\n",
    "    ref_allele_code: str or numeric, optional (default=0)\n",
    "        encoding to use for major alleles\n",
    "    non_ref_allele_code: str or numeric, optional (default=1)\n",
    "        encoding to use for minor alleles\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    np.array\n",
    "        Matrix of size N x L with reference/non-reference allele encoding\n",
    "        for all positions in L\n",
    "    \"\"\"\n",
    "\n",
    "    if encode_gaps:\n",
    "        gap_code=gap_code\n",
    "        print(\"encoding gaps with gap code\", gap_code)\n",
    "    else:\n",
    "        gap_code=ref_allele_code\n",
    "        print(\"encoding gaps with reference allele\", ref_allele_code)\n",
    "\n",
    "\n",
    "    encoded_matrix = np.zeros(shape=matrix.shape, dtype=object)\n",
    "\n",
    "    # For each column\n",
    "    for i in np.arange(0, encoded_matrix.shape[1]):\n",
    "\n",
    "        #grab that column\n",
    "        vec = matrix[:, i]\n",
    "\n",
    "        # create an encoded version of the column, with minor allele code as default\n",
    "        encoded_vec = np.array([non_ref_allele_code] * len(vec), dtype=object)\n",
    "\n",
    "        # If the row has a major allele, replace in encoded vec\n",
    "        is_ref_allele = np.equal(vec, reference_alleles[i])\n",
    "        encoded_vec[is_ref_allele] = ref_allele_code\n",
    "\n",
    "        # If the row has a gap character, replace in encoded vec\n",
    "        is_gap = np.isnan(vec.astype(float))\n",
    "        encoded_vec[is_gap] = gap_code\n",
    "\n",
    "        # save in the encoded matrix\n",
    "        encoded_matrix[:, i] = encoded_vec\n",
    "\n",
    "    return encoded_matrix\n",
    "\n",
    "def prepare_loci_file(matrix,\n",
    "    isolates, snps,  output_loci_file, gap_code=\"NA\", prefix=None):\n",
    "    \"\"\"\n",
    "    Perpares input file for GEMMA\n",
    "    Saves major and minor allele files if prefix is provided\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    matrix: np.array\n",
    "        array of genotypes, N isolates x N positions\n",
    "    isolates: iterable\n",
    "        list of isolate names\n",
    "    snps: iterable\n",
    "        list of position names\n",
    "    output_loci_file: str, optional (default=None)\n",
    "        path to save the output file to\n",
    "    gap_code: str, optional (default=NA)\n",
    "        code with which to encode gaps in the GEMMA input loci file\n",
    "    prefix: str, optional (default=None)\n",
    "        output prefix for major allele and strains file to save\n",
    "    \"\"\"\n",
    "\n",
    "    gm = GenotypeMatrix(\n",
    "        isolates, \n",
    "        snps,         \n",
    "        matrix_data = matrix\n",
    "    )\n",
    "    print(\"Made genotype matrix with shape\", gm.matrix.shape)\n",
    "    # write the input loci file\n",
    "    of = open(output_loci_file, \"w\")\n",
    "    print(\"writing Gemma with gap code\", gap_code)\n",
    "    gm.write_GEMMA(of, gap_code=gap_code)\n",
    "    \n",
    "    \n",
    "    if prefix:\n",
    "        # write the major allele file\n",
    "        major_allele_file = f\"{prefix}_major_alleles.csv\"\n",
    "        maj_alleles = pd.DataFrame(gm.major_alleles, index=gm.positions, columns=[\"major_alleles\"])\n",
    "        maj_alleles.loc[:,\"major_alleles\"] = [ALPHABET[x] for x in maj_alleles.major_alleles]\n",
    "        maj_alleles.to_csv(major_allele_file)\n",
    "\n",
    "        # write the strain file\n",
    "        strain_file= f\"{prefix}_strains.csv\"\n",
    "        strains = pd.DataFrame(gm.strains, columns=[\"strain\"])\n",
    "        strains.to_csv(strain_file, index=None)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "drugs = [\"AMIKACIN\",\"CAPREOMYCIN\",\"ETHAMBUTOL\",\"ETHIONAMIDE\",\"ISONIAZID\",\"KANAMYCIN\",\"MOXIFLOXACIN\",\"PYRAZINAMIDE\",\"RIFAMPICIN\",\"STREPTOMYCIN\"]\n",
    "\n",
    "drug_to_abbrev = {    \n",
    "    \"AMIKACIN\":\"AMI\",\n",
    "     \"CAPREOMYCIN\":\"CAP\",\n",
    "     \"ETHAMBUTOL\":\"EMB\",\n",
    "     \"ETHIONAMIDE\":\"ETA\",\n",
    "     \"ISONIAZID\":\"INH\",\n",
    "     \"KANAMYCIN\":\"KAN\",\n",
    "     \"MOXIFLOXACIN\":\"MOXI\",\n",
    "     \"PYRAZINAMIDE\":\"PZA\",\n",
    "     \"RIFAMPICIN\":\"RIF\",\n",
    "     \"STREPTOMYCIN\":\"STR\"\n",
    "}\n",
    "kwargs={\n",
    "    # \"isolate_annotation_file\":\"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_isolate_annotation.pkl\",\n",
    "    # \"genotype_array_file\":\"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_matrix.npy\",\n",
    "    # \"snp_annotation_file\":\"/n/data1/hms/dbmi/farhat/anna/genome_logreg/variable_selection_20_09/snp_annotation_file_filtered.csv\",\n",
    "    # \"phenotype_file\":\"/n/data1/hms/dbmi/farhat/anna/ABR_epistasis/prep_GEMMA_inputs/MIC_combined_data.csv\",\n",
    "    # \"prefix\":\"gemma_test\", \n",
    "    \"gemma\":\"/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static\",\n",
    "#    \"drug\":\"PYRAZINAMIDE\",\n",
    "#    \"drug_abbrev\": \"PZA\",\n",
    "    \"mtb_ancestor_fasta\": \"../input/MTB_ancestor_reference_modified.fasta\",\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 2: Running for all MAF 0p001 loci\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_pve(drug, **kwargs):\n",
    "    gemma_executable = kwargs[\"gemma\"]\n",
    "    pheno_file = f\"MIC_runs/output/{drug}/{drug}.phenotypes\"\n",
    "    relatedness_loci_file = f\"MIC_runs/output/{drug}/0.input_loci_files/MAF_0p001.loci\"\n",
    "    d_file = f\"MIC_runs/output/{drug}/MAF_0p001.eigenD.txt\"\n",
    "    u_file = f\"MIC_runs/output/{drug}/MAF_0p001.eigenU.txt\"\n",
    "    string = f\"{gemma_executable} -g {relatedness_loci_file} -p {pheno_file} -d {d_file} -u {u_file} -maf 0 -miss 0.2 -lmm -notsnp -o {outfile}\"\n",
    "    process = string.split(\" \")\n",
    "    result = subprocess.run(process, stdout=subprocess.PIPE)\n",
    "\n",
    "    stdout = result.stdout.decode('utf-8').split(\"\\n\")\n",
    "    if len(stdout) > 8:\n",
    "        out1 = stdout[8]\n",
    "        out2 = stdout[9]\n",
    "        assert 'pve estimate' in out1\n",
    "        assert 'se(pve)' in out2\n",
    "        pve = float(out1.split(\"=\")[-1])\n",
    "        se = float(out2.split(\"=\")[-1])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AMIKACIN\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/AMIKACIN/0.input_loci_files/MAF_0p001.loci -p output/AMIKACIN/AMIKACIN.phenotypes -k output/AMIKACIN/MAF_0p001.cXX.txt -maf 0 -miss 0.2 -lmm -notsnp -o heritability_temp.txt\n",
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'error! fail to open phenotype file: output/AMIKACIN/AMIKACIN.phenotypes', 'error! fail to open mean genotype file: output/AMIKACIN/0.input_loci_files/MAF_0p001.loci', 'ERROR: Enforce failed for file_kin output/AMIKACIN/MAF_0p001.cXX.txt: open file in src/param.cpp at line 905 in CheckParam', '']\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'pve' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [4]\u001b[0m, in \u001b[0;36m<cell line: 3>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     24\u001b[0m         se \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mfloat\u001b[39m(out2\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m=\u001b[39m\u001b[38;5;124m\"\u001b[39m)[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m])\n\u001b[1;32m     25\u001b[0m         table\u001b[38;5;241m.\u001b[39mappend([drug, pve, se])\n\u001b[0;32m---> 27\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[43mpve\u001b[49m, se)\n\u001b[1;32m     29\u001b[0m df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(table, columns\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdrug\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpve\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mse_pve\u001b[39m\u001b[38;5;124m\"\u001b[39m])\n\u001b[1;32m     30\u001b[0m df\u001b[38;5;241m.\u001b[39mto_csv(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPVE_all_MAF0p001.csv\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pve' is not defined"
     ]
    }
   ],
   "source": [
    "table = []\n",
    "#for drug in drugs:\n",
    "for drug in [\"AMIKACIN\"]:\n",
    "    print(drug)\n",
    "    \n",
    "    gemma_executable = kwargs[\"gemma\"]\n",
    "    pheno_file = f\"output/{drug}/{drug}.phenotypes\"\n",
    "    relatedness_loci_file = f\"output/{drug}/0.input_loci_files/MAF_0p001.loci\"\n",
    "    cxx_file = f\"output/{drug}/MAF_0p001.cXX.txt\"\n",
    "    outfile = \"heritability_temp.txt\"\n",
    "    string = f\"{gemma_executable} -g {relatedness_loci_file} -p {pheno_file} -k {cxx_file} -maf 0 -miss 0.2 -lmm -notsnp -o {outfile}\"\n",
    "    print(string)\n",
    "    process = string.split(\" \")\n",
    "    result = subprocess.run(process, stdout=subprocess.PIPE)\n",
    "\n",
    "    stdout = result.stdout.decode('utf-8').split(\"\\n\")\n",
    "    print(stdout)\n",
    "    if len(stdout) > 8:\n",
    "        out1 = stdout[9]\n",
    "        out2 = stdout[10]\n",
    "        assert 'pve estimate' in out1\n",
    "        assert 'se(pve)' in out2\n",
    "        pve = float(out1.split(\"=\")[-1])\n",
    "        se = float(out2.split(\"=\")[-1])\n",
    "        table.append([drug, pve, se])\n",
    "    \n",
    "    print(pve, se)\n",
    "    \n",
    "df = pd.DataFrame(table, columns=[\"drug\", \"pve\", \"se_pve\"])\n",
    "df.to_csv(\"PVE_all_MAF0p001.csv\")   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 2.5: Running for all homoplasic loci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# full_range = pd.read_csv(\"../output/05.single_mutations/snps_with_all_annotation.csv\")\n",
    "# full_range = list(full_range.pos)\n",
    "# # reads in the SNP annotations\n",
    "# snp_annotation = pd.read_pickle(\"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_SNP_annotation.pkl\")\n",
    "# snp_annotation['mat_index'] =snp_annotation.index\n",
    "# print('annotation for N Snps:', len(snp_annotation))\n",
    "\n",
    "# print(\"total number of SNPs\", len(snp_annotation.query(\"pos in @full_range\")))\n",
    "# snps_subset = snp_annotation.query(\"pos in @full_range\")   \n",
    "\n",
    "# isolates = pd.read_pickle(kwargs['isolate_annotation_file'])\n",
    "\n",
    "# for drug, drug_abbrev in drug_to_abbrev.items():\n",
    "# #for drug, drug_abbrev in drug_to_a:\n",
    "#     print(drug, drug_abbrev)\n",
    "#     # Generate the cXX file\n",
    "    \n",
    "#     prefix = f\"output/{drug}\"\n",
    "#     output_prefix = f\"{drug}\"\n",
    "#     os.system(f\"mkdir {prefix}\")\n",
    "#     os.system(f\"mkdir {prefix}/0.input_loci_files\")\n",
    "#     os.system(f\"mkdir {prefix}/1.job_submission_files\")\n",
    "#     os.system(f\"mkdir {prefix}/2.run_output\")\n",
    "\n",
    "#     # get the isolates with phenotypes to keep\n",
    "#     phenotypes = pd.read_csv(kwargs['phenotype_file'], index_col=0)\n",
    "#     phenotypes_found_all = phenotypes.loc[~phenotypes[drug_abbrev].isnull(),:]\n",
    "#     phenotypes_found = phenotypes_found_all[[\"ROLLINGDB_ID\", f\"{drug_abbrev}_midpoint\"]]\n",
    "#     phenotypes_found[drug] = np.log(phenotypes_found[f\"{drug_abbrev}_midpoint\"])\n",
    "    \n",
    "#     isolates = pd.read_pickle(kwargs['isolate_annotation_file'])\n",
    "#     isolates_with_phenotype = isolates.query(\"isolate_ID in @phenotypes_found.ROLLINGDB_ID\")\n",
    "#     g = np.load(\"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_matrix.npy\")\n",
    "    \n",
    "#     #### Prepare the phenotype input file ####\n",
    "#     pheno_file = f\"{prefix}/{drug}.phenotypes\"\n",
    "#     write_phenotype_file_MIC(pheno_file, phenotypes_found, isolates_with_phenotype, drug)\n",
    "\n",
    "#     # slice the matrix for only isolates of interest\n",
    "#     geno_mat = g\n",
    "#     matrix = geno_mat[snps_subset.index, :][:, isolates_with_phenotype.index]\n",
    "#     matrix = translate_genotypes(matrix, genotype_translator).T\n",
    "#     print(\"Number of 9's in the matrix\", np.sum(matrix=='-'))\n",
    "\n",
    "    \n",
    "#     # add the ancestral sequence\n",
    "#     matrix, isolates = add_ancestor(kwargs[\"mtb_ancestor_fasta\"], snps_subset, matrix, isolates_with_phenotype)\n",
    "    \n",
    "#     # Generate the locus file\n",
    "#     relatedness_loci_file = f\"{prefix}/0.input_loci_files/homoplasic_sites.loci\"\n",
    "#     print(f\"Creating locus file for {drug} with N positions: {len(snps_subset.pos)}\")\n",
    "#     prepare_loci_file(matrix, isolates, list(snps_subset.pos), relatedness_loci_file, prefix=prefix)\n",
    "\n",
    "#     cxx_file = f\"{prefix}/homoplasic_sites.cXX.txt\"\n",
    "\n",
    "#     outfile = f\"{output_prefix}/homoplasic_sites\"\n",
    "#     string1 = f\"{kwargs['gemma']} -g {relatedness_loci_file} -p {pheno_file} -gk -maf 0 -o {outfile}\"\n",
    "#     print(string1)\n",
    "\n",
    "#     # Run the eigen decomposition\n",
    "#     outfile =f'{output_prefix}/homoplasic_sites'\n",
    "#     string2 = f\"{kwargs['gemma']} -g {relatedness_loci_file} -p {pheno_file} -k {cxx_file} -maf 0 -eigen -notsnp -o {outfile}\"\n",
    "#     print(string2)\n",
    "\n",
    "#     print(\"creating GRM\")\n",
    "#     os.system(string1)\n",
    "\n",
    "#     print(\"running eigendecomp\")\n",
    "#     os.system(string2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# table = []\n",
    "# for drug in drugs:\n",
    "#     print(drug)\n",
    "    \n",
    "#     gemma_executable = kwargs[\"gemma\"]\n",
    "#     pheno_file = f\"output/{drug}/{drug}.phenotypes\"\n",
    "#     relatedness_loci_file = f\"output/{drug}/0.input_loci_files/homoplasic_sites.loci\"\n",
    "#     cxx_file = f\"output/{drug}/homoplasic_sites.cXX.txt\"\n",
    "#     outfile = \"heritability_temp.txt\"\n",
    "    \n",
    "#     string = f\"{gemma_executable} -g {relatedness_loci_file} -p {pheno_file} -k {cxx_file} -maf 0 -miss 0.2 -lmm -notsnp -o {outfile}\"\n",
    "#     print(string)\n",
    "    \n",
    "#     process = string.split(\" \")\n",
    "#     result = subprocess.run(process, stdout=subprocess.PIPE)\n",
    "\n",
    "#     stdout = result.stdout.decode('utf-8').split(\"\\n\")\n",
    "#     print(stdout)\n",
    "#     if len(stdout) > 8:\n",
    "#         out3 = stdout[7]\n",
    "#         out1 = stdout[9]\n",
    "#         out2 = stdout[10]\n",
    "#         assert 'pve estimate' in out1\n",
    "#         assert 'se(pve)' in out2\n",
    "#         assert 'number of analyzed SNPs' in out3\n",
    "#         pve = float(out1.split(\"=\")[-1])\n",
    "#         se = float(out2.split(\"=\")[-1])\n",
    "#         N_snps = float(out3.split(\"=\")[-1])\n",
    "#         table.append([drug, pve, se, N_snps])\n",
    "    \n",
    "#     print(pve, se)\n",
    "    \n",
    "# df = pd.DataFrame(table, columns=[\"drug\", \"pve\", \"se_pve\", \"N_snps\"])\n",
    "# df.to_csv(\"PVE_homoplasic_sites..csv\")   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 3: Running for all loci in antibiotic-associated genes\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drug_to_abbrev.items()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# full_range = pd.read_csv(\"../output/05.single_mutations/snps_with_all_annotation.csv\", index_col=0)\n",
    "# full_range = full_range.query(\"in_known_ABR_position\")\n",
    "# full_range[\"known_drug\"] = [x.split(\",\") for x in full_range.known_drug]\n",
    "\n",
    "# # reads in the SNP annotations\n",
    "# snp_annotation = pd.read_pickle(\"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_SNP_annotation.pkl\")\n",
    "# snp_annotation['mat_index'] =snp_annotation.index\n",
    "# print('annotation for N Snps:', len(snp_annotation))\n",
    "\n",
    "# print(\"total number of SNPs\", len(snp_annotation.query(\"pos in @full_range.pos\")))\n",
    " \n",
    "# isolates = pd.read_pickle(kwargs['isolate_annotation_file'])\n",
    "\n",
    "# for drug, drug_abbrev in drug_to_abbrev.items():\n",
    "#     if drug == \"LEVOFLOXACIN\":\n",
    "#         continue\n",
    "#     print(drug, drug_abbrev)\n",
    "#     relevant_snps = full_range.query(\"@drug_abbrev in known_drug\")\n",
    "    \n",
    "#     _keep = []\n",
    "#     for idx, row in full_range.iterrows():\n",
    "#         if drug_abbrev in row.known_drug:\n",
    "#             _keep.append(idx)\n",
    "            \n",
    "#         elif drug_abbrev=='ETA':\n",
    "#             if \"ETH\" in row.known_drug:\n",
    "#                 _keep.append(idx)\n",
    "#         elif drug_abbrev=='MOXI':\n",
    "#             if \"MXF\" in row.known_drug:\n",
    "#                 _keep.append(idx)\n",
    "#         elif drug_abbrev=='STR':\n",
    "#             if \"STM\" in row.known_drug:\n",
    "#                 _keep.append(idx)\n",
    "                \n",
    "#     positions_found = list(set(full_range.loc[_keep].pos))\n",
    "#     snps_subset = snp_annotation.query(\"pos in @positions_found\")\n",
    "    \n",
    "#     print(f\"found {len(positions_found)} resistance snps\")\n",
    "#     #break\n",
    "#     # Generate the cXX file\n",
    "\n",
    "#     prefix = f\"output/{drug}\"\n",
    "#     output_prefix = f\"{drug}\"\n",
    "#     os.system(f\"mkdir {prefix}\")\n",
    "#     os.system(f\"mkdir {prefix}/0.input_loci_files\")\n",
    "#     os.system(f\"mkdir {prefix}/1.job_submission_files\")\n",
    "#     os.system(f\"mkdir {prefix}/2.run_output\")\n",
    "\n",
    "#     # get the isolates with phenotypes to keep\n",
    "#     phenotypes = pd.read_csv(kwargs['phenotype_file'], index_col=0)\n",
    "#     phenotypes_found_all = phenotypes.loc[~phenotypes[drug_abbrev].isnull(),:]\n",
    "#     phenotypes_found = phenotypes_found_all[[\"ROLLINGDB_ID\", f\"{drug_abbrev}_midpoint\"]]\n",
    "#     phenotypes_found[drug] = np.log(phenotypes_found[f\"{drug_abbrev}_midpoint\"])\n",
    "    \n",
    "#     isolates = pd.read_pickle(kwargs['isolate_annotation_file'])\n",
    "#     isolates_with_phenotype = isolates.query(\"isolate_ID in @phenotypes_found.ROLLINGDB_ID\")\n",
    "#     g = np.load(\"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_matrix.npy\")\n",
    "    \n",
    "#     #### Prepare the phenotype input file ####\n",
    "#     pheno_file = f\"{prefix}/{drug}.phenotypes\"\n",
    "#     write_phenotype_file_MIC(pheno_file, phenotypes_found, isolates_with_phenotype, drug)\n",
    "\n",
    "#     # slice the matrix for only isolates of interest\n",
    "#     geno_mat = g\n",
    "#     matrix = geno_mat[snps_subset.index, :][:, isolates_with_phenotype.index]\n",
    "#     matrix = translate_genotypes(matrix, genotype_translator).T\n",
    "#     print(\"Number of 9's in the matrix\", np.sum(matrix=='-'))\n",
    "\n",
    "    \n",
    "#     # add the ancestral sequence\n",
    "#     matrix, isolates = add_ancestor(kwargs[\"mtb_ancestor_fasta\"], snps_subset, matrix, isolates_with_phenotype)\n",
    "    \n",
    "#     # Generate the locus file\n",
    "#     relatedness_loci_file = f\"{prefix}/0.input_loci_files/ABR_sites.loci\"\n",
    "#     print(f\"Creating locus file for {drug} with N positions: {len(snps_subset.pos)}\")\n",
    "#     prepare_loci_file(matrix, isolates, list(snps_subset.pos), relatedness_loci_file, prefix=prefix)\n",
    "\n",
    "#     cxx_file = f\"{prefix}/ABR_known.cXX.txt\"\n",
    "\n",
    "#     outfile = f\"{output_prefix}/ABR_known\"\n",
    "#     string1 = f\"{kwargs['gemma']} -g {relatedness_loci_file} -p {pheno_file} -gk -maf 0 -o {outfile}\"\n",
    "#     print(string1)\n",
    "\n",
    "#     # Run the eigen decomposition\n",
    "#     outfile =f'{output_prefix}/ABR_known'\n",
    "#     string2 = f\"{kwargs['gemma']} -g {relatedness_loci_file} -p {pheno_file} -k {cxx_file} -maf 0 -eigen -notsnp -o {outfile}\"\n",
    "#     print(string2)\n",
    "\n",
    "#     print(\"creating GRM\")\n",
    "#     os.system(string1)\n",
    "\n",
    "#     print(\"running eigendecomp\")\n",
    "#     os.system(string2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# table = []\n",
    "# for drug in drugs:\n",
    "# #for drug in [\"RIFAMPICIN\"]:\n",
    "#     print(drug)\n",
    "    \n",
    "#     gemma_executable = kwargs[\"gemma\"]\n",
    "#     pheno_file = f\"output/{drug}/{drug}.phenotypes\"\n",
    "#     relatedness_loci_file = f\"output/{drug}/0.input_loci_files/ABR_sites.loci\"\n",
    "#     cxx_file = f\"output/{drug}/ABR_known.cXX.txt\"\n",
    "#     outfile = \"heritability_temp.txt\"\n",
    "    \n",
    "#     string = f\"{gemma_executable} -g {relatedness_loci_file} -p {pheno_file} -k {cxx_file} -maf 0 -miss 0.2 -lmm -notsnp -o {outfile}\"\n",
    "#     print(string)\n",
    "    \n",
    "#     process = string.split(\" \")\n",
    "#     result = subprocess.run(process, stdout=subprocess.PIPE)\n",
    "\n",
    "#     stdout = result.stdout.decode('utf-8').split(\"\\n\")\n",
    "#     print(stdout)\n",
    "#     if len(stdout) > 8:\n",
    "#         out3 = stdout[7]\n",
    "#         out1 = stdout[9]\n",
    "#         out2 = stdout[10]\n",
    "#         assert 'pve estimate' in out1\n",
    "#         assert 'se(pve)' in out2\n",
    "#         assert 'number of analyzed SNPs' in out3\n",
    "#         pve = float(out1.split(\"=\")[-1])\n",
    "#         se = float(out2.split(\"=\")[-1])\n",
    "#         N_snps = float(out3.split(\"=\")[-1])\n",
    "#         table.append([drug, pve, se, N_snps])\n",
    "    \n",
    "#     print(pve, se)\n",
    "    \n",
    "# df = pd.DataFrame(table, columns=[\"drug\", \"pve\", \"se_pve\", \"N_snps\"])\n",
    "# df.to_csv(\"PVE_ABR_known.csv\")   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.5: Possible ABR sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# full_range = pd.read_csv(\"../output/05.single_mutations/snps_with_all_annotation.csv\")\n",
    "# full_range = full_range.query(\"in_possible_ABR_gene\")\n",
    "# full_range[\"possible_drug\"] = [x.split(\",\") for x in full_range.possible_drug]\n",
    "\n",
    "# # reads in the SNP annotations\n",
    "# snp_annotation = pd.read_pickle(\"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_SNP_annotation.pkl\")\n",
    "# snp_annotation['mat_index'] =snp_annotation.index\n",
    "# print('annotation for N Snps:', len(snp_annotation))\n",
    "\n",
    "# print(\"total number of SNPs\", len(snp_annotation.query(\"pos in @full_range\")))\n",
    "# snps_subset = snp_annotation.query(\"pos in @full_range\")   \n",
    "\n",
    "# isolates = pd.read_pickle(kwargs['isolate_annotation_file'])\n",
    "\n",
    "# for drug, drug_abbrev in drug_to_abbrev.items():\n",
    "# #for drug, drug_abbrev in [(\"RIFAMPICIN\", \"RIF\")]:\n",
    "#     print(drug, drug_abbrev)\n",
    "#     # Generate the cXX file\n",
    "    \n",
    "#     if drug == \"LEVOFLOXACIN\":\n",
    "#         continue\n",
    "    \n",
    "#     _keep = []\n",
    "#     for idx, row in full_range.iterrows():\n",
    "#         if drug_abbrev in row.possible_drug:\n",
    "#             _keep.append(idx)\n",
    "#         elif drug_abbrev=='ETA':\n",
    "#             if \"ETH\" in row.possible_drug:\n",
    "#                 _keep.append(idx)\n",
    "#         elif drug_abbrev=='MOXI':\n",
    "#             if \"MXF\" in row.possible_drug:\n",
    "#                 _keep.append(idx)\n",
    "#         elif drug_abbrev=='STR':\n",
    "#             if \"STM\" in row.possible_drug:\n",
    "#                 _keep.append(idx)\n",
    "#     positions_found = list(set(full_range.loc[_keep].pos))\n",
    "#     snps_subset = snp_annotation.query(\"pos in @positions_found\")\n",
    "    \n",
    "#     print(f\"found {len(positions_found)} resistance snps\")\n",
    "#     prefix = f\"output/{drug}\"\n",
    "#     output_prefix = f\"{drug}\"\n",
    "#     os.system(f\"mkdir {prefix}\")\n",
    "#     os.system(f\"mkdir {prefix}/0.input_loci_files\")\n",
    "#     os.system(f\"mkdir {prefix}/1.job_submission_files\")\n",
    "#     os.system(f\"mkdir {prefix}/2.run_output\")\n",
    "\n",
    "#     # get the isolates with phenotypes to keep\n",
    "#     phenotypes = pd.read_csv(kwargs['phenotype_file'], index_col=0)\n",
    "#     phenotypes_found_all = phenotypes.loc[~phenotypes[drug_abbrev].isnull(),:]\n",
    "#     phenotypes_found = phenotypes_found_all[[\"ROLLINGDB_ID\", f\"{drug_abbrev}_midpoint\"]]\n",
    "#     phenotypes_found[drug] = np.log(phenotypes_found[f\"{drug_abbrev}_midpoint\"])\n",
    "    \n",
    "#     isolates = pd.read_pickle(kwargs['isolate_annotation_file'])\n",
    "#     isolates_with_phenotype = isolates.query(\"isolate_ID in @phenotypes_found.ROLLINGDB_ID\")\n",
    "#     g = np.load(\"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_matrix.npy\")\n",
    "    \n",
    "#     #### Prepare the phenotype input file ####\n",
    "#     pheno_file = f\"{prefix}/{drug}.phenotypes\"\n",
    "#     write_phenotype_file_MIC(pheno_file, phenotypes_found, isolates_with_phenotype, drug)\n",
    "\n",
    "#     # slice the matrix for only isolates of interest\n",
    "#     geno_mat = g\n",
    "#     matrix = geno_mat[snps_subset.index, :][:, isolates_with_phenotype.index]\n",
    "#     matrix = translate_genotypes(matrix, genotype_translator).T\n",
    "#     print(\"Number of 9's in the matrix\", np.sum(matrix=='-'))\n",
    "\n",
    "    \n",
    "#     # add the ancestral sequence\n",
    "#     matrix, isolates = add_ancestor(kwargs[\"mtb_ancestor_fasta\"], snps_subset, matrix, isolates_with_phenotype)\n",
    "    \n",
    "#     # Generate the locus file\n",
    "#     relatedness_loci_file = f\"{prefix}/0.input_loci_files/ABR_possible.loci\"\n",
    "#     print(f\"Creating locus file for {drug} with N positions: {len(snps_subset.pos)}\")\n",
    "#     prepare_loci_file(matrix, isolates, list(snps_subset.pos), relatedness_loci_file, prefix=prefix)\n",
    "\n",
    "#     cxx_file = f\"{prefix}/ABR_possible.cXX.txt\"\n",
    "\n",
    "#     outfile = f\"{output_prefix}/ABR_possible\"\n",
    "#     string1 = f\"{kwargs['gemma']} -g {relatedness_loci_file} -p {pheno_file} -gk -maf 0 -o {outfile}\"\n",
    "#     print(string1)\n",
    "\n",
    "#     # Run the eigen decomposition\n",
    "#     outfile =f'{output_prefix}/ABR_possible'\n",
    "#     string2 = f\"{kwargs['gemma']} -g {relatedness_loci_file} -p {pheno_file} -k {cxx_file} -maf 0 -eigen -notsnp -o {outfile}\"\n",
    "#     print(string2)\n",
    "\n",
    "#     print(\"creating GRM\")\n",
    "#     os.system(string1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# table = []\n",
    "# for drug in drugs:\n",
    "#     print(drug)\n",
    "    \n",
    "#     gemma_executable = kwargs[\"gemma\"]\n",
    "#     pheno_file = f\"output/{drug}/{drug}.phenotypes\"\n",
    "#     relatedness_loci_file = f\"output/{drug}/0.input_loci_files/ABR_possible.loci\"\n",
    "#     cxx_file = f\"output/{drug}/ABR_possible.cXX.txt\"\n",
    "#     outfile = \"heritability_temp.txt\"\n",
    "    \n",
    "#     string = f\"{gemma_executable} -g {relatedness_loci_file} -p {pheno_file} -k {cxx_file} -maf 0 -miss 0.2 -lmm -notsnp -o {outfile}\"\n",
    "#     print(string)\n",
    "    \n",
    "#     process = string.split(\" \")\n",
    "#     result = subprocess.run(process, stdout=subprocess.PIPE)\n",
    "\n",
    "#     stdout = result.stdout.decode('utf-8').split(\"\\n\")\n",
    "#     print(stdout)\n",
    "#     if len(stdout) > 8:\n",
    "#         out3 = stdout[7]\n",
    "#         out1 = stdout[9]\n",
    "#         out2 = stdout[10]\n",
    "#         assert 'pve estimate' in out1\n",
    "#         assert 'se(pve)' in out2\n",
    "#         assert 'number of analyzed SNPs' in out3\n",
    "#         pve = float(out1.split(\"=\")[-1])\n",
    "#         se = float(out2.split(\"=\")[-1])\n",
    "#         N_snps = float(out3.split(\"=\")[-1])\n",
    "#         table.append([drug, pve, se, N_snps])\n",
    "    \n",
    "#     print(pve, se)\n",
    "    \n",
    "# df = pd.DataFrame(table, columns=[\"drug\", \"pve\", \"se_pve\", \"N_snps\"])\n",
    "# df.to_csv(\"PVE_ABR_possible.csv\")   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run for SNPs plus other sites that are epistatic (single sites only, subsequent and simultaneous hits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# full_range = pd.read_csv(\"../output/05.single_mutations/snps_with_all_annotation.csv\", index_col=0)\n",
    "# full_range = full_range.query(\"in_known_ABR_position\")\n",
    "# full_range[\"known_drug\"] = [x.split(\",\") for x in full_range.known_drug]\n",
    "\n",
    "# # reads in the SNP annotations\n",
    "# snp_annotation = pd.read_pickle(\"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_SNP_annotation.pkl\")\n",
    "# snp_annotation['mat_index'] =snp_annotation.index\n",
    "# print('annotation for N Snps:', len(snp_annotation))\n",
    "\n",
    "\n",
    "# isolates = pd.read_pickle(kwargs['isolate_annotation_file'])\n",
    "\n",
    "# for drug, drug_abbrev in drug_to_abbrev.items():\n",
    "#     print(drug, drug_abbrev)\n",
    "#     # Generate the cXX file\n",
    "#     _keep = []\n",
    "#     for idx, row in full_range.iterrows():\n",
    "#         if drug_abbrev in row.known_drug:\n",
    "#             _keep.append(idx)\n",
    "#         elif drug_abbrev=='ETA':\n",
    "#             if \"ETH\" in row.known_drug:\n",
    "#                 _keep.append(idx)\n",
    "#         elif drug_abbrev=='MOXI':\n",
    "#             if \"MXF\" in row.known_drug:\n",
    "#                 _keep.append(idx)\n",
    "#         elif drug_abbrev=='STR':\n",
    "#             if \"STM\" in row.known_drug:\n",
    "#                 _keep.append(idx)\n",
    "                \n",
    "#     positions_found = list(set(full_range.loc[_keep].pos))\n",
    "    \n",
    "#     antibiotic_hits = pd.read_csv(\n",
    "#         f\"/n/data1/hms/dbmi/farhat/anna/ABR_epistasis/DependentMutations2/output/04.antibiotic/{drug_abbrev}_antibiotic_hits.csv\"\n",
    "#     )\n",
    "#     positions_found = positions_found + list(antibiotic_hits.position_j) + list(antibiotic_hits.position_i)\n",
    "    \n",
    "#     snps_subset = snp_annotation.query(\"pos in @positions_found\")\n",
    "    \n",
    "#     prefix = f\"output/{drug}\"\n",
    "#     output_prefix = f\"{drug}\"\n",
    "#     os.system(f\"mkdir {prefix}\")\n",
    "#     os.system(f\"mkdir {prefix}/0.input_loci_files\")\n",
    "#     os.system(f\"mkdir {prefix}/1.job_submission_files\")\n",
    "#     os.system(f\"mkdir {prefix}/2.run_output\")\n",
    "\n",
    "#     # get the isolates with phenotypes to keep\n",
    "#     phenotypes = pd.read_csv(kwargs['phenotype_file'], index_col=0)\n",
    "#     phenotypes.replace(\"NA\", np.nan, inplace=True)\n",
    "    \n",
    "#     phenotypes_found_all = phenotypes.loc[~phenotypes[drug_abbrev].isnull(),:]\n",
    "#     phenotypes_found = phenotypes_found_all[[\"ROLLINGDB_ID\", f\"{drug_abbrev}_midpoint\"]]\n",
    "#     phenotypes_found[drug] = np.log(phenotypes_found[f\"{drug_abbrev}_midpoint\"])\n",
    "\n",
    "#     isolates = pd.read_pickle(kwargs['isolate_annotation_file'])\n",
    "#     isolates_with_phenotype = isolates.query(\"isolate_ID in @phenotypes_found.ROLLINGDB_ID\")\n",
    "#     g = np.load(\"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_matrix.npy\")\n",
    "\n",
    "# #     #### Prepare the phenotype input file ####\n",
    "#     pheno_file = f\"{prefix}/{drug}.phenotypes\"\n",
    "#     write_phenotype_file_MIC(pheno_file, phenotypes_found, isolates_with_phenotype, drug)\n",
    "\n",
    "#     # slice the matrix for only isolates of interest\n",
    "#     geno_mat = g\n",
    "#     matrix = geno_mat[snps_subset.index, :][:, isolates_with_phenotype.index]\n",
    "#     matrix = translate_genotypes(matrix, genotype_translator).T    \n",
    "\n",
    "#     # add the ancestral sequence\n",
    "#     matrix, isolates = add_ancestor(kwargs[\"mtb_ancestor_fasta\"], snps_subset, matrix, isolates_with_phenotype)\n",
    "\n",
    "# #     # Generate the locus file\n",
    "    \n",
    "#     relatedness_loci_file = f\"{prefix}/0.input_loci_files/ABR_sites_plus_UK.loci\"\n",
    "\n",
    "#     print(f\"Creating locus file for {drug} with N positions: {len(snps_subset.pos)}\")\n",
    "\n",
    "#     prepare_loci_file(matrix, isolates, list(snps_subset.pos), relatedness_loci_file, prefix=prefix)\n",
    "\n",
    "#     cxx_file = f\"{prefix}/heritability_test_ABR_UK.cXX.txt\"\n",
    "\n",
    "#     outfile = f\"{output_prefix}/heritability_test_ABR_UK\"\n",
    "#     string1 = f\"{kwargs['gemma']} -g {relatedness_loci_file} -p {pheno_file} -gk -maf 0 -o {outfile}\"\n",
    "#     print(string1)\n",
    "\n",
    "#     # Run the eigen decomposition\n",
    "#     outfile =f'{output_prefix}/heritability_test_ABR_UK'\n",
    "#     string2 = f\"{kwargs['gemma']} -g {relatedness_loci_file} -p {pheno_file} -k {cxx_file} -maf 0 -eigen -notsnp -o {outfile}\"\n",
    "#     print(string2)\n",
    "\n",
    "#     print(\"creating GRM\")\n",
    "#     os.system(string1)\n",
    "\n",
    "#     print(\"running eigendecomp\")\n",
    "#     os.system(string2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# table = []\n",
    "# for drug in drugs:\n",
    "#     print(drug)\n",
    "    \n",
    "#     gemma_executable = kwargs[\"gemma\"]\n",
    "#     pheno_file = f\"output/{drug}/{drug}.phenotypes\"\n",
    "#     relatedness_loci_file = f\"output/{drug}/0.input_loci_files/ABR_sites_plus_UK.loci\"\n",
    "#     d_file = f\"output/{drug}/heritability_test_ABR_UK.eigenD.txt\"\n",
    "#     u_file = f\"output/{drug}/heritability_test_ABR_UK.eigenU.txt\"\n",
    "#     outfile = \"heritability_temp.txt\"\n",
    "#     string = f\"{gemma_executable} -g {relatedness_loci_file} -p {pheno_file} -d {d_file} -u {u_file} -maf 0 -miss 0.2 -lmm -notsnp -o {outfile}\"\n",
    "#     print(string)\n",
    "#     process = string.split(\" \")\n",
    "#     result = subprocess.run(process, stdout=subprocess.PIPE)\n",
    "\n",
    "#     stdout = result.stdout.decode('utf-8').split(\"\\n\")\n",
    "\n",
    "#     print(stdout)\n",
    "#     if len(stdout) > 8:\n",
    "#         out3 = stdout[7]\n",
    "#         out1 = stdout[8]\n",
    "#         out2 = stdout[9]\n",
    "#         assert 'pve estimate' in out1\n",
    "#         assert 'se(pve)' in out2\n",
    "#         assert 'number of analyzed SNPs' in out3\n",
    "#         pve = float(out1.split(\"=\")[-1])\n",
    "#         se = float(out2.split(\"=\")[-1])\n",
    "#         N_snps = float(out3.split(\"=\")[-1])\n",
    "#         table.append([drug, pve, se, N_snps])\n",
    "    \n",
    "#     print(pve, se)\n",
    "    \n",
    "# df = pd.DataFrame(table, columns=[\"drug\", \"pve\", \"se_pve\", \"N_snps\"])\n",
    "# df.to_csv(\"PVE_ABR_UK_sites.csv\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run for POSSIBLE ANTIBIOTIC SNPs plus other sites that are epistatic (single sites only, subsequent and simultaneous hits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# full_range = pd.read_csv(\"../output/05.single_mutations/snps_with_all_annotation.csv\", index_col=0)\n",
    "# full_range = full_range.query(\"in_possible_ABR_gene\")\n",
    "# full_range[\"possible_drug\"] = [x.split(\",\") for x in full_range.possible_drug]\n",
    "\n",
    "# # reads in the SNP annotations\n",
    "# snp_annotation = pd.read_pickle(\"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_SNP_annotation.pkl\")\n",
    "# snp_annotation['mat_index'] =snp_annotation.index\n",
    "# print('annotation for N Snps:', len(snp_annotation))\n",
    "\n",
    "\n",
    "# isolates = pd.read_pickle(kwargs['isolate_annotation_file'])\n",
    "\n",
    "# for drug, drug_abbrev in drug_to_abbrev.items():\n",
    "#     print(drug, drug_abbrev)\n",
    "#     # Generate the cXX file\n",
    "#     _keep = []\n",
    "#     for idx, row in full_range.iterrows():\n",
    "#         if drug_abbrev in row.possible_drug:\n",
    "#             _keep.append(idx)\n",
    "#         elif drug_abbrev=='ETA':\n",
    "#             if \"ETH\" in row.possible_drug:\n",
    "#                 _keep.append(idx)\n",
    "#         elif drug_abbrev=='MOXI':\n",
    "#             if \"MXF\" in row.possible_drug:\n",
    "#                 _keep.append(idx)\n",
    "#         elif drug_abbrev=='STR':\n",
    "#             if \"STM\" in row.possible_drug:\n",
    "#                 _keep.append(idx)\n",
    "                \n",
    "#     positions_found = list(set(full_range.loc[_keep].pos))\n",
    "    \n",
    "#     antibiotic_hits = pd.read_csv(\n",
    "#         f\"/n/data1/hms/dbmi/farhat/anna/ABR_epistasis/DependentMutations2/output/04.antibiotic/{drug_abbrev}_antibiotic_hits.csv\"\n",
    "#     )\n",
    "#     positions_found = positions_found + list(antibiotic_hits.position_j) + list(antibiotic_hits.position_i)\n",
    "    \n",
    "#     snps_subset = snp_annotation.query(\"pos in @positions_found\")\n",
    "    \n",
    "#     prefix = f\"output/{drug}\"\n",
    "#     output_prefix = f\"{drug}\"\n",
    "#     os.system(f\"mkdir {prefix}\")\n",
    "#     os.system(f\"mkdir {prefix}/0.input_loci_files\")\n",
    "#     os.system(f\"mkdir {prefix}/1.job_submission_files\")\n",
    "#     os.system(f\"mkdir {prefix}/2.run_output\")\n",
    "\n",
    "#     # get the isolates with phenotypes to keep\n",
    "#     phenotypes = pd.read_csv(kwargs['phenotype_file'], index_col=0)\n",
    "#     phenotypes.replace(\"NA\", np.nan, inplace=True)\n",
    "    \n",
    "#     phenotypes_found_all = phenotypes.loc[~phenotypes[drug_abbrev].isnull(),:]\n",
    "#     phenotypes_found = phenotypes_found_all[[\"ROLLINGDB_ID\", f\"{drug_abbrev}_midpoint\"]]\n",
    "#     phenotypes_found[drug] = np.log(phenotypes_found[f\"{drug_abbrev}_midpoint\"])\n",
    "\n",
    "#     isolates = pd.read_pickle(kwargs['isolate_annotation_file'])\n",
    "#     isolates_with_phenotype = isolates.query(\"isolate_ID in @phenotypes_found.ROLLINGDB_ID\")\n",
    "#     g = np.load(\"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_matrix.npy\")\n",
    "\n",
    "# #     #### Prepare the phenotype input file ####\n",
    "#     pheno_file = f\"{prefix}/{drug}.phenotypes\"\n",
    "#     write_phenotype_file_MIC(pheno_file, phenotypes_found, isolates_with_phenotype, drug)\n",
    "\n",
    "#     # slice the matrix for only isolates of interest\n",
    "#     geno_mat = g\n",
    "#     matrix = geno_mat[snps_subset.index, :][:, isolates_with_phenotype.index]\n",
    "#     matrix = translate_genotypes(matrix, genotype_translator).T    \n",
    "\n",
    "#     # add the ancestral sequence\n",
    "#     matrix, isolates = add_ancestor(kwargs[\"mtb_ancestor_fasta\"], snps_subset, matrix, isolates_with_phenotype)\n",
    "\n",
    "# #     # Generate the locus file\n",
    "    \n",
    "#     relatedness_loci_file = f\"{prefix}/0.input_loci_files/possible_ABR_sites_plus_UK.loci\"\n",
    "\n",
    "#     print(f\"Creating locus file for {drug} with N positions: {len(snps_subset.pos)}\")\n",
    "\n",
    "#     prepare_loci_file(matrix, isolates, list(snps_subset.pos), relatedness_loci_file, prefix=prefix)\n",
    "\n",
    "#     cxx_file = f\"{prefix}/heritability_test_possible_ABR_UK.cXX.txt\"\n",
    "\n",
    "#     outfile = f\"{output_prefix}/heritability_test_possible_ABR_UK\"\n",
    "#     string1 = f\"{kwargs['gemma']} -g {relatedness_loci_file} -p {pheno_file} -gk -maf 0 -o {outfile}\"\n",
    "#     print(string1)\n",
    "\n",
    "#     # Run the eigen decomposition\n",
    "#     outfile =f'{output_prefix}/heritability_test_possible_ABR_UK'\n",
    "#     string2 = f\"{kwargs['gemma']} -g {relatedness_loci_file} -p {pheno_file} -k {cxx_file} -maf 0 -eigen -notsnp -o {outfile}\"\n",
    "#     print(string2)\n",
    "\n",
    "#     print(\"creating GRM\")\n",
    "#     os.system(string1)\n",
    "\n",
    "#     print(\"running eigendecomp\")\n",
    "#     os.system(string2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# table = []\n",
    "# for drug in drugs:\n",
    "#     print(drug)\n",
    "    \n",
    "#     gemma_executable = kwargs[\"gemma\"]\n",
    "#     pheno_file = f\"output/{drug}/{drug}.phenotypes\"\n",
    "#     relatedness_loci_file = f\"output/{drug}/0.input_loci_files/possible_ABR_sites_plus_UK.loci\"\n",
    "#     d_file = f\"output/{drug}/heritability_test_possible_ABR_UK.eigenD.txt\"\n",
    "#     u_file = f\"output/{drug}/heritability_test_possible_ABR_UK.eigenU.txt\"\n",
    "#     outfile = \"heritability_temp.txt\"\n",
    "#     string = f\"{gemma_executable} -g {relatedness_loci_file} -p {pheno_file} -d {d_file} -u {u_file} -maf 0 -miss 0.2 -lmm -notsnp -o {outfile}\"\n",
    "#     print(string)\n",
    "#     process = string.split(\" \")\n",
    "#     result = subprocess.run(process, stdout=subprocess.PIPE)\n",
    "\n",
    "#     stdout = result.stdout.decode('utf-8').split(\"\\n\")\n",
    "\n",
    "#     print(stdout)\n",
    "#     if len(stdout) > 8:\n",
    "#         out3 = stdout[7]\n",
    "#         out1 = stdout[8]\n",
    "#         out2 = stdout[9]\n",
    "#         assert 'pve estimate' in out1\n",
    "#         assert 'se(pve)' in out2\n",
    "#         assert 'number of analyzed SNPs' in out3\n",
    "#         pve = float(out1.split(\"=\")[-1])\n",
    "#         se = float(out2.split(\"=\")[-1])\n",
    "#         N_snps = float(out3.split(\"=\")[-1])\n",
    "#         table.append([drug, pve, se, N_snps])\n",
    "    \n",
    "#     print(pve, se)\n",
    "    \n",
    "# df = pd.DataFrame(table, columns=[\"drug\", \"pve\", \"se_pve\", \"N_snps\"])\n",
    "# df.to_csv(\"possible_PVE_ABR_UK_sites.csv\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run for SNPs + pairs that are in ABR genes or epistatic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "annotation for N Snps: 782565\n"
     ]
    }
   ],
   "source": [
    "full_range = pd.read_csv(\"../output/05.single_mutations/snps_with_all_annotation.csv\", index_col=0)\n",
    "full_range = full_range.query(\"in_possible_ABR_gene\")\n",
    "full_range[\"possible_drug\"] = [x.split(\",\") for x in full_range.possible_drug]\n",
    "\n",
    "# reads in the SNP annotations\n",
    "snp_annotation = pd.read_pickle(\"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_SNP_annotation.pkl\")\n",
    "snp_annotation['mat_index'] =snp_annotation.index\n",
    "print('annotation for N Snps:', len(snp_annotation))\n",
    "\n",
    "isolates = pd.read_pickle(kwargs['isolate_annotation_file'])\n",
    "#g = np.load(\"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_matrix.npy\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AMIKACIN AMI\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘output/AMIKACIN’: File exists\n",
      "mkdir: cannot create directory ‘output/AMIKACIN/0.input_loci_files’: File exists\n",
      "mkdir: cannot create directory ‘output/AMIKACIN/1.job_submission_files’: File exists\n",
      "mkdir: cannot create directory ‘output/AMIKACIN/2.run_output’: File exists\n",
      "/home/ag300/anaconda3/envs/jupytervenv2/lib/python3.7/site-packages/ipykernel_launcher.py:44: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1205\n",
      "1206\n",
      "1205\n",
      "1206\n",
      "Number of isolates 1206\n",
      "Number of SNPS 179\n",
      "encoding gaps with gap code nan\n",
      "<class 'numpy.ndarray'>\n",
      "shape of the genotype matrix (1206, 179)\n",
      "newly created cross matrix dimensions (1206, 280)\n",
      "Creating locus file for AMIKACIN with N positions: 280\n",
      "Made genotype matrix with shape (1206, 280)\n",
      "writing Gemma with gap code NA\n",
      "calculating reference allele matrix with gap code NA\n",
      "encoding gaps with gap code NA\n",
      "<class 'numpy.ndarray'>\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/AMIKACIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/AMIKACIN/AMIKACIN.phenotypes -gk -maf 0 -o AMIKACIN/ABR_with_ABRpairs\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/AMIKACIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/AMIKACIN/AMIKACIN.phenotypes -k output/AMIKACIN/ABR_with_ABRpairs.cXX.txt -maf 0 -eigen -notsnp -o AMIKACIN/ABR_with_ABRpairs\n",
      "creating GRM\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1206\n",
      "## number of analyzed individuals = 1206\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      280\n",
      "## number of analyzed SNPs         =      268\n",
      "Calculating Relatedness Matrix ... \n",
      "================================================== 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running eigendecomp\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1206\n",
      "## number of analyzed individuals = 1206\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      280\n",
      "## number of analyzed SNPs         =      268\n",
      "Start Eigen-Decomposition...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 1009 eigenvalues close to zero\n",
      "**** INFO: Done.\n",
      "mkdir: cannot create directory ‘output/CAPREOMYCIN’: File exists\n",
      "mkdir: cannot create directory ‘output/CAPREOMYCIN/0.input_loci_files’: File exists\n",
      "mkdir: cannot create directory ‘output/CAPREOMYCIN/1.job_submission_files’: File exists\n",
      "mkdir: cannot create directory ‘output/CAPREOMYCIN/2.run_output’: File exists\n",
      "/home/ag300/anaconda3/envs/jupytervenv2/lib/python3.7/site-packages/ipykernel_launcher.py:44: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAPREOMYCIN CAP\n",
      "1101\n",
      "1102\n",
      "1101\n",
      "1102\n",
      "Number of isolates 1102\n",
      "Number of SNPS 148\n",
      "encoding gaps with gap code nan\n",
      "<class 'numpy.ndarray'>\n",
      "shape of the genotype matrix (1102, 148)\n",
      "newly created cross matrix dimensions (1102, 215)\n",
      "Creating locus file for CAPREOMYCIN with N positions: 215\n",
      "Made genotype matrix with shape (1102, 215)\n",
      "writing Gemma with gap code NA\n",
      "calculating reference allele matrix with gap code NA\n",
      "encoding gaps with gap code NA\n",
      "<class 'numpy.ndarray'>\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/CAPREOMYCIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/CAPREOMYCIN/CAPREOMYCIN.phenotypes -gk -maf 0 -o CAPREOMYCIN/ABR_with_ABRpairs\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/CAPREOMYCIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/CAPREOMYCIN/CAPREOMYCIN.phenotypes -k output/CAPREOMYCIN/ABR_with_ABRpairs.cXX.txt -maf 0 -eigen -notsnp -o CAPREOMYCIN/ABR_with_ABRpairs\n",
      "creating GRM\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1102\n",
      "## number of analyzed individuals = 1102\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      215\n",
      "## number of analyzed SNPs         =      206\n",
      "Calculating Relatedness Matrix ... \n",
      "================================================== 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running eigendecomp\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1102\n",
      "## number of analyzed individuals = 1102\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      215\n",
      "## number of analyzed SNPs         =      206\n",
      "Start Eigen-Decomposition...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 946 eigenvalues close to zero\n",
      "**** INFO: Done.\n",
      "mkdir: cannot create directory ‘output/ETHAMBUTOL’: File exists\n",
      "mkdir: cannot create directory ‘output/ETHAMBUTOL/0.input_loci_files’: File exists\n",
      "mkdir: cannot create directory ‘output/ETHAMBUTOL/1.job_submission_files’: File exists\n",
      "mkdir: cannot create directory ‘output/ETHAMBUTOL/2.run_output’: File exists\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ETHAMBUTOL EMB\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ag300/anaconda3/envs/jupytervenv2/lib/python3.7/site-packages/ipykernel_launcher.py:44: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1549\n",
      "1488\n",
      "1549\n",
      "1488\n",
      "Number of isolates 1488\n",
      "Number of SNPS 287\n",
      "encoding gaps with gap code nan\n",
      "<class 'numpy.ndarray'>\n",
      "shape of the genotype matrix (1488, 287)\n",
      "newly created cross matrix dimensions (1488, 560)\n",
      "Creating locus file for ETHAMBUTOL with N positions: 560\n",
      "Made genotype matrix with shape (1488, 560)\n",
      "writing Gemma with gap code NA\n",
      "calculating reference allele matrix with gap code NA\n",
      "encoding gaps with gap code NA\n",
      "<class 'numpy.ndarray'>\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/ETHAMBUTOL/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/ETHAMBUTOL/ETHAMBUTOL.phenotypes -gk -maf 0 -o ETHAMBUTOL/ABR_with_ABRpairs\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/ETHAMBUTOL/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/ETHAMBUTOL/ETHAMBUTOL.phenotypes -k output/ETHAMBUTOL/ABR_with_ABRpairs.cXX.txt -maf 0 -eigen -notsnp -o ETHAMBUTOL/ABR_with_ABRpairs\n",
      "creating GRM\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1488\n",
      "## number of analyzed individuals = 1488\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      560\n",
      "## number of analyzed SNPs         =      550\n",
      "Calculating Relatedness Matrix ... \n",
      "================================================== 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running eigendecomp\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1488\n",
      "## number of analyzed individuals = 1488\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      560\n",
      "## number of analyzed SNPs         =      550\n",
      "Start Eigen-Decomposition...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 1088 eigenvalues close to zero\n",
      "**** INFO: Done.\n",
      "mkdir: cannot create directory ‘output/ETHIONAMIDE’: File exists\n",
      "mkdir: cannot create directory ‘output/ETHIONAMIDE/0.input_loci_files’: File exists\n",
      "mkdir: cannot create directory ‘output/ETHIONAMIDE/1.job_submission_files’: File exists\n",
      "mkdir: cannot create directory ‘output/ETHIONAMIDE/2.run_output’: File exists\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ETHIONAMIDE ETA\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ag300/anaconda3/envs/jupytervenv2/lib/python3.7/site-packages/ipykernel_launcher.py:44: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1212\n",
      "1213\n",
      "1212\n",
      "1213\n",
      "Number of isolates 1213\n",
      "Number of SNPS 104\n",
      "encoding gaps with gap code nan\n",
      "<class 'numpy.ndarray'>\n",
      "shape of the genotype matrix (1213, 104)\n",
      "newly created cross matrix dimensions (1213, 149)\n",
      "Creating locus file for ETHIONAMIDE with N positions: 149\n",
      "Made genotype matrix with shape (1213, 149)\n",
      "writing Gemma with gap code NA\n",
      "calculating reference allele matrix with gap code NA\n",
      "encoding gaps with gap code NA\n",
      "<class 'numpy.ndarray'>\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/ETHIONAMIDE/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/ETHIONAMIDE/ETHIONAMIDE.phenotypes -gk -maf 0 -o ETHIONAMIDE/ABR_with_ABRpairs\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/ETHIONAMIDE/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/ETHIONAMIDE/ETHIONAMIDE.phenotypes -k output/ETHIONAMIDE/ABR_with_ABRpairs.cXX.txt -maf 0 -eigen -notsnp -o ETHIONAMIDE/ABR_with_ABRpairs\n",
      "creating GRM\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1213\n",
      "## number of analyzed individuals = 1213\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      149\n",
      "## number of analyzed SNPs         =      145\n",
      "Calculating Relatedness Matrix ... \n",
      "================================================== 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running eigendecomp\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1213\n",
      "## number of analyzed individuals = 1213\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      149\n",
      "## number of analyzed SNPs         =      145\n",
      "Start Eigen-Decomposition...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 1091 eigenvalues close to zero\n",
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ISONIAZID INH\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘output/ISONIAZID’: File exists\n",
      "mkdir: cannot create directory ‘output/ISONIAZID/0.input_loci_files’: File exists\n",
      "mkdir: cannot create directory ‘output/ISONIAZID/1.job_submission_files’: File exists\n",
      "mkdir: cannot create directory ‘output/ISONIAZID/2.run_output’: File exists\n",
      "/home/ag300/anaconda3/envs/jupytervenv2/lib/python3.7/site-packages/ipykernel_launcher.py:44: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1554\n",
      "1497\n",
      "1554\n",
      "1497\n",
      "Number of isolates 1497\n",
      "Number of SNPS 238\n",
      "encoding gaps with gap code nan\n",
      "<class 'numpy.ndarray'>\n",
      "shape of the genotype matrix (1497, 238)\n",
      "newly created cross matrix dimensions (1497, 411)\n",
      "Creating locus file for ISONIAZID with N positions: 411\n",
      "Made genotype matrix with shape (1497, 411)\n",
      "writing Gemma with gap code NA\n",
      "calculating reference allele matrix with gap code NA\n",
      "encoding gaps with gap code NA\n",
      "<class 'numpy.ndarray'>\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/ISONIAZID/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/ISONIAZID/ISONIAZID.phenotypes -gk -maf 0 -o ISONIAZID/ABR_with_ABRpairs\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/ISONIAZID/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/ISONIAZID/ISONIAZID.phenotypes -k output/ISONIAZID/ABR_with_ABRpairs.cXX.txt -maf 0 -eigen -notsnp -o ISONIAZID/ABR_with_ABRpairs\n",
      "creating GRM\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1497\n",
      "## number of analyzed individuals = 1497\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      411\n",
      "## number of analyzed SNPs         =      403\n",
      "Calculating Relatedness Matrix ... \n",
      "================================================== 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running eigendecomp\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1497\n",
      "## number of analyzed individuals = 1497\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      411\n",
      "## number of analyzed SNPs         =      403\n",
      "Start Eigen-Decomposition...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 1211 eigenvalues close to zero\n",
      "**** INFO: Done.\n",
      "mkdir: cannot create directory ‘output/KANAMYCIN’: File exists\n",
      "mkdir: cannot create directory ‘output/KANAMYCIN/0.input_loci_files’: File exists\n",
      "mkdir: cannot create directory ‘output/KANAMYCIN/1.job_submission_files’: File exists\n",
      "mkdir: cannot create directory ‘output/KANAMYCIN/2.run_output’: File exists\n",
      "/home/ag300/anaconda3/envs/jupytervenv2/lib/python3.7/site-packages/ipykernel_launcher.py:44: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KANAMYCIN KAN\n",
      "1214\n",
      "1215\n",
      "1214\n",
      "1215\n",
      "Number of isolates 1215\n",
      "Number of SNPS 143\n",
      "encoding gaps with gap code nan\n",
      "<class 'numpy.ndarray'>\n",
      "shape of the genotype matrix (1215, 143)\n",
      "newly created cross matrix dimensions (1215, 244)\n",
      "Creating locus file for KANAMYCIN with N positions: 244\n",
      "Made genotype matrix with shape (1215, 244)\n",
      "writing Gemma with gap code NA\n",
      "calculating reference allele matrix with gap code NA\n",
      "encoding gaps with gap code NA\n",
      "<class 'numpy.ndarray'>\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/KANAMYCIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/KANAMYCIN/KANAMYCIN.phenotypes -gk -maf 0 -o KANAMYCIN/ABR_with_ABRpairs\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/KANAMYCIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/KANAMYCIN/KANAMYCIN.phenotypes -k output/KANAMYCIN/ABR_with_ABRpairs.cXX.txt -maf 0 -eigen -notsnp -o KANAMYCIN/ABR_with_ABRpairs\n",
      "creating GRM\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1215\n",
      "## number of analyzed individuals = 1215\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      244\n",
      "## number of analyzed SNPs         =      236\n",
      "Calculating Relatedness Matrix ... \n",
      "================================================== 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running eigendecomp\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1215\n",
      "## number of analyzed individuals = 1215\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      244\n",
      "## number of analyzed SNPs         =      236\n",
      "Start Eigen-Decomposition...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 1048 eigenvalues close to zero\n",
      "**** INFO: Done.\n",
      "mkdir: cannot create directory ‘output/MOXIFLOXACIN’: File exists\n",
      "mkdir: cannot create directory ‘output/MOXIFLOXACIN/0.input_loci_files’: File exists\n",
      "mkdir: cannot create directory ‘output/MOXIFLOXACIN/1.job_submission_files’: File exists\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MOXIFLOXACIN MOXI\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘output/MOXIFLOXACIN/2.run_output’: File exists\n",
      "/home/ag300/anaconda3/envs/jupytervenv2/lib/python3.7/site-packages/ipykernel_launcher.py:44: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1089\n",
      "1090\n",
      "1089\n",
      "1090\n",
      "Number of isolates 1090\n",
      "Number of SNPS 40\n",
      "encoding gaps with gap code nan\n",
      "<class 'numpy.ndarray'>\n",
      "shape of the genotype matrix (1090, 40)\n",
      "newly created cross matrix dimensions (1090, 55)\n",
      "Creating locus file for MOXIFLOXACIN with N positions: 55\n",
      "Made genotype matrix with shape (1090, 55)\n",
      "writing Gemma with gap code NA\n",
      "calculating reference allele matrix with gap code NA\n",
      "encoding gaps with gap code NA\n",
      "<class 'numpy.ndarray'>\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/MOXIFLOXACIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/MOXIFLOXACIN/MOXIFLOXACIN.phenotypes -gk -maf 0 -o MOXIFLOXACIN/ABR_with_ABRpairs\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/MOXIFLOXACIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/MOXIFLOXACIN/MOXIFLOXACIN.phenotypes -k output/MOXIFLOXACIN/ABR_with_ABRpairs.cXX.txt -maf 0 -eigen -notsnp -o MOXIFLOXACIN/ABR_with_ABRpairs\n",
      "creating GRM\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1090\n",
      "## number of analyzed individuals = 1090\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =       55\n",
      "## number of analyzed SNPs         =       50\n",
      "Calculating Relatedness Matrix ... \n",
      "================================================== 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running eigendecomp\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1090\n",
      "## number of analyzed individuals = 1090\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =       55\n",
      "## number of analyzed SNPs         =       50\n",
      "Start Eigen-Decomposition...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 1048 eigenvalues close to zero\n",
      "**** INFO: Done.\n",
      "mkdir: cannot create directory ‘output/PYRAZINAMIDE’: File exists\n",
      "mkdir: cannot create directory ‘output/PYRAZINAMIDE/0.input_loci_files’: File exists\n",
      "mkdir: cannot create directory ‘output/PYRAZINAMIDE/1.job_submission_files’: File exists\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PYRAZINAMIDE PZA\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘output/PYRAZINAMIDE/2.run_output’: File exists\n",
      "/home/ag300/anaconda3/envs/jupytervenv2/lib/python3.7/site-packages/ipykernel_launcher.py:44: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1106\n",
      "1107\n",
      "1106\n",
      "1107\n",
      "Number of isolates 1107\n",
      "Number of SNPS 259\n",
      "encoding gaps with gap code nan\n",
      "<class 'numpy.ndarray'>\n",
      "shape of the genotype matrix (1107, 259)\n",
      "newly created cross matrix dimensions (1107, 388)\n",
      "Creating locus file for PYRAZINAMIDE with N positions: 388\n",
      "Made genotype matrix with shape (1107, 388)\n",
      "writing Gemma with gap code NA\n",
      "calculating reference allele matrix with gap code NA\n",
      "encoding gaps with gap code NA\n",
      "<class 'numpy.ndarray'>\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/PYRAZINAMIDE/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/PYRAZINAMIDE/PYRAZINAMIDE.phenotypes -gk -maf 0 -o PYRAZINAMIDE/ABR_with_ABRpairs\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/PYRAZINAMIDE/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/PYRAZINAMIDE/PYRAZINAMIDE.phenotypes -k output/PYRAZINAMIDE/ABR_with_ABRpairs.cXX.txt -maf 0 -eigen -notsnp -o PYRAZINAMIDE/ABR_with_ABRpairs\n",
      "creating GRM\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1107\n",
      "## number of analyzed individuals = 1107\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      388\n",
      "## number of analyzed SNPs         =      371\n",
      "Calculating Relatedness Matrix ... \n",
      "================================================== 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running eigendecomp\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1107\n",
      "## number of analyzed individuals = 1107\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      388\n",
      "## number of analyzed SNPs         =      371\n",
      "Start Eigen-Decomposition...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 826 eigenvalues close to zero\n",
      "**** INFO: Done.\n",
      "mkdir: cannot create directory ‘output/RIFAMPICIN’: File exists\n",
      "mkdir: cannot create directory ‘output/RIFAMPICIN/0.input_loci_files’: File exists\n",
      "mkdir: cannot create directory ‘output/RIFAMPICIN/1.job_submission_files’: File exists\n",
      "mkdir: cannot create directory ‘output/RIFAMPICIN/2.run_output’: File exists\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RIFAMPICIN RIF\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ag300/anaconda3/envs/jupytervenv2/lib/python3.7/site-packages/ipykernel_launcher.py:44: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1597\n",
      "1522\n",
      "1597\n",
      "1522\n",
      "Number of isolates 1522\n",
      "Number of SNPS 253\n",
      "encoding gaps with gap code nan\n",
      "<class 'numpy.ndarray'>\n",
      "shape of the genotype matrix (1522, 253)\n",
      "newly created cross matrix dimensions (1522, 405)\n",
      "Creating locus file for RIFAMPICIN with N positions: 405\n",
      "Made genotype matrix with shape (1522, 405)\n",
      "writing Gemma with gap code NA\n",
      "calculating reference allele matrix with gap code NA\n",
      "encoding gaps with gap code NA\n",
      "<class 'numpy.ndarray'>\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/RIFAMPICIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/RIFAMPICIN/RIFAMPICIN.phenotypes -gk -maf 0 -o RIFAMPICIN/ABR_with_ABRpairs\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/RIFAMPICIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/RIFAMPICIN/RIFAMPICIN.phenotypes -k output/RIFAMPICIN/ABR_with_ABRpairs.cXX.txt -maf 0 -eigen -notsnp -o RIFAMPICIN/ABR_with_ABRpairs\n",
      "creating GRM\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1522\n",
      "## number of analyzed individuals = 1522\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      405\n",
      "## number of analyzed SNPs         =      402\n",
      "Calculating Relatedness Matrix ... \n",
      "================================================== 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running eigendecomp\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1522\n",
      "## number of analyzed individuals = 1522\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      405\n",
      "## number of analyzed SNPs         =      402\n",
      "Start Eigen-Decomposition...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 1223 eigenvalues close to zero\n",
      "**** INFO: Done.\n",
      "mkdir: cannot create directory ‘output/STREPTOMYCIN’: File exists\n",
      "mkdir: cannot create directory ‘output/STREPTOMYCIN/0.input_loci_files’: File exists\n",
      "mkdir: cannot create directory ‘output/STREPTOMYCIN/1.job_submission_files’: File exists\n",
      "mkdir: cannot create directory ‘output/STREPTOMYCIN/2.run_output’: File exists\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STREPTOMYCIN STR\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ag300/anaconda3/envs/jupytervenv2/lib/python3.7/site-packages/ipykernel_launcher.py:44: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1543\n",
      "1489\n",
      "1543\n",
      "1489\n",
      "Number of isolates 1489\n",
      "Number of SNPS 391\n",
      "encoding gaps with gap code nan\n",
      "<class 'numpy.ndarray'>\n",
      "shape of the genotype matrix (1489, 391)\n",
      "newly created cross matrix dimensions (1489, 607)\n",
      "Creating locus file for STREPTOMYCIN with N positions: 607\n",
      "Made genotype matrix with shape (1489, 607)\n",
      "writing Gemma with gap code NA\n",
      "calculating reference allele matrix with gap code NA\n",
      "encoding gaps with gap code NA\n",
      "<class 'numpy.ndarray'>\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/STREPTOMYCIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/STREPTOMYCIN/STREPTOMYCIN.phenotypes -gk -maf 0 -o STREPTOMYCIN/ABR_with_ABRpairs\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/STREPTOMYCIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/STREPTOMYCIN/STREPTOMYCIN.phenotypes -k output/STREPTOMYCIN/ABR_with_ABRpairs.cXX.txt -maf 0 -eigen -notsnp -o STREPTOMYCIN/ABR_with_ABRpairs\n",
      "creating GRM\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1489\n",
      "## number of analyzed individuals = 1489\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      607\n",
      "## number of analyzed SNPs         =      599\n",
      "Calculating Relatedness Matrix ... \n",
      "================================================== 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running eigendecomp\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1489\n",
      "## number of analyzed individuals = 1489\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      607\n",
      "## number of analyzed SNPs         =      599\n",
      "Start Eigen-Decomposition...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 1041 eigenvalues close to zero\n",
      "**** INFO: Done.\n"
     ]
    }
   ],
   "source": [
    "for drug, drug_abbrev in drug_to_abbrev.items():\n",
    "    \n",
    "    ## \n",
    "    _keep = []\n",
    "    for idx, row in full_range.iterrows():\n",
    "        if drug_abbrev in row.possible_drug:\n",
    "            _keep.append(idx)\n",
    "        elif drug_abbrev=='ETA':\n",
    "            if \"ETH\" in row.possible_drug:\n",
    "                _keep.append(idx)\n",
    "        elif drug_abbrev=='MOXI':\n",
    "            if \"MXF\" in row.possible_drug:\n",
    "                _keep.append(idx)\n",
    "        elif drug_abbrev=='STR':\n",
    "            if \"STM\" in row.possible_drug:\n",
    "                _keep.append(idx)\n",
    "                \n",
    "    positions_found = list(set(full_range.loc[_keep].pos))\n",
    "    \n",
    "    antibiotic_hits = pd.read_csv(\n",
    "        f\"/n/data1/hms/dbmi/farhat/anna/ABR_epistasis/DependentMutations2/output/04.antibiotic/{drug_abbrev}_antibiotic_hits.csv\"\n",
    "    )\n",
    "    positions_found = positions_found + list(antibiotic_hits.position_j) + list(antibiotic_hits.position_i)\n",
    "    snps_subset = snp_annotation.query(\"pos in @positions_found\")\n",
    "    \n",
    "    additional_sites = antibiotic_hits\n",
    "    \n",
    "    print(drug, drug_abbrev)\n",
    "    # Generate the cXX file\n",
    "    \n",
    "    prefix = f\"output/{drug}\"\n",
    "    output_prefix = f\"{drug}\"\n",
    "    os.system(f\"mkdir {prefix}\")\n",
    "    os.system(f\"mkdir {prefix}/0.input_loci_files\")\n",
    "    os.system(f\"mkdir {prefix}/1.job_submission_files\")\n",
    "    os.system(f\"mkdir {prefix}/2.run_output\")\n",
    "\n",
    "    # get the isolates with phenotypes to keep\n",
    "    phenotypes = pd.read_csv(kwargs['phenotype_file'], index_col=0)\n",
    "    phenotypes.replace(\"NA\", np.nan, inplace=True)\n",
    "    \n",
    "    phenotypes_found_all = phenotypes.loc[~phenotypes[drug_abbrev].isnull(),:]\n",
    "    phenotypes_found = phenotypes_found_all[[\"ROLLINGDB_ID\", f\"{drug_abbrev}_midpoint\"]]\n",
    "    phenotypes_found[drug] = np.log(phenotypes_found[f\"{drug_abbrev}_midpoint\"])\n",
    "\n",
    "    isolates = pd.read_pickle(kwargs['isolate_annotation_file'])\n",
    "    isolates_with_phenotype = isolates.query(\"isolate_ID in @phenotypes_found.ROLLINGDB_ID\")\n",
    "    geno_mat = np.load(\"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_matrix.npy\")\n",
    "\n",
    "    #### Prepare the phenotype input file ####\n",
    "    pheno_file = f\"{prefix}/{drug}.phenotypes\"\n",
    "    write_phenotype_file_MIC(pheno_file, phenotypes_found, isolates_with_phenotype, drug)\n",
    "\n",
    "    \n",
    "    matrix = geno_mat[snps_subset.index, :][:, isolates_with_phenotype.index]\n",
    "    \n",
    "    #### Prepare the phenotype input file ####\n",
    "    pheno_file = f\"{prefix}/{drug}.phenotypes\"\n",
    "    write_phenotype_file_MIC(pheno_file, phenotypes_found, isolates_with_phenotype, drug)\n",
    "\n",
    "    \n",
    "    matrix = geno_mat[snps_subset.index, :][:, isolates_with_phenotype.index]\n",
    "    \n",
    "    ## Add the MTB ancestor sequence\n",
    "    \n",
    "    # read the ancestor sequence\n",
    "    ancestral_seq_file=\"../input/MTB_ancestor_reference_modified.fasta\"\n",
    "    # subtract 1 or get wrecked. H37Rv positions are 1-indexed\n",
    "    all_positions = [int(x) - 1 for x in snps_subset.pos]\n",
    "    ancestral_seq = Alignment.from_file(open(ancestral_seq_file, \"r\"))\n",
    "    ancestral_seq_subset = np.array([ancestral_seq.matrix[0,x] for x in all_positions])\n",
    "    \n",
    "    # convert to matrix encoding\n",
    "    ancestral_seq_subset = np.array([genotype_translator_r[x] for x in list(ancestral_seq_subset)])\n",
    "    \n",
    "    # concatenate as FIRST sequence\n",
    "    matrix = np.concatenate([ancestral_seq_subset.reshape(-1,1), matrix], axis=1)\n",
    "    \n",
    "    isolates = [\"MTB_ANC\"] + list(isolates_with_phenotype.isolate_ID)\n",
    "\n",
    "    # add to strain list\n",
    "    \n",
    "    gm = GenotypeMatrix(\n",
    "        position_list = snps_subset.pos,\n",
    "        strain_list = isolates,\n",
    "        matrix_data=matrix.T\n",
    "    )\n",
    "    \n",
    "    print(\"Number of isolates\", len(isolates))\n",
    "    print(\"Number of SNPS\", len(snps_subset))\n",
    "    ## Convert to major allele matrix\n",
    "    geno_mat = calculate_reference_allele_matrix_for_pair_regression(gm.matrix, ancestral_seq_subset) \n",
    "    print(\"shape of the genotype matrix\", geno_mat.shape)\n",
    "    snps_in_analysis = list(snps_subset.pos)\n",
    "    ### To the matrix, add the pairs of SNPs\n",
    "    \n",
    "    matrix = deepcopy(geno_mat)\n",
    "    for idx, row in additional_sites.iterrows():\n",
    "        # lookup the index of the two positions\n",
    "        pos_i = row.position_i\n",
    "        pos_j = row.position_j\n",
    "        \n",
    "        idx_i = gm.position_to_index[pos_i]\n",
    "        idx_j = gm.position_to_index[pos_j]\n",
    "        \n",
    "        new_row = np.multiply(geno_mat[:,idx_i], geno_mat[:, idx_j]).T\n",
    "        matrix = np.concatenate([matrix, new_row.reshape(-1,1)], axis=1)\n",
    "        \n",
    "        snps_in_analysis.append(f\"{pos_i}_{pos_j}\")\n",
    "\n",
    "# #     # Generate the locus file\n",
    "    print(\"newly created cross matrix dimensions\", matrix.shape)\n",
    "    relatedness_loci_file = f\"{prefix}/0.input_loci_files/ABR_sites_plus_all_dependent.loci\"\n",
    "\n",
    "    print(f\"Creating locus file for {drug} with N positions: {len(snps_in_analysis)}\")\n",
    "\n",
    "    prepare_loci_file(matrix, isolates, snps_in_analysis, relatedness_loci_file)\n",
    "\n",
    "    cxx_file = f\"{prefix}/ABR_with_ABRpairs.cXX.txt\"\n",
    "\n",
    "    outfile = f\"{output_prefix}/ABR_with_ABRpairs\"\n",
    "    string1 = f\"{kwargs['gemma']} -g {relatedness_loci_file} -p {pheno_file} -gk -maf 0 -o {outfile}\"\n",
    "    print(string1)\n",
    "\n",
    "    # Run the eigen decomposition\n",
    "    outfile =f'{output_prefix}/ABR_with_ABRpairs'\n",
    "    string2 = f\"{kwargs['gemma']} -g {relatedness_loci_file} -p {pheno_file} -k {cxx_file} -maf 0 -eigen -notsnp -o {outfile}\"\n",
    "    print(string2)\n",
    "\n",
    "    print(\"creating GRM\")\n",
    "    os.system(string1)\n",
    "\n",
    "    print(\"running eigendecomp\")\n",
    "    os.system(string2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AMIKACIN\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/AMIKACIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/AMIKACIN/AMIKACIN.phenotypes -k output/AMIKACIN/ABR_with_ABRpairs.cXX.txt -maf 0 -miss 0.2 -lmm -notsnp -o heritability_temp.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 1009 eigenvalues close to zero\n",
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1206', '## number of analyzed individuals = 1206', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      280', '## number of analyzed SNPs         =      268', 'Start Eigen-Decomposition...', 'pve estimate =0.565059', 'se(pve) =0.0408507', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1206', '## number of analyzed individuals = 1206', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      280', '## number of analyzed SNPs         =      268', 'Start Eigen-Decomposition...', 'pve estimate =0.565059', 'se(pve) =0.0408507', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "0.565059 0.0408507\n",
      "CAPREOMYCIN\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/CAPREOMYCIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/CAPREOMYCIN/CAPREOMYCIN.phenotypes -k output/CAPREOMYCIN/ABR_with_ABRpairs.cXX.txt -maf 0 -miss 0.2 -lmm -notsnp -o heritability_temp.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 946 eigenvalues close to zero\n",
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1102', '## number of analyzed individuals = 1102', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      215', '## number of analyzed SNPs         =      206', 'Start Eigen-Decomposition...', 'pve estimate =0.355128', 'se(pve) =0.0500529', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1102', '## number of analyzed individuals = 1102', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      215', '## number of analyzed SNPs         =      206', 'Start Eigen-Decomposition...', 'pve estimate =0.355128', 'se(pve) =0.0500529', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "0.355128 0.0500529\n",
      "ETHAMBUTOL\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/ETHAMBUTOL/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/ETHAMBUTOL/ETHAMBUTOL.phenotypes -k output/ETHAMBUTOL/ABR_with_ABRpairs.cXX.txt -maf 0 -miss 0.2 -lmm -notsnp -o heritability_temp.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 1088 eigenvalues close to zero\n",
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1488', '## number of analyzed individuals = 1488', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      560', '## number of analyzed SNPs         =      550', 'Start Eigen-Decomposition...', 'pve estimate =0.468603', 'se(pve) =0.0389708', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1488', '## number of analyzed individuals = 1488', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      560', '## number of analyzed SNPs         =      550', 'Start Eigen-Decomposition...', 'pve estimate =0.468603', 'se(pve) =0.0389708', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "0.468603 0.0389708\n",
      "ETHIONAMIDE\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/ETHIONAMIDE/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/ETHIONAMIDE/ETHIONAMIDE.phenotypes -k output/ETHIONAMIDE/ABR_with_ABRpairs.cXX.txt -maf 0 -miss 0.2 -lmm -notsnp -o heritability_temp.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 1091 eigenvalues close to zero\n",
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1213', '## number of analyzed individuals = 1213', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      149', '## number of analyzed SNPs         =      145', 'Start Eigen-Decomposition...', 'pve estimate =0.23046', 'se(pve) =0.0451342', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1213', '## number of analyzed individuals = 1213', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      149', '## number of analyzed SNPs         =      145', 'Start Eigen-Decomposition...', 'pve estimate =0.23046', 'se(pve) =0.0451342', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "0.23046 0.0451342\n",
      "ISONIAZID\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/ISONIAZID/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/ISONIAZID/ISONIAZID.phenotypes -k output/ISONIAZID/ABR_with_ABRpairs.cXX.txt -maf 0 -miss 0.2 -lmm -notsnp -o heritability_temp.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 1211 eigenvalues close to zero\n",
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1497', '## number of analyzed individuals = 1497', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      411', '## number of analyzed SNPs         =      403', 'Start Eigen-Decomposition...', 'pve estimate =0.568829', 'se(pve) =0.0395085', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1497', '## number of analyzed individuals = 1497', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      411', '## number of analyzed SNPs         =      403', 'Start Eigen-Decomposition...', 'pve estimate =0.568829', 'se(pve) =0.0395085', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "0.568829 0.0395085\n",
      "KANAMYCIN\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/KANAMYCIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/KANAMYCIN/KANAMYCIN.phenotypes -k output/KANAMYCIN/ABR_with_ABRpairs.cXX.txt -maf 0 -miss 0.2 -lmm -notsnp -o heritability_temp.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 1048 eigenvalues close to zero\n",
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1215', '## number of analyzed individuals = 1215', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      244', '## number of analyzed SNPs         =      236', 'Start Eigen-Decomposition...', 'pve estimate =0.632181', 'se(pve) =0.0402792', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1215', '## number of analyzed individuals = 1215', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      244', '## number of analyzed SNPs         =      236', 'Start Eigen-Decomposition...', 'pve estimate =0.632181', 'se(pve) =0.0402792', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "0.632181 0.0402792\n",
      "MOXIFLOXACIN\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/MOXIFLOXACIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/MOXIFLOXACIN/MOXIFLOXACIN.phenotypes -k output/MOXIFLOXACIN/ABR_with_ABRpairs.cXX.txt -maf 0 -miss 0.2 -lmm -notsnp -o heritability_temp.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 1048 eigenvalues close to zero\n",
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1090', '## number of analyzed individuals = 1090', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =       55', '## number of analyzed SNPs         =       50', 'Start Eigen-Decomposition...', 'pve estimate =0.492332', 'se(pve) =0.0641681', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1090', '## number of analyzed individuals = 1090', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =       55', '## number of analyzed SNPs         =       50', 'Start Eigen-Decomposition...', 'pve estimate =0.492332', 'se(pve) =0.0641681', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "0.492332 0.0641681\n",
      "PYRAZINAMIDE\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/PYRAZINAMIDE/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/PYRAZINAMIDE/PYRAZINAMIDE.phenotypes -k output/PYRAZINAMIDE/ABR_with_ABRpairs.cXX.txt -maf 0 -miss 0.2 -lmm -notsnp -o heritability_temp.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 826 eigenvalues close to zero\n",
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1107', '## number of analyzed individuals = 1107', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      388', '## number of analyzed SNPs         =      371', 'Start Eigen-Decomposition...', 'pve estimate =0.646346', 'se(pve) =0.0522791', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1107', '## number of analyzed individuals = 1107', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      388', '## number of analyzed SNPs         =      371', 'Start Eigen-Decomposition...', 'pve estimate =0.646346', 'se(pve) =0.0522791', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "0.646346 0.0522791\n",
      "RIFAMPICIN\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/RIFAMPICIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/RIFAMPICIN/RIFAMPICIN.phenotypes -k output/RIFAMPICIN/ABR_with_ABRpairs.cXX.txt -maf 0 -miss 0.2 -lmm -notsnp -o heritability_temp.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 1223 eigenvalues close to zero\n",
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1522', '## number of analyzed individuals = 1522', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      405', '## number of analyzed SNPs         =      402', 'Start Eigen-Decomposition...', 'pve estimate =0.526072', 'se(pve) =0.0364663', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1522', '## number of analyzed individuals = 1522', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      405', '## number of analyzed SNPs         =      402', 'Start Eigen-Decomposition...', 'pve estimate =0.526072', 'se(pve) =0.0364663', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "0.526072 0.0364663\n",
      "STREPTOMYCIN\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/STREPTOMYCIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/STREPTOMYCIN/STREPTOMYCIN.phenotypes -k output/STREPTOMYCIN/ABR_with_ABRpairs.cXX.txt -maf 0 -miss 0.2 -lmm -notsnp -o heritability_temp.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 1041 eigenvalues close to zero\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1489', '## number of analyzed individuals = 1489', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      607', '## number of analyzed SNPs         =      599', 'Start Eigen-Decomposition...', 'pve estimate =0.478958', 'se(pve) =0.0368205', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1489', '## number of analyzed individuals = 1489', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      607', '## number of analyzed SNPs         =      599', 'Start Eigen-Decomposition...', 'pve estimate =0.478958', 'se(pve) =0.0368205', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "0.478958 0.0368205\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** INFO: Done.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug</th>\n",
       "      <th>pve</th>\n",
       "      <th>se_pve</th>\n",
       "      <th>N_snps</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMIKACIN</td>\n",
       "      <td>0.565059</td>\n",
       "      <td>0.040851</td>\n",
       "      <td>268.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAPREOMYCIN</td>\n",
       "      <td>0.355128</td>\n",
       "      <td>0.050053</td>\n",
       "      <td>206.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ETHAMBUTOL</td>\n",
       "      <td>0.468603</td>\n",
       "      <td>0.038971</td>\n",
       "      <td>550.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ETHIONAMIDE</td>\n",
       "      <td>0.230460</td>\n",
       "      <td>0.045134</td>\n",
       "      <td>145.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ISONIAZID</td>\n",
       "      <td>0.568829</td>\n",
       "      <td>0.039509</td>\n",
       "      <td>403.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>KANAMYCIN</td>\n",
       "      <td>0.632181</td>\n",
       "      <td>0.040279</td>\n",
       "      <td>236.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MOXIFLOXACIN</td>\n",
       "      <td>0.492332</td>\n",
       "      <td>0.064168</td>\n",
       "      <td>50.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PYRAZINAMIDE</td>\n",
       "      <td>0.646346</td>\n",
       "      <td>0.052279</td>\n",
       "      <td>371.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>RIFAMPICIN</td>\n",
       "      <td>0.526072</td>\n",
       "      <td>0.036466</td>\n",
       "      <td>402.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>STREPTOMYCIN</td>\n",
       "      <td>0.478958</td>\n",
       "      <td>0.036820</td>\n",
       "      <td>599.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           drug       pve    se_pve  N_snps\n",
       "0      AMIKACIN  0.565059  0.040851   268.0\n",
       "1   CAPREOMYCIN  0.355128  0.050053   206.0\n",
       "2    ETHAMBUTOL  0.468603  0.038971   550.0\n",
       "3   ETHIONAMIDE  0.230460  0.045134   145.0\n",
       "4     ISONIAZID  0.568829  0.039509   403.0\n",
       "5     KANAMYCIN  0.632181  0.040279   236.0\n",
       "6  MOXIFLOXACIN  0.492332  0.064168    50.0\n",
       "7  PYRAZINAMIDE  0.646346  0.052279   371.0\n",
       "8    RIFAMPICIN  0.526072  0.036466   402.0\n",
       "9  STREPTOMYCIN  0.478958  0.036820   599.0"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table = []\n",
    "for drug in drugs:\n",
    "    print(drug)\n",
    "    \n",
    "    gemma_executable = kwargs[\"gemma\"]\n",
    "    pheno_file = f\"output/{drug}/{drug}.phenotypes\"\n",
    "    relatedness_loci_file = f\"output/{drug}/0.input_loci_files/ABR_sites_plus_all_dependent.loci\"\n",
    "    cxx_file = f\"output/{drug}/ABR_with_ABRpairs.cXX.txt\"\n",
    "    outfile = \"heritability_temp.txt\"\n",
    "    #string = f\"{gemma_executable} -g {relatedness_loci_file}  -k {cxx_file} -p {pheno_file} -maf 0 -miss 0.2 -lmm -notsnp -o {outfile}\"\n",
    "    string = f\"{gemma_executable} -g {relatedness_loci_file} -p {pheno_file} -k {cxx_file} -maf 0 -miss 0.2 -lmm -notsnp -o {outfile}\"\n",
    "    print(string)\n",
    "    process = string.split(\" \")\n",
    "    result = subprocess.run(process, stdout=subprocess.PIPE)\n",
    "\n",
    "    stdout = result.stdout.decode('utf-8').split(\"\\n\")\n",
    "    print(stdout)\n",
    "    \n",
    "    print(stdout)\n",
    "    if len(stdout) > 8:\n",
    "        out3 = stdout[7]\n",
    "        out1 = stdout[9]\n",
    "        out2 = stdout[10]\n",
    "        assert 'pve estimate' in out1\n",
    "        assert 'se(pve)' in out2\n",
    "        assert 'number of analyzed SNPs' in out3\n",
    "        pve = float(out1.split(\"=\")[-1])\n",
    "        se = float(out2.split(\"=\")[-1])\n",
    "        N_snps = float(out3.split(\"=\")[-1])\n",
    "        table.append([drug, pve, se, N_snps])\n",
    "    \n",
    "    print(pve, se)\n",
    "    \n",
    "df = pd.DataFrame(table, columns=[\"drug\", \"pve\", \"se_pve\", \"N_snps\"])\n",
    "    \n",
    "df.to_csv(\"PVE_ABR_dep_sites.csv\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Use the antibiotic potentiating hits too"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "annotation for N Snps: 782565\n",
      "AMIKACIN AMI\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘output/AMIKACIN’: File exists\n",
      "mkdir: cannot create directory ‘output/AMIKACIN/0.input_loci_files’: File exists\n",
      "mkdir: cannot create directory ‘output/AMIKACIN/1.job_submission_files’: File exists\n",
      "mkdir: cannot create directory ‘output/AMIKACIN/2.run_output’: File exists\n",
      "/home/ag300/anaconda3/envs/jupytervenv2/lib/python3.7/site-packages/ipykernel_launcher.py:60: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1205\n",
      "1206\n",
      "1205\n",
      "1206\n",
      "Number of isolates 1206\n",
      "Number of SNPS 353\n",
      "encoding gaps with gap code nan\n",
      "<class 'numpy.ndarray'>\n",
      "shape of the genotype matrix (1206, 353)\n",
      "newly created cross matrix dimensions (1206, 793)\n",
      "Creating locus file for AMIKACIN with N positions: 793\n",
      "Made genotype matrix with shape (1206, 793)\n",
      "writing Gemma with gap code NA\n",
      "calculating reference allele matrix with gap code NA\n",
      "encoding gaps with gap code NA\n",
      "<class 'numpy.ndarray'>\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/AMIKACIN/0.input_loci_files/ABR_potentiator_sites_plus_all_dependent.loci -p output/AMIKACIN/AMIKACIN.phenotypes -gk -maf 0 -o AMIKACIN/ABR_with_ABRpairs_potentiator\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/AMIKACIN/0.input_loci_files/ABR_potentiator_sites_plus_all_dependent.loci -p output/AMIKACIN/AMIKACIN.phenotypes -k output/AMIKACIN/ABR_with_ABRpairs_potentiator.cXX.txt -maf 0 -eigen -notsnp -o AMIKACIN/ABR_with_ABRpairs_potentiator\n",
      "creating GRM\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1206\n",
      "## number of analyzed individuals = 1206\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      793\n",
      "## number of analyzed SNPs         =      778\n",
      "Calculating Relatedness Matrix ... \n",
      "================================================== 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running eigendecomp\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1206\n",
      "## number of analyzed individuals = 1206\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      793\n",
      "## number of analyzed SNPs         =      778\n",
      "Start Eigen-Decomposition...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 712 eigenvalues close to zero\n",
      "**** INFO: Done.\n",
      "mkdir: cannot create directory ‘output/CAPREOMYCIN’: File exists\n",
      "mkdir: cannot create directory ‘output/CAPREOMYCIN/0.input_loci_files’: File exists\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAPREOMYCIN CAP\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘output/CAPREOMYCIN/1.job_submission_files’: File exists\n",
      "mkdir: cannot create directory ‘output/CAPREOMYCIN/2.run_output’: File exists\n",
      "/home/ag300/anaconda3/envs/jupytervenv2/lib/python3.7/site-packages/ipykernel_launcher.py:60: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1101\n",
      "1102\n",
      "1101\n",
      "1102\n",
      "Number of isolates 1102\n",
      "Number of SNPS 334\n",
      "encoding gaps with gap code nan\n",
      "<class 'numpy.ndarray'>\n",
      "shape of the genotype matrix (1102, 334)\n",
      "newly created cross matrix dimensions (1102, 639)\n",
      "Creating locus file for CAPREOMYCIN with N positions: 639\n",
      "Made genotype matrix with shape (1102, 639)\n",
      "writing Gemma with gap code NA\n",
      "calculating reference allele matrix with gap code NA\n",
      "encoding gaps with gap code NA\n",
      "<class 'numpy.ndarray'>\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/CAPREOMYCIN/0.input_loci_files/ABR_potentiator_sites_plus_all_dependent.loci -p output/CAPREOMYCIN/CAPREOMYCIN.phenotypes -gk -maf 0 -o CAPREOMYCIN/ABR_with_ABRpairs_potentiator\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/CAPREOMYCIN/0.input_loci_files/ABR_potentiator_sites_plus_all_dependent.loci -p output/CAPREOMYCIN/CAPREOMYCIN.phenotypes -k output/CAPREOMYCIN/ABR_with_ABRpairs_potentiator.cXX.txt -maf 0 -eigen -notsnp -o CAPREOMYCIN/ABR_with_ABRpairs_potentiator\n",
      "creating GRM\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1102\n",
      "## number of analyzed individuals = 1102\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      639\n",
      "## number of analyzed SNPs         =      620\n",
      "Calculating Relatedness Matrix ... \n",
      "================================================== 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running eigendecomp\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1102\n",
      "## number of analyzed individuals = 1102\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      639\n",
      "## number of analyzed SNPs         =      620\n",
      "Start Eigen-Decomposition...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 688 eigenvalues close to zero\n",
      "**** INFO: Done.\n",
      "mkdir: cannot create directory ‘output/ETHAMBUTOL’: File exists\n",
      "mkdir: cannot create directory ‘output/ETHAMBUTOL/0.input_loci_files’: File exists\n",
      "mkdir: cannot create directory ‘output/ETHAMBUTOL/1.job_submission_files’: File exists\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ETHAMBUTOL EMB\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘output/ETHAMBUTOL/2.run_output’: File exists\n",
      "/home/ag300/anaconda3/envs/jupytervenv2/lib/python3.7/site-packages/ipykernel_launcher.py:60: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1549\n",
      "1488\n",
      "1549\n",
      "1488\n",
      "Number of isolates 1488\n",
      "Number of SNPS 366\n",
      "encoding gaps with gap code nan\n",
      "<class 'numpy.ndarray'>\n",
      "shape of the genotype matrix (1488, 366)\n",
      "newly created cross matrix dimensions (1488, 896)\n",
      "Creating locus file for ETHAMBUTOL with N positions: 896\n",
      "Made genotype matrix with shape (1488, 896)\n",
      "writing Gemma with gap code NA\n",
      "calculating reference allele matrix with gap code NA\n",
      "encoding gaps with gap code NA\n",
      "<class 'numpy.ndarray'>\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/ETHAMBUTOL/0.input_loci_files/ABR_potentiator_sites_plus_all_dependent.loci -p output/ETHAMBUTOL/ETHAMBUTOL.phenotypes -gk -maf 0 -o ETHAMBUTOL/ABR_with_ABRpairs_potentiator\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/ETHAMBUTOL/0.input_loci_files/ABR_potentiator_sites_plus_all_dependent.loci -p output/ETHAMBUTOL/ETHAMBUTOL.phenotypes -k output/ETHAMBUTOL/ABR_with_ABRpairs_potentiator.cXX.txt -maf 0 -eigen -notsnp -o ETHAMBUTOL/ABR_with_ABRpairs_potentiator\n",
      "creating GRM\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1488\n",
      "## number of analyzed individuals = 1488\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      896\n",
      "## number of analyzed SNPs         =      885\n",
      "Calculating Relatedness Matrix ... \n",
      "================================================== 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running eigendecomp\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1488\n",
      "## number of analyzed individuals = 1488\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      896\n",
      "## number of analyzed SNPs         =      885\n",
      "Start Eigen-Decomposition...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 865 eigenvalues close to zero\n",
      "**** INFO: Done.\n",
      "mkdir: cannot create directory ‘output/ETHIONAMIDE’: File exists\n",
      "mkdir: cannot create directory ‘output/ETHIONAMIDE/0.input_loci_files’: File exists\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ETHIONAMIDE ETA\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘output/ETHIONAMIDE/1.job_submission_files’: File exists\n",
      "mkdir: cannot create directory ‘output/ETHIONAMIDE/2.run_output’: File exists\n",
      "/home/ag300/anaconda3/envs/jupytervenv2/lib/python3.7/site-packages/ipykernel_launcher.py:60: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1212\n",
      "1213\n",
      "1212\n",
      "1213\n",
      "Number of isolates 1213\n",
      "Number of SNPS 174\n",
      "encoding gaps with gap code nan\n",
      "<class 'numpy.ndarray'>\n",
      "shape of the genotype matrix (1213, 174)\n",
      "newly created cross matrix dimensions (1213, 391)\n",
      "Creating locus file for ETHIONAMIDE with N positions: 391\n",
      "Made genotype matrix with shape (1213, 391)\n",
      "writing Gemma with gap code NA\n",
      "calculating reference allele matrix with gap code NA\n",
      "encoding gaps with gap code NA\n",
      "<class 'numpy.ndarray'>\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/ETHIONAMIDE/0.input_loci_files/ABR_potentiator_sites_plus_all_dependent.loci -p output/ETHIONAMIDE/ETHIONAMIDE.phenotypes -gk -maf 0 -o ETHIONAMIDE/ABR_with_ABRpairs_potentiator\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/ETHIONAMIDE/0.input_loci_files/ABR_potentiator_sites_plus_all_dependent.loci -p output/ETHIONAMIDE/ETHIONAMIDE.phenotypes -k output/ETHIONAMIDE/ABR_with_ABRpairs_potentiator.cXX.txt -maf 0 -eigen -notsnp -o ETHIONAMIDE/ABR_with_ABRpairs_potentiator\n",
      "creating GRM\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1213\n",
      "## number of analyzed individuals = 1213\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      391\n",
      "## number of analyzed SNPs         =      386\n",
      "Calculating Relatedness Matrix ... \n",
      "================================================== 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running eigendecomp\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1213\n",
      "## number of analyzed individuals = 1213\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      391\n",
      "## number of analyzed SNPs         =      386\n",
      "Start Eigen-Decomposition...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 958 eigenvalues close to zero\n",
      "**** INFO: Done.\n",
      "mkdir: cannot create directory ‘output/ISONIAZID’: File exists\n",
      "mkdir: cannot create directory ‘output/ISONIAZID/0.input_loci_files’: File exists\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ISONIAZID INH\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘output/ISONIAZID/1.job_submission_files’: File exists\n",
      "mkdir: cannot create directory ‘output/ISONIAZID/2.run_output’: File exists\n",
      "/home/ag300/anaconda3/envs/jupytervenv2/lib/python3.7/site-packages/ipykernel_launcher.py:60: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1554\n",
      "1497\n",
      "1554\n",
      "1497\n",
      "Number of isolates 1497\n",
      "Number of SNPS 329\n",
      "encoding gaps with gap code nan\n",
      "<class 'numpy.ndarray'>\n",
      "shape of the genotype matrix (1497, 329)\n",
      "newly created cross matrix dimensions (1497, 629)\n",
      "Creating locus file for ISONIAZID with N positions: 629\n",
      "Made genotype matrix with shape (1497, 629)\n",
      "writing Gemma with gap code NA\n",
      "calculating reference allele matrix with gap code NA\n",
      "encoding gaps with gap code NA\n",
      "<class 'numpy.ndarray'>\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/ISONIAZID/0.input_loci_files/ABR_potentiator_sites_plus_all_dependent.loci -p output/ISONIAZID/ISONIAZID.phenotypes -gk -maf 0 -o ISONIAZID/ABR_with_ABRpairs_potentiator\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/ISONIAZID/0.input_loci_files/ABR_potentiator_sites_plus_all_dependent.loci -p output/ISONIAZID/ISONIAZID.phenotypes -k output/ISONIAZID/ABR_with_ABRpairs_potentiator.cXX.txt -maf 0 -eigen -notsnp -o ISONIAZID/ABR_with_ABRpairs_potentiator\n",
      "creating GRM\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1497\n",
      "## number of analyzed individuals = 1497\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      629\n",
      "## number of analyzed SNPs         =      620\n",
      "Calculating Relatedness Matrix ... \n",
      "================================================== 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running eigendecomp\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1497\n",
      "## number of analyzed individuals = 1497\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      629\n",
      "## number of analyzed SNPs         =      620\n",
      "Start Eigen-Decomposition...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 1034 eigenvalues close to zero\n",
      "**** INFO: Done.\n",
      "mkdir: cannot create directory ‘output/KANAMYCIN’: File exists\n",
      "mkdir: cannot create directory ‘output/KANAMYCIN/0.input_loci_files’: File exists\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KANAMYCIN KAN\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘output/KANAMYCIN/1.job_submission_files’: File exists\n",
      "mkdir: cannot create directory ‘output/KANAMYCIN/2.run_output’: File exists\n",
      "/home/ag300/anaconda3/envs/jupytervenv2/lib/python3.7/site-packages/ipykernel_launcher.py:60: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1214\n",
      "1215\n",
      "1214\n",
      "1215\n",
      "Number of isolates 1215\n",
      "Number of SNPS 320\n",
      "encoding gaps with gap code nan\n",
      "<class 'numpy.ndarray'>\n",
      "shape of the genotype matrix (1215, 320)\n",
      "newly created cross matrix dimensions (1215, 756)\n",
      "Creating locus file for KANAMYCIN with N positions: 756\n",
      "Made genotype matrix with shape (1215, 756)\n",
      "writing Gemma with gap code NA\n",
      "calculating reference allele matrix with gap code NA\n",
      "encoding gaps with gap code NA\n",
      "<class 'numpy.ndarray'>\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/KANAMYCIN/0.input_loci_files/ABR_potentiator_sites_plus_all_dependent.loci -p output/KANAMYCIN/KANAMYCIN.phenotypes -gk -maf 0 -o KANAMYCIN/ABR_with_ABRpairs_potentiator\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/KANAMYCIN/0.input_loci_files/ABR_potentiator_sites_plus_all_dependent.loci -p output/KANAMYCIN/KANAMYCIN.phenotypes -k output/KANAMYCIN/ABR_with_ABRpairs_potentiator.cXX.txt -maf 0 -eigen -notsnp -o KANAMYCIN/ABR_with_ABRpairs_potentiator\n",
      "creating GRM\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1215\n",
      "## number of analyzed individuals = 1215\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      756\n",
      "## number of analyzed SNPs         =      744\n",
      "Calculating Relatedness Matrix ... \n",
      "================================================== 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running eigendecomp\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1215\n",
      "## number of analyzed individuals = 1215\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      756\n",
      "## number of analyzed SNPs         =      744\n",
      "Start Eigen-Decomposition...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 755 eigenvalues close to zero\n",
      "**** INFO: Done.\n",
      "mkdir: cannot create directory ‘output/MOXIFLOXACIN’: File exists\n",
      "mkdir: cannot create directory ‘output/MOXIFLOXACIN/0.input_loci_files’: File exists\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MOXIFLOXACIN MOXI\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘output/MOXIFLOXACIN/1.job_submission_files’: File exists\n",
      "mkdir: cannot create directory ‘output/MOXIFLOXACIN/2.run_output’: File exists\n",
      "/home/ag300/anaconda3/envs/jupytervenv2/lib/python3.7/site-packages/ipykernel_launcher.py:60: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1089\n",
      "1090\n",
      "1089\n",
      "1090\n",
      "Number of isolates 1090\n",
      "Number of SNPS 171\n",
      "encoding gaps with gap code nan\n",
      "<class 'numpy.ndarray'>\n",
      "shape of the genotype matrix (1090, 171)\n",
      "newly created cross matrix dimensions (1090, 425)\n",
      "Creating locus file for MOXIFLOXACIN with N positions: 425\n",
      "Made genotype matrix with shape (1090, 425)\n",
      "writing Gemma with gap code NA\n",
      "calculating reference allele matrix with gap code NA\n",
      "encoding gaps with gap code NA\n",
      "<class 'numpy.ndarray'>\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/MOXIFLOXACIN/0.input_loci_files/ABR_potentiator_sites_plus_all_dependent.loci -p output/MOXIFLOXACIN/MOXIFLOXACIN.phenotypes -gk -maf 0 -o MOXIFLOXACIN/ABR_with_ABRpairs_potentiator\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/MOXIFLOXACIN/0.input_loci_files/ABR_potentiator_sites_plus_all_dependent.loci -p output/MOXIFLOXACIN/MOXIFLOXACIN.phenotypes -k output/MOXIFLOXACIN/ABR_with_ABRpairs_potentiator.cXX.txt -maf 0 -eigen -notsnp -o MOXIFLOXACIN/ABR_with_ABRpairs_potentiator\n",
      "creating GRM\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1090\n",
      "## number of analyzed individuals = 1090\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      425\n",
      "## number of analyzed SNPs         =      415\n",
      "Calculating Relatedness Matrix ... \n",
      "================================================== 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running eigendecomp\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1090\n",
      "## number of analyzed individuals = 1090\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      425\n",
      "## number of analyzed SNPs         =      415\n",
      "Start Eigen-Decomposition...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 806 eigenvalues close to zero\n",
      "**** INFO: Done.\n",
      "mkdir: cannot create directory ‘output/PYRAZINAMIDE’: File exists\n",
      "mkdir: cannot create directory ‘output/PYRAZINAMIDE/0.input_loci_files’: File exists\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PYRAZINAMIDE PZA\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘output/PYRAZINAMIDE/1.job_submission_files’: File exists\n",
      "mkdir: cannot create directory ‘output/PYRAZINAMIDE/2.run_output’: File exists\n",
      "/home/ag300/anaconda3/envs/jupytervenv2/lib/python3.7/site-packages/ipykernel_launcher.py:60: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1106\n",
      "1107\n",
      "1106\n",
      "1107\n",
      "Number of isolates 1107\n",
      "Number of SNPS 464\n",
      "encoding gaps with gap code nan\n",
      "<class 'numpy.ndarray'>\n",
      "shape of the genotype matrix (1107, 464)\n",
      "newly created cross matrix dimensions (1107, 1465)\n",
      "Creating locus file for PYRAZINAMIDE with N positions: 1465\n",
      "Made genotype matrix with shape (1107, 1465)\n",
      "writing Gemma with gap code NA\n",
      "calculating reference allele matrix with gap code NA\n",
      "encoding gaps with gap code NA\n",
      "<class 'numpy.ndarray'>\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/PYRAZINAMIDE/0.input_loci_files/ABR_potentiator_sites_plus_all_dependent.loci -p output/PYRAZINAMIDE/PYRAZINAMIDE.phenotypes -gk -maf 0 -o PYRAZINAMIDE/ABR_with_ABRpairs_potentiator\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/PYRAZINAMIDE/0.input_loci_files/ABR_potentiator_sites_plus_all_dependent.loci -p output/PYRAZINAMIDE/PYRAZINAMIDE.phenotypes -k output/PYRAZINAMIDE/ABR_with_ABRpairs_potentiator.cXX.txt -maf 0 -eigen -notsnp -o PYRAZINAMIDE/ABR_with_ABRpairs_potentiator\n",
      "creating GRM\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1107\n",
      "## number of analyzed individuals = 1107\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =     1465\n",
      "## number of analyzed SNPs         =     1438\n",
      "Calculating Relatedness Matrix ... \n",
      "================================================== 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running eigendecomp\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1107\n",
      "## number of analyzed individuals = 1107\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =     1465\n",
      "## number of analyzed SNPs         =     1438\n",
      "Start Eigen-Decomposition...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 553 eigenvalues close to zero\n",
      "**** INFO: Done.\n",
      "mkdir: cannot create directory ‘output/RIFAMPICIN’: File exists\n",
      "mkdir: cannot create directory ‘output/RIFAMPICIN/0.input_loci_files’: File exists\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RIFAMPICIN RIF\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘output/RIFAMPICIN/1.job_submission_files’: File exists\n",
      "mkdir: cannot create directory ‘output/RIFAMPICIN/2.run_output’: File exists\n",
      "/home/ag300/anaconda3/envs/jupytervenv2/lib/python3.7/site-packages/ipykernel_launcher.py:60: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1597\n",
      "1522\n",
      "1597\n",
      "1522\n",
      "Number of isolates 1522\n",
      "Number of SNPS 418\n",
      "encoding gaps with gap code nan\n",
      "<class 'numpy.ndarray'>\n",
      "shape of the genotype matrix (1522, 418)\n",
      "newly created cross matrix dimensions (1522, 996)\n",
      "Creating locus file for RIFAMPICIN with N positions: 996\n",
      "Made genotype matrix with shape (1522, 996)\n",
      "writing Gemma with gap code NA\n",
      "calculating reference allele matrix with gap code NA\n",
      "encoding gaps with gap code NA\n",
      "<class 'numpy.ndarray'>\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/RIFAMPICIN/0.input_loci_files/ABR_potentiator_sites_plus_all_dependent.loci -p output/RIFAMPICIN/RIFAMPICIN.phenotypes -gk -maf 0 -o RIFAMPICIN/ABR_with_ABRpairs_potentiator\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/RIFAMPICIN/0.input_loci_files/ABR_potentiator_sites_plus_all_dependent.loci -p output/RIFAMPICIN/RIFAMPICIN.phenotypes -k output/RIFAMPICIN/ABR_with_ABRpairs_potentiator.cXX.txt -maf 0 -eigen -notsnp -o RIFAMPICIN/ABR_with_ABRpairs_potentiator\n",
      "creating GRM\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1522\n",
      "## number of analyzed individuals = 1522\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      996\n",
      "## number of analyzed SNPs         =      993\n",
      "Calculating Relatedness Matrix ... \n",
      "================================================== 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running eigendecomp\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1522\n",
      "## number of analyzed individuals = 1522\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =      996\n",
      "## number of analyzed SNPs         =      993\n",
      "Start Eigen-Decomposition...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 841 eigenvalues close to zero\n",
      "**** INFO: Done.\n",
      "mkdir: cannot create directory ‘output/STREPTOMYCIN’: File exists\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STREPTOMYCIN STR\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘output/STREPTOMYCIN/0.input_loci_files’: File exists\n",
      "mkdir: cannot create directory ‘output/STREPTOMYCIN/1.job_submission_files’: File exists\n",
      "mkdir: cannot create directory ‘output/STREPTOMYCIN/2.run_output’: File exists\n",
      "/home/ag300/anaconda3/envs/jupytervenv2/lib/python3.7/site-packages/ipykernel_launcher.py:60: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1543\n",
      "1489\n",
      "1543\n",
      "1489\n",
      "Number of isolates 1489\n",
      "Number of SNPS 595\n",
      "encoding gaps with gap code nan\n",
      "<class 'numpy.ndarray'>\n",
      "shape of the genotype matrix (1489, 595)\n",
      "newly created cross matrix dimensions (1489, 1200)\n",
      "Creating locus file for STREPTOMYCIN with N positions: 1200\n",
      "Made genotype matrix with shape (1489, 1200)\n",
      "writing Gemma with gap code NA\n",
      "calculating reference allele matrix with gap code NA\n",
      "encoding gaps with gap code NA\n",
      "<class 'numpy.ndarray'>\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/STREPTOMYCIN/0.input_loci_files/ABR_potentiator_sites_plus_all_dependent.loci -p output/STREPTOMYCIN/STREPTOMYCIN.phenotypes -gk -maf 0 -o STREPTOMYCIN/ABR_with_ABRpairs_potentiator\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/STREPTOMYCIN/0.input_loci_files/ABR_potentiator_sites_plus_all_dependent.loci -p output/STREPTOMYCIN/STREPTOMYCIN.phenotypes -k output/STREPTOMYCIN/ABR_with_ABRpairs_potentiator.cXX.txt -maf 0 -eigen -notsnp -o STREPTOMYCIN/ABR_with_ABRpairs_potentiator\n",
      "creating GRM\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1489\n",
      "## number of analyzed individuals = 1489\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =     1200\n",
      "## number of analyzed SNPs         =     1192\n",
      "Calculating Relatedness Matrix ... \n",
      "================================================== 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running eigendecomp\n",
      "GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018\n",
      "Reading Files ... \n",
      "## number of total individuals = 1489\n",
      "## number of analyzed individuals = 1489\n",
      "## number of covariates = 1\n",
      "## number of phenotypes = 1\n",
      "## number of total SNPs/var        =     1200\n",
      "## number of analyzed SNPs         =     1192\n",
      "Start Eigen-Decomposition...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 625 eigenvalues close to zero\n",
      "**** INFO: Done.\n"
     ]
    }
   ],
   "source": [
    "full_range = pd.read_csv(\"../output/05.single_mutations/snps_with_all_annotation.csv\", index_col=0)\n",
    "full_range = full_range.query(\"in_possible_ABR_gene\")\n",
    "full_range[\"possible_drug\"] = [x.split(\",\") for x in full_range.possible_drug]\n",
    "\n",
    "# reads in the SNP annotations\n",
    "snp_annotation = pd.read_pickle(\"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_SNP_annotation.pkl\")\n",
    "snp_annotation['mat_index'] =snp_annotation.index\n",
    "print('annotation for N Snps:', len(snp_annotation))\n",
    "\n",
    "isolates = pd.read_pickle(kwargs['isolate_annotation_file'])\n",
    "#g = np.load(\"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_matrix.npy\")\n",
    "\n",
    "for drug, drug_abbrev in drug_to_abbrev.items():\n",
    "    \n",
    "    ## \n",
    "    _keep = []\n",
    "    for idx, row in full_range.iterrows():\n",
    "        if drug_abbrev in row.possible_drug:\n",
    "            _keep.append(idx)\n",
    "        elif drug_abbrev=='ETA':\n",
    "            if \"ETH\" in row.possible_drug:\n",
    "                _keep.append(idx)\n",
    "        elif drug_abbrev=='MOXI':\n",
    "            if \"MXF\" in row.possible_drug:\n",
    "                _keep.append(idx)\n",
    "        elif drug_abbrev=='STR':\n",
    "            if \"STM\" in row.possible_drug:\n",
    "                _keep.append(idx)\n",
    "                \n",
    "    positions_found = list(set(full_range.loc[_keep].pos))\n",
    "    \n",
    "    antibiotic_hits = pd.read_csv(\n",
    "        f\"/n/data1/hms/dbmi/farhat/anna/ABR_epistasis/DependentMutations2/output/04.antibiotic/{drug_abbrev}_antibiotic_hits.csv\"\n",
    "    )\n",
    "    antibiotic_hits1 = pd.read_csv(\n",
    "        f\"/n/data1/hms/dbmi/farhat/anna/ABR_epistasis/DependentMutations2/output/04.antibiotic/{drug_abbrev}_potentiator_antibiotic_hits.csv\"\n",
    "    )\n",
    "    antibiotic_hits = pd.concat([antibiotic_hits, antibiotic_hits1])\n",
    "    positions_found = positions_found + list(antibiotic_hits.position_j) + list(antibiotic_hits.position_i)\n",
    "    snps_subset = snp_annotation.query(\"pos in @positions_found\")\n",
    "    \n",
    "    additional_sites = antibiotic_hits\n",
    "    \n",
    "    print(drug, drug_abbrev)\n",
    "    # Generate the cXX file\n",
    "    \n",
    "    prefix = f\"output/{drug}\"\n",
    "    output_prefix = f\"{drug}\"\n",
    "    os.system(f\"mkdir {prefix}\")\n",
    "    os.system(f\"mkdir {prefix}/0.input_loci_files\")\n",
    "    os.system(f\"mkdir {prefix}/1.job_submission_files\")\n",
    "    os.system(f\"mkdir {prefix}/2.run_output\")\n",
    "\n",
    "    # get the isolates with phenotypes to keep\n",
    "    phenotypes = pd.read_csv(kwargs['phenotype_file'], index_col=0)\n",
    "    phenotypes.replace(\"NA\", np.nan, inplace=True)\n",
    "    \n",
    "    phenotypes_found_all = phenotypes.loc[~phenotypes[drug_abbrev].isnull(),:]\n",
    "    phenotypes_found = phenotypes_found_all[[\"ROLLINGDB_ID\", f\"{drug_abbrev}_midpoint\"]]\n",
    "    phenotypes_found[drug] = np.log(phenotypes_found[f\"{drug_abbrev}_midpoint\"])\n",
    "\n",
    "    isolates = pd.read_pickle(kwargs['isolate_annotation_file'])\n",
    "    isolates_with_phenotype = isolates.query(\"isolate_ID in @phenotypes_found.ROLLINGDB_ID\")\n",
    "    geno_mat = np.load(\"/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered_2/genotypes_matrix.npy\")\n",
    "\n",
    "    #### Prepare the phenotype input file ####\n",
    "    pheno_file = f\"{prefix}/{drug}.phenotypes\"\n",
    "    write_phenotype_file_MIC(pheno_file, phenotypes_found, isolates_with_phenotype, drug)\n",
    "\n",
    "    \n",
    "    matrix = geno_mat[snps_subset.index, :][:, isolates_with_phenotype.index]\n",
    "    \n",
    "    #### Prepare the phenotype input file ####\n",
    "    pheno_file = f\"{prefix}/{drug}.phenotypes\"\n",
    "    write_phenotype_file_MIC(pheno_file, phenotypes_found, isolates_with_phenotype, drug)\n",
    "\n",
    "    \n",
    "    matrix = geno_mat[snps_subset.index, :][:, isolates_with_phenotype.index]\n",
    "    \n",
    "    ## Add the MTB ancestor sequence\n",
    "    \n",
    "    # read the ancestor sequence\n",
    "    ancestral_seq_file=\"../input/MTB_ancestor_reference_modified.fasta\"\n",
    "    # subtract 1 or get wrecked. H37Rv positions are 1-indexed\n",
    "    all_positions = [int(x) - 1 for x in snps_subset.pos]\n",
    "    ancestral_seq = Alignment.from_file(open(ancestral_seq_file, \"r\"))\n",
    "    ancestral_seq_subset = np.array([ancestral_seq.matrix[0,x] for x in all_positions])\n",
    "    \n",
    "    # convert to matrix encoding\n",
    "    ancestral_seq_subset = np.array([genotype_translator_r[x] for x in list(ancestral_seq_subset)])\n",
    "    \n",
    "    # concatenate as FIRST sequence\n",
    "    matrix = np.concatenate([ancestral_seq_subset.reshape(-1,1), matrix], axis=1)\n",
    "    \n",
    "    isolates = [\"MTB_ANC\"] + list(isolates_with_phenotype.isolate_ID)\n",
    "\n",
    "    # add to strain list\n",
    "    \n",
    "    gm = GenotypeMatrix(\n",
    "        position_list = snps_subset.pos,\n",
    "        strain_list = isolates,\n",
    "        matrix_data=matrix.T\n",
    "    )\n",
    "    \n",
    "    print(\"Number of isolates\", len(isolates))\n",
    "    print(\"Number of SNPS\", len(snps_subset))\n",
    "    ## Convert to major allele matrix\n",
    "    geno_mat = calculate_reference_allele_matrix_for_pair_regression(gm.matrix, ancestral_seq_subset) \n",
    "    print(\"shape of the genotype matrix\", geno_mat.shape)\n",
    "    snps_in_analysis = list(snps_subset.pos)\n",
    "    ### To the matrix, add the pairs of SNPs\n",
    "    \n",
    "    matrix = deepcopy(geno_mat)\n",
    "    for idx, row in additional_sites.iterrows():\n",
    "        # lookup the index of the two positions\n",
    "        pos_i = row.position_i\n",
    "        pos_j = row.position_j\n",
    "        \n",
    "        idx_i = gm.position_to_index[pos_i]\n",
    "        idx_j = gm.position_to_index[pos_j]\n",
    "        \n",
    "        new_row = np.multiply(geno_mat[:,idx_i], geno_mat[:, idx_j]).T\n",
    "        matrix = np.concatenate([matrix, new_row.reshape(-1,1)], axis=1)\n",
    "        \n",
    "        snps_in_analysis.append(f\"{pos_i}_{pos_j}\")\n",
    "\n",
    "# #     # Generate the locus file\n",
    "    print(\"newly created cross matrix dimensions\", matrix.shape)\n",
    "    relatedness_loci_file = f\"{prefix}/0.input_loci_files/ABR_potentiator_sites_plus_all_dependent.loci\"\n",
    "\n",
    "    print(f\"Creating locus file for {drug} with N positions: {len(snps_in_analysis)}\")\n",
    "\n",
    "    prepare_loci_file(matrix, isolates, snps_in_analysis, relatedness_loci_file)\n",
    "\n",
    "    cxx_file = f\"{prefix}/ABR_with_ABRpairs_potentiator.cXX.txt\"\n",
    "\n",
    "    outfile = f\"{output_prefix}/ABR_with_ABRpairs_potentiator\"\n",
    "    string1 = f\"{kwargs['gemma']} -g {relatedness_loci_file} -p {pheno_file} -gk -maf 0 -o {outfile}\"\n",
    "    print(string1)\n",
    "\n",
    "    # Run the eigen decomposition\n",
    "    outfile =f'{output_prefix}/ABR_with_ABRpairs_potentiator'\n",
    "    string2 = f\"{kwargs['gemma']} -g {relatedness_loci_file} -p {pheno_file} -k {cxx_file} -maf 0 -eigen -notsnp -o {outfile}\"\n",
    "    print(string2)\n",
    "\n",
    "    print(\"creating GRM\")\n",
    "    os.system(string1)\n",
    "\n",
    "    print(\"running eigendecomp\")\n",
    "    os.system(string2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AMIKACIN\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/AMIKACIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/AMIKACIN/AMIKACIN.phenotypes -k output/AMIKACIN/ABR_with_ABRpairs_potentiator.cXX.txt -maf 0 -miss 0.2 -lmm -notsnp -o heritability_temp.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 712 eigenvalues close to zero\n",
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1206', '## number of analyzed individuals = 1206', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      793', '## number of analyzed SNPs         =      778', 'Start Eigen-Decomposition...', 'pve estimate =0.537309', 'se(pve) =0.0410739', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1206', '## number of analyzed individuals = 1206', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      793', '## number of analyzed SNPs         =      778', 'Start Eigen-Decomposition...', 'pve estimate =0.537309', 'se(pve) =0.0410739', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "0.537309 0.0410739\n",
      "CAPREOMYCIN\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/CAPREOMYCIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/CAPREOMYCIN/CAPREOMYCIN.phenotypes -k output/CAPREOMYCIN/ABR_with_ABRpairs_potentiator.cXX.txt -maf 0 -miss 0.2 -lmm -notsnp -o heritability_temp.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 688 eigenvalues close to zero\n",
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1102', '## number of analyzed individuals = 1102', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      639', '## number of analyzed SNPs         =      620', 'Start Eigen-Decomposition...', 'pve estimate =0.407109', 'se(pve) =0.0529406', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1102', '## number of analyzed individuals = 1102', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      639', '## number of analyzed SNPs         =      620', 'Start Eigen-Decomposition...', 'pve estimate =0.407109', 'se(pve) =0.0529406', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "0.407109 0.0529406\n",
      "ETHAMBUTOL\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/ETHAMBUTOL/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/ETHAMBUTOL/ETHAMBUTOL.phenotypes -k output/ETHAMBUTOL/ABR_with_ABRpairs_potentiator.cXX.txt -maf 0 -miss 0.2 -lmm -notsnp -o heritability_temp.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 865 eigenvalues close to zero\n",
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1488', '## number of analyzed individuals = 1488', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      896', '## number of analyzed SNPs         =      885', 'Start Eigen-Decomposition...', 'pve estimate =0.453898', 'se(pve) =0.0439552', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1488', '## number of analyzed individuals = 1488', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      896', '## number of analyzed SNPs         =      885', 'Start Eigen-Decomposition...', 'pve estimate =0.453898', 'se(pve) =0.0439552', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "0.453898 0.0439552\n",
      "ETHIONAMIDE\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/ETHIONAMIDE/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/ETHIONAMIDE/ETHIONAMIDE.phenotypes -k output/ETHIONAMIDE/ABR_with_ABRpairs_potentiator.cXX.txt -maf 0 -miss 0.2 -lmm -notsnp -o heritability_temp.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 958 eigenvalues close to zero\n",
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1213', '## number of analyzed individuals = 1213', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      391', '## number of analyzed SNPs         =      386', 'Start Eigen-Decomposition...', 'pve estimate =0.471663', 'se(pve) =0.055662', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1213', '## number of analyzed individuals = 1213', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      391', '## number of analyzed SNPs         =      386', 'Start Eigen-Decomposition...', 'pve estimate =0.471663', 'se(pve) =0.055662', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "0.471663 0.055662\n",
      "ISONIAZID\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/ISONIAZID/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/ISONIAZID/ISONIAZID.phenotypes -k output/ISONIAZID/ABR_with_ABRpairs_potentiator.cXX.txt -maf 0 -miss 0.2 -lmm -notsnp -o heritability_temp.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 1034 eigenvalues close to zero\n",
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1497', '## number of analyzed individuals = 1497', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      629', '## number of analyzed SNPs         =      620', 'Start Eigen-Decomposition...', 'pve estimate =0.67816', 'se(pve) =0.0336789', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1497', '## number of analyzed individuals = 1497', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      629', '## number of analyzed SNPs         =      620', 'Start Eigen-Decomposition...', 'pve estimate =0.67816', 'se(pve) =0.0336789', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "0.67816 0.0336789\n",
      "KANAMYCIN\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/KANAMYCIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/KANAMYCIN/KANAMYCIN.phenotypes -k output/KANAMYCIN/ABR_with_ABRpairs_potentiator.cXX.txt -maf 0 -miss 0.2 -lmm -notsnp -o heritability_temp.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 755 eigenvalues close to zero\n",
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1215', '## number of analyzed individuals = 1215', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      756', '## number of analyzed SNPs         =      744', 'Start Eigen-Decomposition...', 'pve estimate =0.621469', 'se(pve) =0.0406408', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1215', '## number of analyzed individuals = 1215', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      756', '## number of analyzed SNPs         =      744', 'Start Eigen-Decomposition...', 'pve estimate =0.621469', 'se(pve) =0.0406408', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "0.621469 0.0406408\n",
      "MOXIFLOXACIN\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/MOXIFLOXACIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/MOXIFLOXACIN/MOXIFLOXACIN.phenotypes -k output/MOXIFLOXACIN/ABR_with_ABRpairs_potentiator.cXX.txt -maf 0 -miss 0.2 -lmm -notsnp -o heritability_temp.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 806 eigenvalues close to zero\n",
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1090', '## number of analyzed individuals = 1090', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =       55', '## number of analyzed SNPs         =       50', 'Start Eigen-Decomposition...', 'pve estimate =0.598275', 'se(pve) =0.0508725', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1090', '## number of analyzed individuals = 1090', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =       55', '## number of analyzed SNPs         =       50', 'Start Eigen-Decomposition...', 'pve estimate =0.598275', 'se(pve) =0.0508725', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "0.598275 0.0508725\n",
      "PYRAZINAMIDE\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/PYRAZINAMIDE/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/PYRAZINAMIDE/PYRAZINAMIDE.phenotypes -k output/PYRAZINAMIDE/ABR_with_ABRpairs_potentiator.cXX.txt -maf 0 -miss 0.2 -lmm -notsnp -o heritability_temp.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 553 eigenvalues close to zero\n",
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1107', '## number of analyzed individuals = 1107', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      388', '## number of analyzed SNPs         =      371', 'Start Eigen-Decomposition...', 'pve estimate =0.622346', 'se(pve) =0.0547998', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1107', '## number of analyzed individuals = 1107', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      388', '## number of analyzed SNPs         =      371', 'Start Eigen-Decomposition...', 'pve estimate =0.622346', 'se(pve) =0.0547998', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "0.622346 0.0547998\n",
      "RIFAMPICIN\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/RIFAMPICIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/RIFAMPICIN/RIFAMPICIN.phenotypes -k output/RIFAMPICIN/ABR_with_ABRpairs_potentiator.cXX.txt -maf 0 -miss 0.2 -lmm -notsnp -o heritability_temp.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 841 eigenvalues close to zero\n",
      "**** INFO: Done.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1522', '## number of analyzed individuals = 1522', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      405', '## number of analyzed SNPs         =      402', 'Start Eigen-Decomposition...', 'pve estimate =0.588582', 'se(pve) =0.0380766', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1522', '## number of analyzed individuals = 1522', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      405', '## number of analyzed SNPs         =      402', 'Start Eigen-Decomposition...', 'pve estimate =0.588582', 'se(pve) =0.0380766', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "0.588582 0.0380766\n",
      "STREPTOMYCIN\n",
      "/n/data1/hms/dbmi/farhat/anna/software/gemma-0.98.1-linux-static -g output/STREPTOMYCIN/0.input_loci_files/ABR_sites_plus_all_dependent.loci -p output/STREPTOMYCIN/STREPTOMYCIN.phenotypes -k output/STREPTOMYCIN/ABR_with_ABRpairs_potentiator.cXX.txt -maf 0 -miss 0.2 -lmm -notsnp -o heritability_temp.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** WARNING: Matrix G has 625 eigenvalues close to zero\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1489', '## number of analyzed individuals = 1489', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      607', '## number of analyzed SNPs         =      599', 'Start Eigen-Decomposition...', 'pve estimate =0.406087', 'se(pve) =0.041125', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "['GEMMA 0.98.1 (2018-12-10) by Xiang Zhou and team (C) 2012-2018', 'Reading Files ... ', '## number of total individuals = 1489', '## number of analyzed individuals = 1489', '## number of covariates = 1', '## number of phenotypes = 1', '## number of total SNPs/var        =      607', '## number of analyzed SNPs         =      599', 'Start Eigen-Decomposition...', 'pve estimate =0.406087', 'se(pve) =0.041125', '                                                   0%\\r================================================== 100%\\r================================================== 100%\\r', '']\n",
      "0.406087 0.041125\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "**** INFO: Done.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug</th>\n",
       "      <th>pve</th>\n",
       "      <th>se_pve</th>\n",
       "      <th>N_snps</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMIKACIN</td>\n",
       "      <td>0.537309</td>\n",
       "      <td>0.041074</td>\n",
       "      <td>778.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAPREOMYCIN</td>\n",
       "      <td>0.407109</td>\n",
       "      <td>0.052941</td>\n",
       "      <td>620.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ETHAMBUTOL</td>\n",
       "      <td>0.453898</td>\n",
       "      <td>0.043955</td>\n",
       "      <td>885.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ETHIONAMIDE</td>\n",
       "      <td>0.471663</td>\n",
       "      <td>0.055662</td>\n",
       "      <td>386.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ISONIAZID</td>\n",
       "      <td>0.678160</td>\n",
       "      <td>0.033679</td>\n",
       "      <td>620.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>KANAMYCIN</td>\n",
       "      <td>0.621469</td>\n",
       "      <td>0.040641</td>\n",
       "      <td>744.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MOXIFLOXACIN</td>\n",
       "      <td>0.598275</td>\n",
       "      <td>0.050873</td>\n",
       "      <td>50.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PYRAZINAMIDE</td>\n",
       "      <td>0.622346</td>\n",
       "      <td>0.054800</td>\n",
       "      <td>371.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>RIFAMPICIN</td>\n",
       "      <td>0.588582</td>\n",
       "      <td>0.038077</td>\n",
       "      <td>402.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>STREPTOMYCIN</td>\n",
       "      <td>0.406087</td>\n",
       "      <td>0.041125</td>\n",
       "      <td>599.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           drug       pve    se_pve  N_snps\n",
       "0      AMIKACIN  0.537309  0.041074   778.0\n",
       "1   CAPREOMYCIN  0.407109  0.052941   620.0\n",
       "2    ETHAMBUTOL  0.453898  0.043955   885.0\n",
       "3   ETHIONAMIDE  0.471663  0.055662   386.0\n",
       "4     ISONIAZID  0.678160  0.033679   620.0\n",
       "5     KANAMYCIN  0.621469  0.040641   744.0\n",
       "6  MOXIFLOXACIN  0.598275  0.050873    50.0\n",
       "7  PYRAZINAMIDE  0.622346  0.054800   371.0\n",
       "8    RIFAMPICIN  0.588582  0.038077   402.0\n",
       "9  STREPTOMYCIN  0.406087  0.041125   599.0"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table = []\n",
    "for drug in drugs:\n",
    "    print(drug)\n",
    "    \n",
    "    gemma_executable = kwargs[\"gemma\"]\n",
    "    pheno_file = f\"output/{drug}/{drug}.phenotypes\"\n",
    "    relatedness_loci_file = f\"output/{drug}/0.input_loci_files/ABR_sites_plus_all_dependent.loci\"\n",
    "    cxx_file = f\"output/{drug}/ABR_with_ABRpairs_potentiator.cXX.txt\"\n",
    "    outfile = \"heritability_temp.txt\"\n",
    "    #string = f\"{gemma_executable} -g {relatedness_loci_file}  -k {cxx_file} -p {pheno_file} -maf 0 -miss 0.2 -lmm -notsnp -o {outfile}\"\n",
    "    string = f\"{gemma_executable} -g {relatedness_loci_file} -p {pheno_file} -k {cxx_file} -maf 0 -miss 0.2 -lmm -notsnp -o {outfile}\"\n",
    "    print(string)\n",
    "    process = string.split(\" \")\n",
    "    result = subprocess.run(process, stdout=subprocess.PIPE)\n",
    "\n",
    "    stdout = result.stdout.decode('utf-8').split(\"\\n\")\n",
    "    print(stdout)\n",
    "    \n",
    "    print(stdout)\n",
    "    if len(stdout) > 8:\n",
    "        out3 = stdout[7]\n",
    "        out1 = stdout[9]\n",
    "        out2 = stdout[10]\n",
    "        assert 'pve estimate' in out1\n",
    "        assert 'se(pve)' in out2\n",
    "        assert 'number of analyzed SNPs' in out3\n",
    "        pve = float(out1.split(\"=\")[-1])\n",
    "        se = float(out2.split(\"=\")[-1])\n",
    "        N_snps = float(out3.split(\"=\")[-1])\n",
    "        table.append([drug, pve, se, N_snps])\n",
    "    \n",
    "    print(pve, se)\n",
    "    \n",
    "df = pd.DataFrame(table, columns=[\"drug\", \"pve\", \"se_pve\", \"N_snps\"])\n",
    "    \n",
    "df.to_csv(\"PVE_ABR_dep_sites_with_potentiators.csv\")\n",
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
